---
title: "Plots and Figures for Comparative Hummingbird Blood paper"
author: "Jessie Williamson"
date: "7/21/2020"
output:
  md_document:
    variant: markdown_github
---

#### 

PLOTS & FIGS SCRIPT 

####

This script contains exploratory plots, polished plots, and figures for allhummingbird_blood_comparison.Rmd. These are separated into a standalone script to not bog down main script. 


# Load packages
```{R}
library(reshape)
library(reshape2)
library(plyr)
library(dplyr)
library(car)
library(GGally)
library(Hmisc)
library(gridExtra)
library(stats)
library(gplots)
library(ggplot2)
library(stats4) # Forces knitr to work when it's being wonky
library(PMCMR) #Allows Kruskal-Wallis post-hocs
library(effects)
library(gridExtra)
library(lattice)
library(survival)
library(fmsb)
library(faraway)
library(ape)
library(plotly) # for 3D plots
library(viridis)
library(patchwork)
library(htmlwidgets)

# Mapping 
library(raster)
library(sp)
library(rgdal)
library(RStoolbox)
library(prettymapr)
library(viridis)
library(rasterVis)

# Modeling packages 
library(nlme)
library(lme4)
library(AICcmodavg)
library(MuMIn)
library(glmulti)
library(lsmeans)
library(rsq) # get r-squared values from GLM
library(r2glmm) # for R^2 values from lmer() and glmer()
library(multcompView) # related to multiple comparisons?
library(jtools) # interaction plots 
library(interactions) # interaction plots 
library(broom)
library(stargazer) # model output tables
library(ggeffects) # for estimating model predictions from mixed effects models
library(brms)
library(MCMCglmm)
library(rstan)
library(bayesplot)

# Phylo packages 
library(phytools)
```


---

# Clear workspace and set WD
```{R}
rm(list=ls(all=TRUE)) # clear workspace 
setwd("/Users/Jessie/Dropbox (MSBbirds)/Rdirectory/ComparativeHummingbirdBlood")
```



# load in data 
```{R}
# # Load functions 
# source("ada_functions.R") # Erik's ADA functions for clean & collated lm diag plots

# Final hummingbird blood dataset
# Don't need to load if already loaded from another script 
final <- read.csv("AllHummingbirdBlood_Wrangled_withBioClim_andPCA_03-01-21.csv")
final <- final[ , !(colnames(final) %in% c("X"))] # drop weird X column 1 if reading in from read.csv

spmeans <- read.csv("AllHummingbird_SpeciesMeanBloodData_03-01-21.csv")
final <- final[ , !(colnames(final) %in% c("X"))] # drop weird X column 1 if reading in from read.csv

# Patagona-free subset (for possible modeling of mass, where Patagona is a huge body size outlier):
final.nopgig <- read.csv("HumBlood.final.nopgig_03-01-21.csv")
final.nopgig <- final.nopgig[ , !(colnames(final.nopgig) %in% c("X"))] # drop weird X column 1 if reading in from read.csv

# Phylogeny pruned from McGuire and adjusted (see script HumBlood_Phylogeny.Rmd)
#tree <- read.tree("McGuirePruned_AllHummingbirdComparativeTree_FINAL.tre") # a list 
# tree$tip.label # Check 77 tips
# plotTree(tree, ftype="i") # Check this pre-final tree w/ adjusted branch lengths 
```


# Set levels to ensure that colors plot properly
Make sure levels correspond to clade label order in phylogeny from top to bottom
```{r}
# Set clade levels/make sure these are correct
final$Clade <- factor(final$Clade
                       , levels = c("Coquette",
                                    "Brilliant", 
                                    "Patagona", 
                                    "Bee", 
                                    "Mountain-gem", 
                                    "Emerald", 
                                    "Mango",  
                                    "Topaz",
                                    "Hermit"))
```



# COLOR PALETTES 
```{r}
# Colors from Figure 1 map phylo fig (last revised 2/23/21); see Illustrator file
# Turquoise - 4FBBB6 (coquette) **may need to change because very similar to points on the map
# Gray - A5A5AC (brilliant)
# Red - EE4024 (patagona)
# Gold - D9CA44 (bees)
# Pink - E786B7 (mountain gems)
# Green - 37B453 (emeralds)
# Orange - F9AC1B (mangos)
# Blue - 00A5CB (topazes)
# purple - 9E6EAE (hermits)

# HUMMINGBIRD COLORS 
hum_colors <- c( "#04A5A5", # Turquoise (Coquette)
                 "#9E9285", # brown (Brilliant)
                 "#EE4024", # red (Patagona)
                 "#F4CD0B", # Gold (Bee)
                 "#E786B7", # pink (Mountain-gem)
                 "#97B25B", # green (Emerald)  
                 "#F79E0F", # orange (Mango)
                 "#03446B", # navy blue (Topaz)
                 "#916EF9" # purple (Hermit)
                    )

# Old Hum colors 
# hum_colors <- c( "#4FBBB6", # Turquoise (Coquette)
#                  "#A5A5AC", # gray (Brilliant)
#                  "#EE4024", # red (Patagona)
#                  "#D9CA44", # Gold (Bee)
#                  "#E786B7", # pink (Mountain-gem)
#                  "#37B453", # green (Emerald)  
#                  "#F9AC1B", # orange (Mango)
#                  "#00A5CB", # blue (Topaz)
#                  "#9E6EAE" # purple (Hermit)
#                             )

# Oreotrochilus chimborazo color palette: 
# Deep purple: #6E1EDC
# Aqua gorget: #1CD2D8; another color: #0DAFA9
# pollen forehead: #D49A76
# gray patch brown: #665E54; another version #726A61; another version #8C847B
# deep green: #344D3F
# chuquiragua buds: #E2833F; #F4893C
# chuquiragua leaves: #6F8F17
# snowy breast: #D1D1D9
# Iridescent cheeks: #BA51FA
#scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87

# Ochimborazo_palette <- c( "#0E0E17", # Black 
#                           "#1CD2D8", # turquoise aqua gorget
#                           "#8C847B", # browny flanks 
#                           "#F4893C", # Chuquiragua buds 
#                           "#166975", # Darker part of gorget
#                           "#FDC31F", # Chuquiragua pollen  
#                           "#BA51FA", # Pinkish shine to gorget; Dark green wing patches: #344D3F
#                           "#6E1EDC", # Deep purple 
#                           "#6F8F17" # Chuquiragua leaves 
#                             )
# 
# peru_palette <- c( "#628454", # sage green
#                    "#FB4301", # deep orange-red
#                    "#6D7B9C", # skirt blue
#                    "#966D56", # Adobe brown
#                    "#E7860E", # golden orange-yellow
#                    "#264142", # Deep sea green
#                    "#BFD1F8", # periwinkle on shirt
#                    "#952842", # cabernet
#                    "#F871AA" # pink
#                             )
# 
# # Peru chimborazo: inspired by Peruvian textiles and Oreotrochilus chimborazo colors 
# peru_chimborazo <- c( "#FDC31F", # Chuquiragua pollen
#                       "#1CD2D8", # turquoise aqua gorget
#                       "#8C847B", # browny flanks 
#                       "#9AB243", # Chuquiragua gree
#                        "#FB4301", # Deep orange-red
#                       "#264142", # Deep sea green
#                       "#F871AA", # pink
#                       "#6E1EDC", # Deep purple 
#                       "#6D7B9C" # skirt blue # played around w/ subbing in this too:   "#A8A9AD" # gray breast
#                             )
```


# Basic plot of sampling across elevations
```{r}
# SUPER BASIC PLOTS, missing a lot of formatting/aesthetics 
# Sampling density across elevation 
(p <- ggplot(final, aes(x=elev, y=..density..)) +
      geom_density(alpha=.2, fill="#FF6666") + 
      geom_histogram(binwidth = 200) + 
    #  scale_color_manual(values = hum_colors) +
      theme_classic() + 
      labs(x="Elevation (m)", y="Sampling Density") + 
      theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )

# Sampling Depth across elevation  
(p <- ggplot(final, aes(x=elev)) +
   #   geom_density(alpha=.2, fill="#FF6666") + 
      geom_histogram(binwidth = 200) + 
    #  scale_color_manual(values = hum_colors) +
      theme_classic() + 
      labs(x="Elevation (m)", y="Sampling Depth") + 
      theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )

# Overlaid histograms
# ggplot(final, aes(x=elev)) +
#   geom_histogram(fill="black", alpha=0.5, position="identity")
# 
# ggplot(final, aes(elev, colour = Clade)) +
#   geom_freqpoly(binwidth = 500)
# 
# ggplot(final, aes(elev)) +
#   geom_histogram(bins = 100) + 
#   theme_classic() 
```



# FIGURE 2 - BLOOD CHARACTERISTICS WITH ELEVATION 
Figure 2 - 6-panel fig w/ relationship between predictors and elevation 
```{r}
# Scatterplots of each of the 6 blood characteristics with elevation, colored by clade

# Code to set shapes by clade, but this was totally overwhelming
  # scale_shape_manual(values=c("Bee"=4, "Brilliant"=17, "Coquette"=15, # Shapes are a little overwhelming, honestly
  #                             "Emerald"=18, "Hermit"=25, "Mango"=8, "Mountain-gem"=3,
  #                             "Patagona"=16, "Topaz"=9)) +
 #   scale_shape_manual(values=shapes) + # Another way to do this is set colors above 

# Scatterplot: Hb v. elev 
(p1 <- ggplot(final, aes(x=elev, y=hb, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="[Hb] g/dl") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p1, filename="p1_ScatterplotBasic_Elev_by_Hb.pdf", height=7, width=9, units="in")
# The 166 rows of missing values come from many NA Hb values; don't worry about these 

# Scatterplot: Hct v. elev
(p2 <- ggplot(final, aes(x=elev, y=hct, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  # scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p2, filename="p3_ScatterplotBasic_Elev_by_Hct.pdf", height=7, width=9, units="in")

# Scatterplot: TRBC v. elev
(p3 <- ggplot(final, aes(x=elev, y=trbc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
   labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     x="") + # Removed "Elevation (m)" because we don't want Panel A to have label
    # in expression world, ~ is space and mathematical symbols like parenthesis need to be set of w/ " " 
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p3, filename="p3_ScatterplotBasic_Elev_by_TRBC.pdf", height=7, width=9, units="in")

# Scatterplot: mcv v. elev
(p4 <- ggplot(final, aes(x=elev, y=mcv, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  #theme(legend.box.spacing = unit(c(0.2,5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p4, filename="p4_ScatterplotBasic_Elev_by_MCV.pdf", height=7, width=9, units="in")

# ANNOYING: space between plot area and legend is supposedly adjusted w/ "legend.box.spacing" arg but little documentation and it's
# being stubborn. I manually exported legend, will make 6-panel fig w/out legend and will add legend on in Illustrator

# Scatterplot: MCH v. elev
(p5 <- ggplot(final, aes(x=elev, y=mch, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", y="MCH (pg)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("E") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p5, filename="p5_ScatterplotBasic_Elev_by_MCH.pdf", height=7, width=9, units="in")


# Scatterplot: MCHC v. elev
(p6 <- ggplot(final, aes(x=elev, y=mchc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", y="MCHC (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("F") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.2,0.2,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p6, filename="p6_ScatterplotBasic_Elev_by_MCHC.pdf", height=7, width=9, units="in")


library(patchwork)
# Combine all plots to make one 6-panel figure:
# Fig2_BloodVElev_6panel <- ((p1 + p2) / (p3 + p4) / (p5 + p6) +  # the / literally means put p14 on top of p15
#                                   plot_layout(guides = "collect")) # "collates" and combines one common legend 
#                                   # In order for plot_layout to "collect" legend, must have printed legend for indiv plots above
# plot(Fig2_BloodVElev_6panel)
# ggsave(Fig2_BloodVElev_6panel, filename = "Fig2_BloodVElev_6panel_03-01-21.pdf", bg="transparent", height=7, width=7, units="in")
# This looks ok, but horizontal format (below) is better

# Alternative layout option = BETTER, USE THIS ONE: 
Fig2_BloodVElev_6panel_v2 <- (p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(guides = "collect"))
plot(Fig2_BloodVElev_6panel_v2)
ggsave(Fig2_BloodVElev_6panel_v2, filename = "Fig2_BloodVElev_6panel_v2_03-30-21.pdf", bg="transparent", height=7, width=11, units="in")
# Why I like this format: primary blood indices align clearly in Panels A-C; secondary indices on bottom in D-F

# Fig2_BloodVElev_6panel <- grid.arrange(p1,p2,p3,p4,p5,p6, ncol=2) # Old method 
```

Use patchwork to arrange plots: https://gotellilab.github.io/GotelliLabMeetingHacks/NickGotelli/ggplotPatchwork.html
Many say this is a better alternative to grid.arrange() 

Tutorial to using patchwork to make plots: https://patchwork.data-imaginist.com/articles/guides/layout.html



# FIGURE 2 WITH SPECIES MEANS - BLOOD CHARACTERISTICS WITH ELEVATION 
Figure 2 - 6-panel fig w/ relationship between predictors and elevation (USING SPECIES MEAN DATA)
```{r}
# Scatterplots of each of the 6 blood characteristics with elevation, colored by clade

# MAKE SURE THESE DATASETS ARE READ IN FROM SPECIES MEAN MODELING FILE OR UP TOP IN THIS SCRIPT! 

# Scatterplot: Hb v. elev 
(p21 <- ggplot(spm.hb, aes(x=elev.mean, y=hb.mean, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.6, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="[Hb] g/dl") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p21, filename="p21_ScatterplotBasic_Elev_by_Hb_SPM.pdf", height=7, width=9, units="in")
# The 166 rows of missing values come from many NA Hb values; don't worry about these 

# Scatterplot: Hct v. elev
(p22 <- ggplot(spm.hct, aes(x=elev.mean, y=hct.mean, colour=Clade)) + # Add shape=Clade here to shapes to work properly
  geom_point(size=1.6, alpha=0.8) +
  # scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p22, filename="p22_ScatterplotBasic_Elev_by_Hct_SPM.pdf", height=7, width=9, units="in")

# Scatterplot: TRBC v. elev
(p23 <- ggplot(spm.trbc, aes(x=elev.mean, y=trbc.mean, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.6, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
   labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     x="") + # Removed "Elevation (m)" because we don't want Panel A to have label
    # in expression world, ~ is space and mathematical symbols like parenthesis need to be set of w/ " " 
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p23, filename="p23_ScatterplotBasic_Elev_by_TRBC_SPM.pdf", height=7, width=9, units="in")

# Scatterplot: mcv v. elev
(p24 <- ggplot(spm.mcv, aes(x=elev.mean, y=mcv.mean, colour=Clade)) + # shape=Clade here to get shapes to work properly
  geom_point(size=1.6, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  #theme(legend.box.spacing = unit(c(0.2,5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p24, filename="p24_ScatterplotBasic_Elev_by_MCV_SPM.pdf", height=7, width=9, units="in")

# ANNOYING: space between plot area and legend is supposedly adjusted w/ "legend.box.spacing" arg but little documentation and it's
# being stubborn. I manually exported legend, will make 6-panel fig w/out legend and will add legend on in Illustrator

# Scatterplot: MCH v. elev
(p25 <- ggplot(spm.mch, aes(x=elev.mean, y=mch.mean,  colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.6, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", y="MCH (pg)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("E") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p25, filename="p25_ScatterplotBasic_Elev_by_MCH_SPM.pdf", height=7, width=9, units="in")


# Scatterplot: MCHC v. elev
(p26 <- ggplot(spm.mchc, aes(x=elev.mean, y=mchc.mean, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.6, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", y="MCHC (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("F") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.2,0.2,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p26, filename="p26_ScatterplotBasic_Elev_by_MCHC_SPM.pdf", height=7, width=9, units="in")


library(patchwork)
# Combine all plots to make one 6-panel figure:
# Fig2_BloodVElev_6panel <- ((p1 + p2) / (p3 + p4) / (p5 + p6) +  # the / literally means put p14 on top of p15
#                                   plot_layout(guides = "collect")) # "collates" and combines one common legend 
#                                   # In order for plot_layout to "collect" legend, must have printed legend for indiv plots above
# plot(Fig2_BloodVElev_6panel)
# ggsave(Fig2_BloodVElev_6panel, filename = "Fig2_BloodVElev_6panel_03-01-21.pdf", bg="transparent", height=7, width=7, units="in")
# This looks ok, but horizontal format (below) is better

# Alternative layout option = BETTER, USE THIS ONE: 
Fig2_BloodVElev_6panel_Species_mean <- (p21 + p22 + p23 + p24 + p25 + p26 + plot_layout(guides = "collect"))
plot(Fig2_BloodVElev_6panel_Species_mean)
ggsave(Fig2_BloodVElev_6panel_Species_mean, filename = "Fig2_BloodVElev_6panel_Species_mean_03-29-21.pdf", bg="transparent", height=7, width=11, units="in")
# Why I like this format: primary blood indices align clearly in Panels A-C; secondary indices on bottom in D-F

# Fig2_BloodVElev_6panel <- grid.arrange(p1,p2,p3,p4,p5,p6, ncol=2) # Old method 
```




# FIGURE 3: PLOTS FOR RELATIONSHIPS AMONG BLOOD PARAMETERS
Quick workup of predictors: 
```{r}
# full dataset
p <- ggpairs(subset(final, select = c(elev, mass, hb, hct, trbc, mcv, mch, mchc))) 
print(p)

# Subset final dataset to just have hb, mcv, trbc, mch
final.hmt <- subset(final, select=c(species, rowID, nk, elev, hb, mcv, trbc, mch)) 
final.hmt <- na.omit(final.hmt)

# For plotting blood-O2 carrying capacity as background of plot, make variable called o2cc
# test <- final %>% mutate(o2cc = mcv * trbc * (1/3))
# As it turns out, we don't need this; see plot below
```

Mock up of Figure 3, relationships among blood parameters 
See Hawkey et al. 1991 for inspiration.
```{r}
# Need to remove NA values from 'final', but need to subset to hb, trbc, mcv first; running naomit on 'final' wipes all
# observations because so many NA values throughout the dataset

# Panel A: RBC and MCV - redo of plot in Hawkey et al. 1991
(p14 <- ggplot(final.hmt, aes(x=mcv, y=trbc, colour=hb)) +
      scale_colour_viridis(option="rocket", discrete=FALSE, name="[Hb] (g/dl)") +
     # p <- p + geom_boxplot(alpha = 0.5, outlier.size=0.1) # Add this back if categorical x-axis variable 
    geom_point(size=2, alpha=1.0) +
    # scale_color_manual(values=hum_colors) +
    # stat_smooth(data=final, aes(y=trbc), size=1.2, method="lm") +
    #geom_jitter(width=0.8, height=0.8, alpha=0.8, size=2) + # jitter size should match geom_point size or it looks weird
    labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     x="") + # Removed x-axis label because we don't want Panel A to have label
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    #scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
    theme_classic() +
    theme(legend.position = "right") + 
    ggtitle("A") + # Assign panel number/header
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + #makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.2,0.2,0.0,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p14, filename="Fig3_PanelA_RBC_vs_MCV_Scatterplot_ColoredByElev.pdf", height=7, width=9, units="in") 
# Wow, this is amazing. See Hawkey et al. 1991 for some dynamite interpretation
# Essentially this means: species have either evolved many small erythrocytes or few large erythrocytes 
# Seems to be a sweet spot of a moderate number of a moderate sized erythrocyte

# Panel B: MCH and MCV - redo of plot in Hawksey et al. 1991
(p15 <- ggplot(final.hmt, aes(x=mcv, y=mch, colour=elev)) +
    scale_colour_viridis(option="rocket", discrete=FALSE, name="Elevation (m)") +
    # p <- p + geom_boxplot(alpha = 0.5, outlier.size=0.1) # Add this back if categorical x-axis variable 
    geom_point(size=2, alpha=1.0) +
    # stat_smooth(data=full1, aes(y=trbc), size=1.2, method="lm") +
    # geom_jitter(width=0.8, height=0.8, alpha=0.8, size=2) + # jitter size should match geom_point size or it looks weird
    labs(y="MCH (pg)", 
         x="MCV (fl)") + # fix RBC title/notation
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    #scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
    theme_classic() +
    theme(legend.position = "right") + # Note: orange = spring, blue = winter 
    ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.0,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p15, filename="Fig3_PanelB_RBC_vs_MCH_Scatterplot_ColoredByElev.pdf", height=7, width=9, units="in") 


# 2 PANEL FIGURE 
Fig3_BloodComparison_2panel_v2 <- (p14 / p15 +   # the / literally means put p14 on top of p15
                                  plot_layout(guides = "collect")) # "collates" and combines one common legend
                                  # In order for plot_layout to "collect" legend, must have printed legend for indiv plots above
plot(Fig3_BloodComparison_2panel_v2)
ggsave(Fig3_BloodComparison_2panel_v2, filename = "Fig3_BloodComparison_2panel_ColoredByElev_04-26-21.pdf", bg="transparent", height=6, width=5, units="in")

# Old method for combining plots, but I was having trouble getting a common legend
# Fig3_BloodComparison_2panel_v2 <- grid.arrange(p14, p15, ncol=1)
```


# ALTERNATIVE FIGURE 3 WITH HEAT MAP OF BLOOD O2 CARRYING CAPACITY
Creating a heat map/color gradient background in ggplot isn't super straightforward. Through a back-channel google images search, I ended up finding this advice, which is exactly what we want: 
https://stackoverflow.com/questions/46354513/diagonal-gradient-in-a-ggplot-background
(Believe it or not, google image searching "ggplot panel.background color gradient" was more helpful than web searching)
```{r}
# PANEL A: MCH v. MCV, no color to points (yes, this was panel B from previous figure version)
(p16 <- ggplot(final.hmt, aes(x=mcv, y=mch)) +
    # scale_colour_viridis(option="rocket", discrete=FALSE, name="Elevation (m)") +
    # p <- p + geom_boxplot(alpha = 0.5, outlier.size=0.1) # Add this back if categorical x-axis variable 
    geom_point(size=1.8, alpha=0.8) +
    # stat_smooth(data=full1, aes(y=trbc), size=1.2, method="lm") +
    # geom_jitter(width=0.8, height=0.8, alpha=0.8, size=2) + # jitter size should match geom_point size or it looks weird
    labs(y="MCH (pg)", 
         x=" ") + 
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    #scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
    theme_classic() +
    theme(legend.position = "right") + # Note: orange = spring, blue = winter 
    ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.0,0.2,0.0,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p16, filename="Fig3_v2_PanelA_RBC_vs_MCH_Scatterplot.pdf", height=7, width=9, units="in") 


# PANEL B: TRBC v. MCV w/ background shaded by heat map of blood O2 carrying capacity 
# To get a heat map of blood-O2 carrying capacity, we need to make a contour of circles to plot behind our data 
# Formula for O2 carrying capacity is: oc22 = MCV x TRBC x 1/3
#min(final$mcv) * min(final$trbc) * (1/3) # min o2cc; can't set this up this way because NAs, I think
o2cc.min <- 58.48877 * 3.58 * (1/3); o2cc.min # This is min MCV * min TRBC * 1/3
o2cc.max <- 147.6209 * 10.130 * (1/2); o2cc.max # Max MCV * max TBBC * 1/3

# Generate heat map color gradient of values: 
xColor <- seq(from=55, to=557.3333, length.out=100) # scale of color for x; length.out = desired length of sequence
yColor <- seq(from=55, to=557.3333,length.out=100) # scale of color for y 
  # For both, min value WAX min(mcv) * min(trbc) * 1/3; max = max(mcv) * max(trbc) * 1/3
  # But that just gives us min/max values for our data; we want min/max for plot space 
  # so multiply min(ylim)*min(xlim)*1/3 and max(ylim)*max(xlim)*1/3 to get plot space dimensions
  # For xColor: 55*3*1/3; For yColor: 152*11*1/3
x <- seq(from=55, to=152, length.out=100) # on x axis; min MCV: 58.48877; max MCV: 147.6209
y <- seq(from=3, to=11, length.out=100) # on y axis; min TRBC: 3.58; max TRBC: 10.130
  # This is adjusting the plot space that the background will fill 
  # Essentially, this is xlim and ylim + buffer to expand beyond min and max values to fill entire plot space 
df <- cbind(expand.grid(x=xColor, y=yColor), expand.grid(x=x, y=y)) #grid for colors
colnames(df)<-c("xColor", "yColor", "x", "y")
df$zColor <- (df$xColor+df$yColor) # the color factor for radius; tutorial used ^2 for each, but we don't want that scale
  # Adding xColor+yColor without xColor^2+yColor^2 plots real x-y oxygen carrying capacity values

# Color gradient banana plot 
(p17 <- ggplot(final.hmt, aes(x=mcv, y=trbc)) +
    geom_raster(data =df, aes(x, y,fill = zColor))+
    scale_fill_viridis(option="rocket", discrete=FALSE, direction=1, name="O2\ncarrying\ncapacity")+ 
      # direction=1 goes cool to warm; direcdtion=-1 goes warm to cool
    # scale_colour_viridis(option="rocket", discrete=FALSE, name="[Hb] (g/dl)") +
    geom_point(size=1.8, alpha=0.8, colour="black") + # aliceblue, ivory1 and snow1 are also nice, slightly warmer
    labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
         x="MCV (fl)") + # Removed x-axis label because we don't want Panel A to have label
    theme_classic() +
   # xlim(56, 150) + ylim(2.58, 11)+
    coord_cartesian(expand = FALSE) + # FALSE = no padding for border
    theme(legend.position = "right") + 
    ggtitle("B") + # Assign panel number/header
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + #makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.0,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p17, filename="Blood_O2_CarryingCapacity_HeatMapPlot_04-26-21.pdf", height=7, width=9, units="in") 


# 2 PANEL FIGURE 
Fig3_BloodComparison_2panel_v3 <- (p16 / p17)     # the / literally means put p16 on top of p17
                                #   plot_layout(guides = "collect")) # "collates" and combines one common legend
                                  # In order for plot_layout to "collect" legend, must have printed legend for indiv plots above
plot(Fig3_BloodComparison_2panel_v3)
ggsave(Fig3_BloodComparison_2panel_v3, filename = "Fig3_v3_BloodComparison_2panel_WithHeatMap_04-26-21.pdf", bg="transparent", height=6, width=5, units="in")
```



# 3D plots
We abandoned these attempts because they don't provide meaningful information about relationships. See rest of tinkering with 3D plots in the CUT script. 
```{r}
# library(plotly)
# # for playing around with colors: 
# # library(scales)
# # show_col(viridis_pal(option="B")(10)) # show all colors in viridis palette 
# 
# # 3d plot of elev, mcv, trbc
# fig <- plot_ly(final, x = ~hb, y = ~mcv, z = ~trbc, color = ~elev, colors = viridis_pal(option = "F")(100)) # rocket
# fig <- fig %>% add_markers()
# fig <- fig %>% layout(scene = list(xaxis = list(title = '[Hb] (g/dl)'),
#                      yaxis = list(title = 'MCV (fl)'),
#                      zaxis = list(title = 'TRBC')))
# fig
# ggsave(fig, filename="test.pdf", height=7, width=9, units="in")
# 
# library(htmlwidgets)
# htmlwidgets::saveWidget(fig, "3D_hb_mcv_trbc.html")
# # plotly.io.write_html("/Users/Jessie/Dropbox (MSBbirds)/Rdirectory/ComparativeHummingbirdBlood/3D_hb_mcv_trbc.html")
# # Couldn't get plotly save code to work
# 
# # 3d plot of elev, mcv, trbc - SPECIES MEAN DATA 
# fig <- plot_ly(cens.pca.spm.all, x = ~hb.mean, y = ~mcv.mean, z = ~trbc.mean, color = ~elev.mean, colors = viridis_pal(option = "F")(100)) # rocket, new palette
# fig <- fig %>% add_markers()
# fig <- fig %>% layout(scene = list(xaxis = list(title = '[Hb] (g/dl)'),
#                      yaxis = list(title = 'MCV (fl)'),
#                      zaxis = list(title = 'TRBC')))
# fig
# # Need to figure out how to reverse Hb axis (because "-hb" DOES reverse, but it also makes numbers negative)
# htmlwidgets::saveWidget(fig, "3D_hb_mcv_trbc.html")
```


# Trial: Sideways plot of species x TRBC w/ points sized by MCV and colored by Hb
...like a species dot plot 
From this Stack Exchange tutorial: https://stackoverflow.com/questions/33169149/how-do-i-visualize-a-3-dimensional-matrix
```{r}
library("ggplot2") 
theme_set(theme_bw())
library("viridis")

(p30 <- ggplot(final.hmt) +
    geom_point(aes(x=trbc, y=species, colour=hb, size=mcv)) +
    #facet_wrap(~Clade, ncol=3) + 
   # scale_color_viridis() +
    scale_colour_viridis(option="magma", discrete=FALSE) +
    scale_size(range=c(0.4,6)) + 
    # labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis
    #  x="") + # Removed x-axis label because we don't want Panel A to have label
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    #scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
   # theme_classic() +
    theme_set(theme_bw()) + 
    theme(legend.position = "right") + 
    #ggtitle("A") + # Assign panel number/header
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + #makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.2,0.2,0.0,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p30, filename="p30_Species_Hb_MCV_TRBC.pdf", height=7, width=9, units="in") 

# Make a final.sub dataset that drops sampling outliers and considers Patagona as two functional species: 
final.sub <- final # Make the final dataset copy that you'll use to wrangle species mean data 
sp.count.final.sub <- final.sub %>% group_by(species) %>% summarise(count = length(nk))

# Loop through dataset and assign "Patagona_gigas_S" to Chilean birds and "Patagona_gigas_N" to Peru birds
for(i in 1:nrow(final.sub)){
    if((final.sub$species[i] == "Patagona_gigas") & (final.sub$department[i] == "Región Valparaíso") ){
        final.sub$species[i] <- "Patagona_gigas_S"
    }else if((final.sub$species[i] == "Patagona_gigas")){ 
      final.sub$species[i] <- "Patagona_gigas_N"
    }
}

# Drop sampling outliers (n=1); I don't have a total_samples column in final dataset, so doing this the clunky way: 
final.sub <- final.sub[-which(final.sub$species == "Amazilia_franciae"),] 
final.sub <- final.sub[-which(final.sub$species == "Calliphlox_amethystina"),] 
final.sub <- final.sub[-which(final.sub$species == "Chaetocercus_mulsant"),] 
final.sub <- final.sub[-which(final.sub$species == "Heliodoxa_rubinoides"),] 
final.sub <- final.sub[-which(final.sub$species == "Heliomaster_longirostris"),] 
final.sub <- final.sub[-which(final.sub$species == "Myrmia_micrura"),] 
final.sub <- final.sub[-which(final.sub$species == "Phaethornis_atrimentalis"),] 
final.sub <- final.sub[-which(final.sub$species == "Phaethornis_ruber"),] 
final.sub <- final.sub[-which(final.sub$species == "Phaethornis_stuarti"),] 

# subset to be able to remove NAs
final.sub2 <- subset(final.sub, select=c(species, rowID, nk, elev, hb, mcv, trbc, mch)) 
final.sub2 <- na.omit(final.sub2)

(p30 <- ggplot(final.sub2) +
    geom_point(aes(x=trbc, y=species, colour=hb, size=mcv)) +
    #facet_wrap(~Clade, ncol=3) + 
    scale_colour_viridis(option="rocket", discrete=FALSE) +
    scale_size(range=c(0.4,6)) + 
    labs(x=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for TRBC units
      y="Species") + 
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    #scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
   # theme_classic() +
    coord_cartesian(expand = TRUE) + # FALSE = no padding for border
    theme_set(theme_bw()) + 
    theme(legend.position = "right") + 
    #ggtitle("A") + # Assign panel number/header
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + #makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.2,0.2,0.0,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p30, filename="p30_Species_Hb_MCV_TRBC.pdf", height=8, width=7, units="in") 
# Gray points indicate no Hb; just TRBC and MCV (these shouldn't be in there now because I did NA omit)
# If opting to go this direction, species should really be orderd in phylogenetic order 
```


***All radar chart stuff has been moved to CUT script 


# PLOTS OF BLOOD CHARACTERISTICS WITH ELEVATION_POSITION 
****THIS IS FIG. 2 redone w/ elev position! 
```{r}
# Scatterplot: Hb v. elev 
(p16 <- ggplot(final, aes(x=elev_position, y=hb, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="[Hb] g/dl") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p16, filename="p16_ScatterplotBasic_ElevPos_by_Hb.pdf", height=7, width=9, units="in")
# The 166 rows of missing values come from many NA Hb values; don't worry about these 

# Scatterplot: Hct v. elev
(p17 <- ggplot(final, aes(x=elev_position, y=hct, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  # scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p17, filename="p17_ScatterplotBasic_ElevPosition_by_Hct.pdf", height=7, width=9, units="in")

# Scatterplot: TRBC v. elev
(p18 <- ggplot(final, aes(x=elev_position, y=trbc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
   labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     x="") + # Removed "Elevation (m)" because we don't want Panel A to have label
    # in expression world, ~ is space and mathematical symbols like parenthesis need to be set of w/ " " 
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p18, filename="p18_ScatterplotBasic_ElevPosition_by_TRBC.pdf", height=7, width=9, units="in")

# Scatterplot: mcv v. elev
(p19 <- ggplot(final, aes(x=elev_position, y=mcv, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation Position", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  #theme(legend.box.spacing = unit(c(0.2,5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p19, filename="p19_ScatterplotBasic_ElevPosition_by_MCV.pdf", height=7, width=9, units="in")

# Scatterplot: MCH v. elev_pos
(p20 <- ggplot(final, aes(x=elev_position, y=mch, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation Position", y="MCH (pg)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("E") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p20, filename="p20_ScatterplotBasic_ElevPosition_by_MCH.pdf", height=7, width=9, units="in")

# Scatterplot: MCHC v. elev_pos
(p21 <- ggplot(final, aes(x=elev_position, y=mchc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation Position", y="MCHC (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("F") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.2,0.2,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p21, filename="p21_ScatterplotBasic_ElevPosition_by_MCHC.pdf", height=7, width=9, units="in")


library(patchwork)
# Combine all plots to make one 6-panel figure:
# Fig2_BloodVElev_6panel <- ((p1 + p2) / (p3 + p4) / (p5 + p6) +  # the / literally means put p14 on top of p15
#                                   plot_layout(guides = "collect")) # "collates" and combines one common legend 
#                                   # In order for plot_layout to "collect" legend, must have printed legend for indiv plots above
# plot(Fig2_BloodVElev_6panel)
# ggsave(Fig2_BloodVElev_6panel, filename = "Fig2_BloodVElev_6panel_03-01-21.pdf", bg="transparent", height=7, width=7, units="in")
# This looks ok, but horizontal format (below) is better

# Alternative layout option = BETTER, USE THIS ONE: 
Fig_BloodVElevPosition_6panel <- (p16 + p17 + p18 + p19 + p20 + p21 + plot_layout(guides = "collect"))
plot(Fig_BloodVElevPosition_6panel)
ggsave(Fig_BloodVElevPosition_6panel, filename = "Fig_BloodVElevPosition_6panel_03-01-21.pdf", bg="transparent", height=7, width=11, units="in")
# Why I like this format: primary blood indices align clearly in Panels A-C; secondary indices on bottom in D-F
```



JESSIE WORKING HERE 

# Trial with species mean values 
If you want to do this, need to read in species mean value dataset from modeling rmd file 
```{r}
library(dplyr)

test1 <- final %>% group_by(species) %>% dplyr::summarize(mean.mcv = mean(mcv, na.rm = TRUE), count =n())
test2 <- final %>% group_by(species) %>% dplyr::summarize(mean.elev = mean(elev, na.rm = TRUE))
test2 <- test2[ , !(colnames(test2) %in% c("species"))] # drop species col
test3 <- cbind(test1, test2)
                                  
# plot species mean MCV and species mean elevation; for most species it will be ok

#  Scatterplot: Hb v. elev
(p <- ggplot(test3, aes(x=mean.elev, y=mean.mcv)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.5, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) +
 # scale_color_manual(values = hum_colors) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # Don't jitter; creates the appearance of duplicate points
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), size = 1,colour="firebrick3", size=1.0) + # line used to be dark turquoise
  theme_classic() +
  labs(x="Mean Elev (m)", y="Mean MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("(MCV quadratic)") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
#ggsave(p, filename="p1_ScatterplotBasic_Elev_by_Hb.pdf", height=7, width=9, units="in")


# all indivdual data
# Scatterplot: mcv v. elev
(p4 <- ggplot(final, aes(x=elev, y=mcv, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  #theme(legend.box.spacing = unit(c(0.2,5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p4, filename="p4_ScatterplotBasic_Elev_by_MCV.pdf", height=7, width=9, units="in")


# Scatterplot: Hb v. elev 
# (p <- ggplot(species.means, aes(x=mean_mass, y=mean_hb, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
#   geom_point(size=1.5, alpha=0.8) +
#   #scale_colour_viridis(discrete=TRUE) + 
#   scale_color_manual(values = hum_colors) +
#  # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # Don't jitter; creates the appearance of duplicate points
#   geom_smooth(method = "lm", colour="firebrick3", size=1.0) + # line used to be dark turquoise 
#   theme_classic() + 
#   labs(x="Mass (g)", y="[Hb] g/dl") +
#        # theme(axis.text.x = element_text(hjust = 1)) +
#   theme(legend.position = "right") + 
#   ggtitle("(a)") + # Assign panel number/header; this will be (a) because first in series of 3
#   theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
#          plot.title = element_text(face="bold")) + # This makes panel header bold 
#        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
#   theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
#  )
#ggsave(p, filename="p1_ScatterplotBasic_Elev_by_Hb.pdf", height=7, width=9, units="in")
```



# PRELIM DATA EXPLORATION 
```{r}

# Boxplot of differences in hb w/ species
(p22 <- ggplot(final, aes(x=Clade, y=hb)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(Clade)), outlier.size=2.0, alpha=0.7) +
        geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=6) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        labs(y="[Hb] g/dl", # Hella confusing formatting for x and y axis labels 
             x="Species") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
#ggsave(p1, filename="p1_Boxplot_Hb_by_Clade.pdf", height=7, width=9, units="in")

# Boxplot of differences in hb w/ genotype
(p23 <- ggplot(final, aes(x=b13b83.genotype, y=hb)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(b13b83.genotype)), outlier.size=2.0, alpha=0.7) +
        geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=6) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        labs(y="[Hb] g/dl", # Hella confusing formatting for x and y axis labels 
             x="Species") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
#ggsave(p1, filename="p1_Boxplot_Hb_by_Clade.pdf", height=7, width=9, units="in")
 


# Extremely basic plot of elev range summary
# Maybe interesting to get mean Hb value per species and plot range on x-axis and mean Hb on y? 
plot(elev.range.summary)

# Basic scatterplot of mass v. Hb 
(p <- ggplot(final, aes(x=mass, y=hb, colour=Clade)) + # uncenter Tamin values
  geom_point() +
  scale_colour_viridis(discrete=TRUE) +
  geom_smooth(method = "lm", colour="darkturquoise", size=1.2) + 
  theme_classic() + 
  labs(x="Mass (g)", y="[Hb] g/dl") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  # ggtitle("(a)") + # Assign panel number/header; this will be (a) because first in series of 3
  # theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
  #       plot.title = element_text(face="bold")) + # This makes panel header bold 
  #     # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p3, filename="p3_ScatterplotBasic_Mass_by_Hb.pdf", height=7, width=9, units="in")
# This is hilarious: this plot is like "Patagona"...and everything else. 
# This brings up questions about how to analyze mass data w/ Patagona involved and whether this


# Basic scatterplot of intraspecific variation in mass (note: need to have read in modeling steps)
(p <- ggplot(final, aes(x=elev, y=intravar.mass, colour=Clade)) + # uncenter Tamin values
  geom_point() +
  scale_colour_viridis(discrete=TRUE) +
  geom_smooth(method = "lm", colour="darkturquoise", size=1.2) + 
  theme_classic() + 
  labs(x="Elevation (m)", y="Intraspecific Mass Variation (g)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") +
  # ggtitle("(a)") + # Assign panel number/header; this will be (a) because first in series of 3
  # theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
  #       plot.title = element_text(face="bold")) + # This makes panel header bold 
  #     # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p3, filename="p3_ScatterplotBasic_Mass_by_Hb.pdf", height=7, width=9, units="in")
# This is hilarious: this plot is like "Patagona"...and everything else. 
# This brings up questions about how to analyze mass data w/ Patagona involved and whether this 



# Basic scatterplot of mass v. Hb 
(p <- ggplot(final.nopgig, aes(x=mass, y=hb)) + # uncenter Tamin values
  geom_point() +
  geom_smooth(method = "lm", colour="darkturquoise", size=1.2) + 
  theme_classic() + 
  labs(x="Mass (g)", y="[Hb] g/dl") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  # ggtitle("(a)") + # Assign panel number/header; this will be (a) because first in series of 3
  # theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
  #       plot.title = element_text(face="bold")) + # This makes panel header bold 
  #     # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
   )
ggsave(p4, filename="p4_ScatterplotBasic_Mass_by_Hb_NOPATAGONA.pdf", height=7, width=9, units="in")
# Def interesting trend of increasing Hb w/ mass 


# Break these plots out by species (I prefer clade; ~species is too much and a shitshow to look at)
# not sure this is too useful except to show sampling across elev gradient 
(p5 <- ggplot(final, aes(x=elev, y=hb)) +
  facet_wrap(~Clade, scales="free") +
  geom_point(pch=21, stroke=1, aes(color=Clade), show.legend = FALSE) +
  geom_smooth(method="lm", se=FALSE, linetype="dashed", color="black") +
  theme_classic() +
  labs(x="Elevation (m)", y="[Hb] g/dl") +
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p5, filename="p5_Scatterplot_Elev_by_Hb_byCLADE.pdf", height=7, width=9, units="in")


# now just by po2 specialists: 
# Break these plots out by species 
# not sure this is too useful except to show sampling across elev gradient 
(p5.1 <- ggplot(po2.specialist, aes(x=elev, y=hb)) +
  facet_wrap(~species, scales="free") + # split by species 
  geom_point(pch=21, stroke=1, aes(color=clade), show.legend = FALSE) + # color by clade 
  geom_smooth(method="lm", se=FALSE, linetype="dashed", color="black") +
  theme_classic() +
  labs(x="Elevation (m)", y="[Hb] g/dl") +
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p5.1, filename="p5.1_Scatterplot_Elev_by_Hb_PO2-specialists.pdf", height=7, width=9, units="in")


# now just by po2 generalists: 
# Break these plots out by species 
# not sure this is too useful except to show sampling across elev gradient 
(p5.2 <- ggplot(po2.generalist, aes(x=elev, y=hb)) +
  facet_wrap(~species, scales="free") + # split by species 
  geom_point(pch=21, stroke=1, aes(color=clade), show.legend = FALSE) + # color by clade 
  geom_smooth(method="lm", se=FALSE, linetype="dashed", color="black") +
  theme_classic() +
  labs(x="Elevation (m)", y="[Hb] g/dl") +
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p5.2, filename="p5.2_Scatterplot_Elev_by_Hb_PO2-generalists.pdf", height=7, width=9, units="in")
# Huh, interesting for C. coruscans and R. vesper: both are generalists because found <1000 m
# but we've never actually sampled them at lowest elevs, hence both starting on x-axis here >2000 m




# Break these plots out by species but remove colors to simplify 
# not sure this is too useful except to show sampling across elev gradient 
(p6 <- ggplot(final, aes(x=elev, y=hb)) +
  facet_wrap(~Clade, scales="free") +
  geom_point(stroke=1, show.legend = FALSE) +
  geom_smooth(method="lm", se=FALSE, color="turquoise") +
  theme_classic() +
  labs(x="Elevation (m)", y="[Hb] g/dl") +
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p6, filename="p2_Scatterplot_Elev_by_Hb_byCLADE_simple.pdf", height=7, width=9, units="in")



# Boxplot of differences in hb w/ species
(p7 <- ggplot(final, aes(x=sex, y=hb)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(sex)), outlier.size=2.0, alpha=0.7) +
        geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=6) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        labs(y="[Hb] g/dl", # Hella confusing formatting for x and y axis labels 
             x="Sex") +
   #    labs(x="Season",y="Tbmin (°C)") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
  #      ggtitle("(d)") + # Assign panel number/header
  #      theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
   #           plot.title = element_text(face="bold")) + # This makes panel header bold 
      # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p7, filename="p7_Boxplot_Hb_by_Sex.pdf", height=7, width=9, units="in")
# Good news is that Hb pretty similar between sexes 
# Might consider removing unknowns to plot this cleanly


# Hb differences w/ age boxplot 
(p8 <- ggplot(final, aes(x=age, y=hb)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(age)), outlier.size=2.0, alpha=0.7) +
        geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=6) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        labs(y="[Hb] g/dl", # Hella confusing formatting for x and y axis labels 
             x="Sex") +
   #    labs(x="Season",y="Tbmin (°C)") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
  #      ggtitle("(d)") + # Assign panel number/header
  #      theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
   #           plot.title = element_text(face="bold")) + # This makes panel header bold 
      # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p8, filename="p8_Boxplot_Hb_by_Age.pdf", height=7, width=9, units="in")
# Needs massaging to clean this up
# Need to assign NAs to unknowns
# Adults a bit higher than juvs, but not sure differences are significant (they might be tho)
# Add age as a random effect in models? 


(p9 <- ggplot(final, aes(x=locality, y=hb)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(locality)), outlier.size=2.0, alpha=0.7) +
        geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=6) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        labs(y="[Hb] g/dl", # Hella confusing formatting for x and y axis labels 
             x="Locality") +
   #    labs(x="Season",y="Tbmin (°C)") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
  #      ggtitle("(d)") + # Assign panel number/header
  #      theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
   #           plot.title = element_text(face="bold")) + # This makes panel header bold 
      # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(p9, filename="p9_Boxplot_Hb_by_Locality.pdf", height=7, width=9, units="in")
# As we'd expect, lots of locality differences; notably, Algarrobo (point 4) is super low! 

# Is there some sort of latitude effect????

# Oreotrochilus chimborazo color palette: 
# Deep purple: #6E1EDC
# Aqua gorget: #1CD2D8; another color: #0DAFA9
# pollen forehead: #D49A76
# gray patch brown: #665E54; another version #726A61; another version #8C847B
# deep green: #344D3F
# chuquiragua buds: #E2833F
# chuquiragua leaves
# snowy breast: #D1D1D9
scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87


# INTERACTION PLOT -GENERALIST SPECIALIST HB v. ELEV
(p10 <- ggplot(final, aes(x=elev, y=hb, colour=po2_class, linetype=po2_class)) +
# p <- p + geom_boxplot(alpha = 0.5, outlier.size=0.1) # Add this back if categorical x-axis variable 
    geom_point(data=final, aes(y=hb), size=2, alpha=0.8) +
    stat_smooth(data=final, aes(y=hb, group=po2_class), size=1.2, method="lm") +
    geom_jitter(width=0.8, height=0.8, alpha=0.8, size=2) + # jitter size should match geom_point size or it looks weird
    labs(y="[Hb] g/dl", x="Elevation (m)") +
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
    theme_classic() +
    theme(legend.position = "right") + # Note: orange = spring, blue = winter 
   # ggtitle("(a)") + # Assign panel number/header
  #  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
  #        plot.title = element_text(face="bold")) + # This makes panel header bold 
    # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p10, filename="p10_Hb_vs_Elev_Scatterplot_PO2GeneralistSpecialist.pdf", height=7, width=9, units="in") 



# Now subset by just generalist and specialist and see what this looks like 
genspec <- final[which(final$po2_class == "generalist" | final$po2_class=="specialist"),]

# INTERACTION PLOT -GENERALIST SPECIALIST ONLY SUBSET HB v. ELEV
(p11 <- ggplot(genspec, aes(x=elev, y=hb, colour=po2_class, linetype=po2_class)) +
# p <- p + geom_boxplot(alpha = 0.5, outlier.size=0.1) # Add this back if categorical x-axis variable 
    geom_point(data=genspec, aes(y=hb), size=2, alpha=0.8) +
    stat_smooth(data=genspec, aes(y=hb, group=po2_class), size=1.2, method="lm") +
    geom_jitter(width=0.5, height=0.5, alpha=0.5, size=2) + # jitter size should match geom_point size or it looks weird
    labs(y="[Hb] g/dl", x="Elevation (m)") +
    scale_color_manual(values=c(generalist="#ffb20a", specialist="#853FCD")) + #336B87 (dark)
    theme_classic() +
    theme(legend.position = "right") + # Note: orange = spring, blue = winter 
   # ggtitle("(a)") + # Assign panel number/header
  #  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
  #        plot.title = element_text(face="bold")) + # This makes panel header bold 
    # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=14))
)
ggsave(p11, filename="p11_Hb_vs_Elev_Scatterplot_PO2GeneralistSpecialistONLY.pdf", height=7, width=9, units="in") 

# light blue: #90AFC5; R 144, G 175, B 197
# dark teal: #336B87; R51, G107, B135
# purple orchid: #853FCD; R133, G63, B205


# Elev and Hb, faceted by species 
(p2 <- ggplot(final, aes(x=elev, y=hb)) + 
       facet_wrap(~species) +   # ok, way too many species right now to see any sort of clear pattern
       geom_point() +
       geom_smooth(method = "lm", colour="darkturquoise", size=1.2) + 
       theme_classic() + 
       labs(x="Elevation (m)", y="[Hb] g/dl") +
       # theme(axis.text.x = element_text(hjust = 1)) +
       theme(legend.position = "none") +
      # ggtitle("(a)") + # Assign panel number/header; this will be (a) because first in series of 3
      # theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
      #       plot.title = element_text(face="bold")) + # This makes panel header bold 
      #     # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
       theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
        )
#ggsave(p2, filename="m1.p1_TbMin_Vs_TaMin_Scatterplot_Simple.pdf", height=7, width=9, units="in")

# Boxplot of Hb v. sex w/ elevation
boxplot (hb ~ sex, data=final, xlab="Elevation (m)", ylab="[Hb] (g/dl)", boxwex=0.3, ylim=c(14,26))
stripchart(hb ~ sex, vertical = TRUE, data=full, method = "jitter", add = TRUE, pch = 20, col=c("gold", "purple", "olivedrab3"))
  # rgb(0.85,0.6,0,0.56),(rgb(0,0,0.95,0.4))))
# anovaHWIm <- aov(hb ~ sex, data=psub) 
# summary(anovaHWIm)
# text(1.5,24, "*", cex=3, col=("red"))

# I need to clean up the sex variable mega 
# There's also something weird about the way sexes are appearing: olive drab looks like female? (only a few unknown points)

# Hb v. elev
library(ggplot2)
p1 <- ggplot(final, aes(x=elev, y=hb))
p1 <- p1 + geom_jitter(position = position_jitter(width = 0.1), alpha = 1/4)
p1 <- p1 + stat_smooth(method = lm)
p1 <- p1 + theme_bw()
p1 <- p1 + labs(title = "[Hb] v. elevation")
# p1 <- p1 + facet_wrap(. ~ country)
print(p1)

# Hct v. elev
library(ggplot2)
p1 <- ggplot(final, aes(x=elev, y=hct))
p1 <- p1 + geom_jitter(position = position_jitter(width = 0.1), alpha = 1/4)
p1 <- p1 + stat_smooth(method = lm)
p1 <- p1 + theme_bw()
p1 <- p1 + labs(title = "Hct v. elevation")
# p1 <- p1 + facet_wrap(. ~ country)
print(p1)

# MCV v. elev
library(ggplot2)
p1 <- ggplot(final, aes(x=elev, y=mcv))
p1 <- p1 + geom_jitter(position = position_jitter(width = 0.1), alpha = 1/4)
p1 <- p1 + stat_smooth(method = lm)
p1 <- p1 + theme_bw()
p1 <- p1 + labs(title = "MCV v. elevation")
print(p1)

# MCH v. elev
library(ggplot2)
p1 <- ggplot(final, aes(x=elev, y=mch))
p1 <- p1 + geom_jitter(position = position_jitter(width = 0.1), alpha = 1/4)
p1 <- p1 + stat_smooth(method = lm)
p1 <- p1 + theme_bw()
p1 <- p1 + labs(title = "MCH v. elevation")
print(p1)


# MCHC v. elev
library(ggplot2)
p1 <- ggplot(final, aes(x=elev, y=mchc))
p1 <- p1 + geom_jitter(position = position_jitter(width = 0.1), alpha = 1/4)
p1 <- p1 + stat_smooth(method = lm)
p1 <- p1 + theme_bw()
p1 <- p1 + labs(title = "MCHC v. elevation")
print(p1)


# Hb v. elev_position
library(ggplot2)
p <- ggplot(final, aes(x=elev_position, y=hb))
p <- p + geom_jitter(position = position_jitter(width = 0.1), alpha = 1/4)
p <- p + stat_smooth(method = lm)
p <- p + theme_bw()
p <- p + labs(title = "[Hb] v. elevation_position")
# p1 <- p1 + facet_wrap(. ~ country)
print(p)

# Hct v. elev_position
library(ggplot2)
p <- ggplot(final, aes(x=elev_position, y=hb))
p <- p + geom_jitter(position = position_jitter(width = 0.1), alpha = 1/4)
p <- p + stat_smooth(method = lm)
p <- p + theme_bw()
p <- p + labs(title = "Hct v. elevation_position")
# p1 <- p1 + facet_wrap(. ~ country)
print(p)

```



# Plot for Chris Brown Bag: 
If you want to plot this, read in final.hct from modeling file
```{r}
# Scatterplot: Hct v. elev
(bbag_plot <- ggplot(final.hct, aes(x=elev_position, y=hct, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final.hct, size=1.5, alpha=0.8) +
  scale_colour_viridis(discrete=TRUE) + 
  #scale_color_manual(values = peru_chimborazo) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="firebrick3", size=1.4) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation Position", y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  # ggtitle("(b)") + # Assign panel number/header; this will be (a) because first in series of 3
  # theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
  #        plot.title = element_text(face="bold")) + # This makes panel header bold 
  #      # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(axis.text.y=element_text(size=12), axis.text.x=element_text(size=12), axis.title=element_text(size=16))
 )
ggsave(bbag_plot, filename="bbag_plot_ScatterplotBasic_ElevPosition_by_Hct.pdf", height=3, width=4, units="in")
```





# Print environment for reproducibility
```{r}
sessionInfo() # List of packages and versions in use 
```

###########

## END 
