---
title: "Plots and Figures for Comparative Hummingbird Blood paper"
author: "Jessie Williamson"
date: "7/21/2020; last revised 2023-04-20"
output:
  md_document:
    variant: markdown_github
---

#### 

PLOTS & FIGS SCRIPT 

####

This script contains plots and figures for Williamson et al. 2023, Hummingbird blood traits track oxygen availability across space and time, *Ecology Letters*. These are separated into a standalone script to not bog down main scripts because some multi-panel figs are beefy. 

4/20/23 note: While going through this script and proofreading, I realized that some figure numbers never got updated to correspond with the final Ecology Letters supplement order. SO: In this script they're ordered correctly, but keep in mind that the number prefixes listed in pdf output files here probably don't always correspond to the pdf files on your hard drive; just keyword search other parts of the fig title and you'll find the files.  


# Load packages
```{R}
library(reshape)
library(reshape2)
library(plyr)
library(dplyr)
library(car)
library(GGally)
library(Hmisc)
library(gridExtra)
library(stats)
library(gplots)
library(ggplot2)
library(stats4) # Forces knitr to work when it's being wonky
library(PMCMR) #Allows Kruskal-Wallis post-hocs
library(effects)
library(gridExtra)
library(lattice)
library(survival)
library(fmsb)
library(faraway)
library(ape)
library(plotly) # for 3D plots
library(viridis)
library(patchwork)
library(htmlwidgets)
library(ggsignif)

# Mapping 
library(raster)
library(sp)
library(rgdal)
library(RStoolbox)
library(prettymapr)
library(viridis)
library(rasterVis)

# Modeling packages 
library(nlme)
library(lme4)
library(AICcmodavg)
library(MuMIn)
library(glmulti)
library(lsmeans)
library(rsq) # get r-squared values from GLM
library(r2glmm) # for R^2 values from lmer() and glmer()
library(multcompView) # related to multiple comparisons?
library(jtools) # interaction plots 
library(interactions) # interaction plots 
library(broom)
library(stargazer) # model output tables
library(ggeffects) # for estimating model predictions from mixed effects models
library(brms)
library(MCMCglmm)
library(rstan)
library(bayesplot)

# Phylo packages 
library(phytools)
```


---

# Clear workspace and set WD
```{R}
rm(list=ls(all=TRUE)) # clear workspace 
setwd("/Users/Jessie/Dropbox (MSBbirds)/Rdirectory/ComparativeHummingbirdBlood")
```



# load in data 
```{R}
# # Load functions 
# source("ada_functions.R") # Erik's ADA functions for clean & collated lm diag plots

# Final hummingbird blood dataset
final <- read.csv("ComparativeHummingbirdBlood_FinalDataset_ForModeling_06-14-21.csv")
final <- final[ , !(colnames(final) %in% c("X"))] # drop weird X column 1 if reading in from read.csv

spmeans <- read.csv("AllHummingbird_BloodData_SpeciesMeans_70SpeciesWithNoNAData_2022-01-17.csv")
# I have many iterations of species mean datasets; This one was made in HumBlood_Modeling_SpeciesMeans.Rmd
# It includes data from 70 no NA species for all 6 blood traits (so keep in mind that species means may be slightly 
# different for trait-specific datasets)

# Patagona-free subset (for possible modeling of mass, where Patagona is a huge body size outlier):
final.nopgig <- read.csv("HumBlood.final.nopgig_06-14-21.csv")
final.nopgig <- final.nopgig[ , !(colnames(final.nopgig) %in% c("X"))] # drop weird X column 1 if reading in from read.csv

# Phylogeny pruned from McGuire and adjusted (see script HumBlood_Phylogeny.Rmd)
#tree <- read.tree("McGuirePruned_AllHummingbirdComparativeTree_FINAL.tre") # a list 
# tree$tip.label # Check 77 tips
# plotTree(tree, ftype="i") # Check this pre-final tree w/ adjusted branch lengths 
```


# Set levels to ensure that colors plot properly
Make sure levels correspond to clade label order in phylogeny from top to bottom
```{r}
# Set clade levels/make sure these are correct
final$Clade <- factor(final$Clade
                       , levels = c("Coquette",
                                    "Brilliant", 
                                    "Patagona", 
                                    "Bee", 
                                    "Mountain-gem", 
                                    "Emerald", 
                                    "Mango",  
                                    "Topaz",
                                    "Hermit"))
```



# COLOR PALETTES 
```{r}
# Colors from Figure 1 map phylo fig (last revised 2/23/21); see Illustrator file
# Turquoise - 4FBBB6 (coquette) **may need to change because very similar to points on the map
# Gray - A5A5AC (brilliant)
# Red - EE4024 (patagona)
# Gold - D9CA44 (bees)
# Pink - E786B7 (mountain gems)
# Green - 37B453 (emeralds)
# Orange - F9AC1B (mangos)
# Blue - 00A5CB (topazes)
# purple - 9E6EAE (hermits)

# HUMMINGBIRD COLORS 
hum_colors <- c( "#04A5A5", # Turquoise (Coquette)
                 "#9E9285", # brown (Brilliant)
                 "#EE4024", # red (Patagona)
                 "#F4CD0B", # Gold (Bee)
                 "#E786B7", # pink (Mountain-gem)
                 "#97B25B", # green (Emerald)  
                 "#F79E0F", # orange (Mango)
                 "#03446B", # navy blue (Topaz)
                 "#916EF9" # purple (Hermit)
                    )

# Old Hum colors 
# hum_colors <- c( "#4FBBB6", # Turquoise (Coquette)
#                  "#A5A5AC", # gray (Brilliant)
#                  "#EE4024", # red (Patagona)
#                  "#D9CA44", # Gold (Bee)
#                  "#E786B7", # pink (Mountain-gem)
#                  "#37B453", # green (Emerald)  
#                  "#F9AC1B", # orange (Mango)
#                  "#00A5CB", # blue (Topaz)
#                  "#9E6EAE" # purple (Hermit)
#                             )

# Oreotrochilus chimborazo color palette: 
# Deep purple: #6E1EDC
# Aqua gorget: #1CD2D8; another color: #0DAFA9
# pollen forehead: #D49A76
# gray patch brown: #665E54; another version #726A61; another version #8C847B
# deep green: #344D3F
# chuquiragua buds: #E2833F; #F4893C
# chuquiragua leaves: #6F8F17
# snowy breast: #D1D1D9
# Iridescent cheeks: #BA51FA
#scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87

# Ochimborazo_palette <- c( "#0E0E17", # Black 
#                           "#1CD2D8", # turquoise aqua gorget
#                           "#8C847B", # browny flanks 
#                           "#F4893C", # Chuquiragua buds 
#                           "#166975", # Darker part of gorget
#                           "#FDC31F", # Chuquiragua pollen  
#                           "#BA51FA", # Pinkish shine to gorget; Dark green wing patches: #344D3F
#                           "#6E1EDC", # Deep purple 
#                           "#6F8F17" # Chuquiragua leaves 
#                             )
# 
# peru_palette <- c( "#628454", # sage green
#                    "#FB4301", # deep orange-red
#                    "#6D7B9C", # skirt blue
#                    "#966D56", # Adobe brown
#                    "#E7860E", # golden orange-yellow
#                    "#264142", # Deep sea green
#                    "#BFD1F8", # periwinkle on shirt
#                    "#952842", # cabernet
#                    "#F871AA" # pink
#                             )
# 
# # Peru chimborazo: inspired by Peruvian textiles and Oreotrochilus chimborazo colors 
# peru_chimborazo <- c( "#FDC31F", # Chuquiragua pollen
#                       "#1CD2D8", # turquoise aqua gorget
#                       "#8C847B", # browny flanks 
#                       "#9AB243", # Chuquiragua gree
#                        "#FB4301", # Deep orange-red
#                       "#264142", # Deep sea green
#                       "#F871AA", # pink
#                       "#6E1EDC", # Deep purple 
#                       "#6D7B9C" # skirt blue # played around w/ subbing in this too:   "#A8A9AD" # gray breast
#                             )
```


# Basic plot of sampling across elevations
```{r}
# SUPER BASIC PLOTS, missing a lot of formatting/aesthetics 
# Sampling density across elevation 
(p <- ggplot(final, aes(x=elev, y=..density..)) +
      geom_density(alpha=.2, fill="#FF6666") + 
      geom_histogram(binwidth = 200) + 
    #  scale_color_manual(values = hum_colors) +
      theme_classic() + 
      labs(x="Elevation (m)", y="Sampling Density") + 
      theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )

# Sampling Depth across elevation  
(p <- ggplot(final, aes(x=elev)) +
   #   geom_density(alpha=.2, fill="#FF6666") + 
      geom_histogram(binwidth = 200) + 
    #  scale_color_manual(values = hum_colors) +
      theme_classic() + 
      labs(x="Elevation (m)", y="Sampling Depth") + 
      theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )

# Overlaid histograms
# ggplot(final, aes(x=elev)) +
#   geom_histogram(fill="black", alpha=0.5, position="identity")
# 
# ggplot(final, aes(elev, colour = Clade)) +
#   geom_freqpoly(binwidth = 500)
# 
# ggplot(final, aes(elev)) +
#   geom_histogram(bins = 100) + 
#   theme_classic() 
```


# FIGURE 1: Was made by producing raw maps and species elev ranges in HumBlood_PhyloElevRangeMapFig.Rmd; then by exporting raw graphics to Illustrator to make the final figure. 


# FIGURE 2: COMBINED INDIVIDUAL-LEVEL AND SPECIES MEAN DATA IN ONE MASTER PLOT 
Figure 2, a 6-panel fig w/ relationship between predictors and elevation. Linear models fit to main plots (break point analysis is in the supplement)
```{r}
# IMPORTANT! Read in 'final' dataset, set levels and colors for hum_colors above

# Read in species mean datasets (made in HumBlood_Modeling_SpeciesMeans.Rmd)
read.csv("AllHummingbirdBlood_Wrangled_withBioClim_andPCA_03-01-21.csv")
spm.hb <- read.csv("HumBlood_SpeciesMeans_spm.hb_06-14-21.csv")
spm.hct <- read.csv("HumBlood_SpeciesMeans_spm.hct_06-14-21.csv")
spm.trbc <- read.csv("HumBlood_SpeciesMeans_spm.trbc_06-14-21.csv")
spm.mcv <- read.csv("HumBlood_SpeciesMeans_spm.mcv_06-14-21.csv")
spm.mch <- read.csv("HumBlood_SpeciesMeans_spm.mch_06-14-21.csv")
spm.mchc <- read.csv("HumBlood_SpeciesMeans_spm.mchc_06-14-21.csv")

# Set levels. I'm sure there's a FAR less tedious and ultra-repetitive way to do this, but I'm in the mood to make plots, 
# not struggle w/ syntax challenges. 
# NO MOUNTAIN-GEM IN SPECIES MEAN DATA (because of n=1)
spm.hb$Clade <- factor(spm.hb$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.hct$Clade <- factor(spm.hct$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.trbc$Clade <- factor(spm.trbc$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.mcv$Clade <- factor(spm.mcv$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.mch$Clade <- factor(spm.mch$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.mchc$Clade <- factor(spm.mchc$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))

# species_mean_colors
spec_mean_colors <- c( "#04A5A5", # Turquoise (Coquette)
                 "#9E9285", # brown (Brilliant)
                 "#EE4024", # red (Patagona)
                 "#F4CD0B", # Gold (Bee)
                # "#E786B7", # pink (Mountain-gem)
                 "#97B25B", # green (Emerald)  
                 "#F79E0F", # orange (Mango)
                 "#03446B", # navy blue (Topaz)
                 "#916EF9" # purple (Hermit)
                    )
```


# Plot master figure 2
```{r}
# Scatterplots of each of the 6 blood characteristics with elevation, colored by clade
# Code to set shapes by clade, but this was totally overwhelming
  # scale_shape_manual(values=c("Bee"=4, "Brilliant"=17, "Coquette"=15, # Shapes are a little overwhelming, honestly
  #                             "Emerald"=18, "Hermit"=25, "Mango"=8, "Mountain-gem"=3,
  #                             "Patagona"=16, "Topaz"=9)) +
 #   scale_shape_manual(values=shapes) + # Another way to do this is set colors above 

# PANEL A: Scatterplot: Hb v. elev 
(p1 <- ggplot(final, aes(x=elev, y=hb, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
 # facet_wrap(~Clade) + 
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  #geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above4ksub) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="[Hb] (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  annotate("text", x=4200, y=24.2, size=3, label = "italic(R) ^ 2==0.21", parse = TRUE) + # R^2 from full model
  ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p1, filename="p1_ScatterplotBasic_Elev_by_Hb_2023-02-27.pdf", height=7, width=9, units="in")
# The 166 rows of missing values come from many NA Hb values; don't worry about these 

# Get R^2 for Panel A; R^2 reported from Bayesian model
# fig2.panelA.r2 <- glm(hb ~ elev, data=final, family=gaussian())
# summary(fig2.panelA.r2)
# round((1 - (fig2.panelA.r2$deviance/fig2.panelA.r2$null.deviance)),2) # R^2 = 0.08

##### 

# PANEL B: Scatterplot: Hct v. elev
(p2 <- ggplot(final, aes(x=elev, y=hct, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  # scale_colour_viridis(discrete=TRUE) + 
  #facet_wrap(~Clade) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4200, y=78, size=3, label = "italic(R) ^ 2==0.17", parse = TRUE) + # R^2 from full model
  ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p2, filename="p3_ScatterplotBasic_Elev_by_Hct_2023-02-27.pdf", height=7, width=9, units="in")

# # Get R^2 for Panel B; R^2 reported from Bayesian model
# fig2.panelB.r2 <- glm(hct ~ elev, data=final, family=gaussian())
# summary(fig2.panelA.r2)
# round((1 - (fig2.panelB.r2$deviance/fig2.panelB.r2$null.deviance)),2) # R^2 = 0.06

#####

# PANEL C: Scatterplot: TRBC v. elev
(p3 <- ggplot(final, aes(x=elev, y=trbc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  #facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise
  #geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above4ksub) + # line used to be dark turquoise
  theme_classic() + 
   labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     x="Elevation (m)") + # Removed "Elevation (m)" because we don't want Panel A to have label
    # in expression world, ~ is space and mathematical symbols like parenthesis need to be set of w/ " " 
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4200, y=9.8, size=3, label = "italic(R) ^ 2==0.17", parse = TRUE) + # R^2 from full model
  ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,right=5,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p3, filename="p3_ScatterplotBasic_Elev_by_TRBC_2023-02-27.pdf", height=7, width=9, units="in")

# Get R^2 for Panel C; R^2 reported from Bayesian model
# fig2.panelC.r2 <- glm(trbc ~ elev, data=final, family=gaussian())
# round((1 - (fig2.panelC.r2$deviance/fig2.panelC.r2$null.deviance)),2) # R^2 = 0.01

#####

# PANEL D: Scatterplot: mcv v. elev
(p4 <- ggplot(final, aes(x=elev, y=mcv, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  # facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
  # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, linetype=2) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4200, y=150, size=3, label = "italic(R) ^ 2==0.12", parse = TRUE) + # R^2 from full model
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  #theme(legend.box.spacing = unit(c(0.2,5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p4, filename="p4_ScatterplotBasic_Elev_by_MCV_2022-01-17.pdf", height=7, width=9, units="in")


# PANEL E: Scatterplot: MCH v. elev
(p5 <- ggplot(final, aes(x=elev, y=mch, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
 # facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", y="MCH (pg)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4200, y=52, size=3, label = "italic(R) ^ 2==0.09", parse = TRUE) + # R^2 from full model
  ggtitle("E") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p5, filename="p5_ScatterplotBasic_Elev_by_MCH_2021-11-29.pdf", height=7, width=9, units="in")

# Get R^2 for Panel E; R^2 reported from Bayesian model
# fig2.panelE.r2 <- glm(mch ~ elev, data=final, family=gaussian())
# round((1 - (fig2.panelE.r2$deviance/fig2.panelE.r2$null.deviance)),2) # R^2 = 0.01

#####

# PANEL F: Scatterplot: MCHC v. elev
(p6 <- ggplot(final, aes(x=elev, y=mchc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
 #  facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  #geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", y="MCHC (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  ggtitle("F") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,right=5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p6, filename="p6_ScatterplotBasic_Elev_by_MCHC_2021-11-29.pdf", height=7, width=9, units="in")


### SPECIES MEAN PLOTS ####

# Scatterplots of each of the 6 blood characteristics with elevation, colored by clade but w/ species mean data 

# PANEL G: Scatterplot: Hb v. elev 
(p21 <- ggplot(spm.hb, aes(x=elev.mean, y=hb.mean, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=0.8, alpha=0.8) +
  # facet_wrap(~Clade) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = spec_mean_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  #geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=0.6) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="[Hb] (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  annotate("text", x=1300, y=20.4, size=1.9, label = "italic(R) ^ 2==0.18", parse = TRUE) + # R^2 from full model
  ggtitle("G") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  # theme(plot.margin = unit(c(0.0,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.5,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=6), axis.text.x=element_text(size=6), axis.title.y=element_text(size=6, margin = 
  margin(r=-8)), axis.title.x=element_text(size=6))
# r=0 specifies that you want the right side of the y-axis label flush with tick marks
 )
ggsave(p21, filename="p21_ScatterplotBasic_Elev_by_Hb_SPM_2021-11-29.pdf", height=7, width=9, units="in")
# The 166 rows of missing values come from many NA Hb values; don't worry about these 

# Get R^2 for Panel G; R^2 reported from Bayesian model
# fig2.panelG.r2 <- glm(hb.mean ~ elev.mean, data=spm.hb, family=gaussian())
# round((1 - (fig2.panelG.r2$deviance/fig2.panelG.r2$null.deviance)),2) # R^2 = 0.18

#####  


# PANEL H: Scatterplot: Hct v. elev
(p22 <- ggplot(spm.hct, aes(x=elev.mean, y=hct.mean, colour=Clade)) + # Add shape=Clade here to shapes to work properly
  geom_point(size=0.8, alpha=0.8) +
 #  facet_wrap(~Clade) +
  # scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = spec_mean_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  #geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=0.6) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=1300, y=65, size=1.9, label = "italic(R) ^ 2==0.24", parse = TRUE) + # R^2 from full model
  ggtitle("H") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  # theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.5,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=6), axis.text.x=element_text(size=6), axis.title.y=element_text(size=6, margin = 
  margin(r=-8)), axis.title.x=element_text(size=6))
  # r=0 specifies that you want the right side of the y-axis label flush with tick marks
 )
ggsave(p22, filename="p22_ScatterplotBasic_Elev_by_Hct_SPM_2021-11-29.pdf", height=7, width=9, units="in")

# Get R^2 for Panel H; R^2 reported from Bayesian model
# fig2.panelH.r2 <- glm(hct.mean ~ elev.mean, data=spm.hct, family=gaussian())
# round((1 - (fig2.panelH.r2$deviance/fig2.panelH.r2$null.deviance)),2) # R^2 = 0.09

##### 


# PANEL I: Scatterplot: TRBC v. elev
(p23 <- ggplot(spm.trbc, aes(x=elev.mean, y=trbc.mean, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=0.8, alpha=0.8) +
  # facet_wrap(~Clade) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = spec_mean_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  # geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + 
  #geom_smooth(method = "lm", formula = y ~ x, colour="black", size=0.6) + 
  geom_smooth(method = "lm", formula =  y ~ x + I(x^2), colour="black", size=0.6) + # we opted for solid bc Fig. 4 
    # USING LINEAR FIT BECUASE QUADRATIC OVERLAPPED ZERO IN FULL SPECIES MEAN MODEL
  theme_classic() + 
   labs(y=expression(atop("TRBC","("~RBC~x~10^{6}~"/"~mm^{3}~")")), # atop() is KEY for splitting expression onto 2 lines
  # labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # one line y-axis formatting  
     x="") + # Removed "Elevation (m)" because we don't want Panel A to have label
    # in expression world, ~ is space and mathematical symbols like parenthesis need to be set of w/ " " 
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=1300, y=8.2, size=1.9, label = "italic(R) ^ 2==0.22", parse = TRUE) + # R^2 from full model
  ggtitle("I") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  # theme(plot.margin = unit(c(0.0,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.5,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=6), axis.text.x=element_text(size=6), axis.title.y=element_text(size=6, margin = 
  margin(r=-6)), axis.title.x=element_text(size=6))
  # r=0 specifies that you want the right side of the y-axis label flush with tick marks
 )
ggsave(p23, filename="p23_ScatterplotBasic_Elev_by_TRBC_SPM_2022-01-18.pdf", height=7, width=9, units="in")

# # Get R^2 for Panel I; R^2 reported from Bayesian model
# fig2.panelI.r2 <- glm(trbc.mean ~ elev.mean, data=spm.trbc, family=gaussian())
# round((1 - (fig2.panelI.r2$deviance/fig2.panelI.r2$null.deviance)),2) # R^2 = 0.01

##### 


# PANEL J: Scatterplot: mcv v. elev
(p24 <- ggplot(spm.mcv, aes(x=elev.mean, y=mcv.mean, colour=Clade)) + # shape=Clade here to get shapes to work properly
  geom_point(size=0.8, alpha=0.8) +
 #  facet_wrap(~Clade) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = spec_mean_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=0.6, linetype=2) + 
     # REMOVE TREND LINE BECAUSE NOT A SIGNIFICANT PREDICTOR IN MODELS
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=3900, y=122, size=1.9, label = "italic(R) ^ 2==0.20", parse = TRUE) + # Add R^2 value from model
  ggtitle("J") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  #theme(legend.box.spacing = unit(c(0.2,5,0.2,0.2), "cm")) +  # top, right, bottom, left
  # theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.5,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=6), axis.text.x=element_text(size=6), axis.title.y=element_text(size=6, margin = 
  margin(r=-8)), axis.title.x=element_text(size=6))
  # r=0 specifies that you want the right side of the y-axis label flush with tick marks
 )
ggsave(p24, filename="p24_ScatterplotBasic_Elev_by_MCV_SPM_2022-01-17.pdf", height=7, width=9, units="in")


# PANEL K: Scatterplot: MCH v. elev
(p25 <- ggplot(spm.mch, aes(x=elev.mean, y=mch.mean,  colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=0.8, alpha=0.8) +
 #   facet_wrap(~Clade) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = spec_mean_colors) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=0.6) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="", y="MCH (pg)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=3900, y=39.6, size=1.9, label = "italic(R) ^ 2==0.22", parse = TRUE) + # Add R^2 value from model
  ggtitle("K") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  # theme(plot.margin = unit(c(0.0,0.1,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.5,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=6), axis.text.x=element_text(size=6), axis.title.y=element_text(size=6, margin = 
  margin(r=-8)), axis.title.x=element_text(size=6))
  # r=0 specifies that you want the right side of the y-axis label flush with tick marks
 )
ggsave(p25, filename="p25_ScatterplotBasic_Elev_by_MCH_SPM_2021-11-29.pdf", height=7, width=9, units="in")

# Get R^2 for Panel K; R^2 reported from Bayesian model
# fig2.panelK.r2 <- glm(mch.mean ~ elev.mean + I(elev.mean^2), data=spm.mch, family=gaussian())
# round((1 - (fig2.panelK.r2$deviance/fig2.panelK.r2$null.deviance)),2) # R^2 = 0.10

##### 


# PANEL L: Scatterplot: MCHC v. elev
(p26 <- ggplot(spm.mchc, aes(x=elev.mean, y=mchc.mean, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=0.8, alpha=0.8) +
 #  facet_wrap(~Clade) +
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = spec_mean_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  # geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=1.0) + # line used to be dark turquoise 
    # REMOVE TREND LINE BECAUSE NOT A SIGNIFICANT PREDICTOR IN MODELS
  #geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation (m)", y="MCHC (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  ggtitle("L") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  # theme(plot.margin = unit(c(0.0,0.2,0.2,0.1), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.5,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=6), axis.text.x=element_text(size=6), axis.title.y=element_text(size=6, margin = 
  margin(r=-8)), axis.title.x=element_text(size=10))
  # r=0 specifies that you want the right side of the y-axis label flush with tick marks
 )
ggsave(p26, filename="p26_ScatterplotBasic_Elev_by_MCHC_SPM_2021-11-29.pdf", height=7, width=9, units="in")


# COMBINE ALL PLOTS TO MAKE 12-PANEL FIGURE: 
# library(patchwork)
Fig2_v2_all_data <- (
          p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(widths = c(0.8,0.8,0.8))
          + inset_element(p21 / p22 / p23 / p24 / p25 / p26, 
                          left = 1.0, # Increased #s move left edge to the right; decreased to the left
                          bottom = -0.19, 
                          right = 1.72, # Increased #s move right edge right; decreased to the left
                          top = 2.38) 
          )  
print(Fig2_v2_all_data)
ggsave(Fig2_v2_all_data, filename = "Fig2_v2_all_data_Dashed-and-Solid_2023-02-27.pdf", bg="transparent", height=6.7, width=11, units="in")
# A little janky, but I set up the species mean panel as an inset in order to get the formatting to work
# SYNTAX VERY TRICKY! 
# Only big problem with this is that I can't figure out how to get guides="collect" to put common A-F legend on RIGHT
# side of my giant inset - legend wants to appear over the top of the inset. 



# OLD SYNTAX
# Fig2_v2_all_data <- (
#                     p1 + p2 + p3 + p4 + p5 + p6 |
#                     p21 / p22 / p23 / p24 / p25 / p26
#                     #+ plot_layout(guides = "collect")
#                      #, heights = unit(c(5,5, 1), c('cm', 'cm', 'null')))
#                     #+ plot_layout(widths = c(2,2,2,1))
#                     )
# plot(Fig2_v2_all_data)
# Don't want ncol argument becuase ncol=3 without parenthesis in the right place treats the | pipe as separate
# test <- p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(widths = c(2, 2, 1))
# plot(test)
# This works fine and legends are ok (would look fine if you commented out species mean legends maybe)
```



### FIGURE 3 - 6-panel blood posterior probability fig (intraspecific model main panesls w/ interspecific model insets)
A 6-panel posterior probability plot figure with individual-level data showing full models of species-mean data in insets within each plot. 

#Set up some figure aesthetics and load in data: 
```{r}
library(tidyverse)
library(patchwork)

# Load in individual-level models for plots 
load("Rdata_files/hb.m4_FullModel_SpeciesREOnly.RData")
load("Rdata_files/hct.m4_FullModel_SpeciesREOnly.RData")
load("Rdata_files/trbc.m4_FullModel_SpeciesREOnly.RData") 
load("Rdata_files/mcv.m4_FullModel_SpeciesREOnly.RData")
load("Rdata_files/mch.m4_FullModel_SpeciesREOnly.RData")
load("Rdata_files/mchc.m4_FullModel_SpeciesREOnly.RData") 

# Quick load to be able to make plots for big species mean comparison plot: 
# For quick pasting for final plot 
load("Rdata_files/spm.hb.m2_FullModel.RData") 
load("Rdata_files/spm.hct.m2_FullModel.RData") 
load("Rdata_files/spm.trbc.m4_FullModelWithQuadratic.RData")
load("Rdata_files/spm.mcv.m4_FullModelWithQuadratic.RData") 
load("Rdata_files/spm.mch.m4_FullModelWithQuadratic.RData")  
load("Rdata_files/spm.mchc.m4_FullModelWithQuadratic.RData")

# Set color scheme 
color_scheme_set("viridisD")
# color_scheme_set("viridisE") # Looks most sophisticated for fig; ViridisC looks a little clownish
# Options: blue, teal, red, brightblue, purple, gray, mix-blue-red, pink; ViridisC & E are my favorite
# See help function of color_scheme_set for full list and how to mix colors 

# Note: if you use font_family "Arial" ggsave will have trouble writing this to a pdf because it makes the plot w/ Arial 
# Narrow and pdf is slightly different; either use family "Helvetica" (or any other), or follow instructions here: 
# https://github.com/thomasp85/ggraph/issues/152
```

PRINT FIGURE 3. 
```{R}
# HB - panel A
color_scheme_set("viridisD")
fig3a_hb_main <- mcmc_plot(hb.m4, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer","b_elev.pos.z",  "b_intravar.temp.z", "b_intravar.precip.z", "b_intravar.mass.z"), 
            prob_outer=0.95, # 95% outer CI 
            prob=0.89, # 50% inner CI 
            point_est="mean") + # mean point est; default is median 
            #mcmc_plot(hb.m4, variable=c("b_elev.z"), prob_outer=0.95, prob=0.50, point_est="mean", col="red") + 
            labs(x="Effect on [Hb] (within species)", title = " ") + 
            scale_y_discrete(labels=c("Elev", "Temp", "Precip", "Mass", "Genotype (Ser-Ser)", "Genotype (Gly-Ser)", 
            "Elev Position", "Intra Var: Temp", "Intra Var: Precip", "Intra Var: Mass")) +
            theme(plot.title = element_text(family = "Helvetica", color="black", size=12)) + # font = panel header size
            theme_bw() +  #theme with white background
            geom_point(aes(x=0.52, y=1), color="red", size=3.35) + # Add red circle 4 important predictor (elev)
            theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
            theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
            theme(text = element_text(family="Helvetica",size=10)) + # y axis labels
            ggtitle("A") + # Assign panel number/header
            theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
            plot.title = element_text(face="bold"))  # This makes panel header bold 
            # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
            #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left 
fig3a_hb_main
ggsave(fig3a_hb_main, filename = "fig3a_hb_main_IndivLevel.pdf", bg="transparent", height=7, width=9, units="in")

# HB Inset plot (Final model, M4)
color_scheme_set("brightblue")
fig3a_hb_inset <- mcmc_plot(spm.hb.m2, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer"),
            prob_outer=0.95, # 95% outer CI
            prob=0.89, # 50% inner CI
            point_est="mean", 
              point_size=2, inner_size=1) +
            labs(x="Effect on [Hb]\n(among species)", title = " ") +
            scale_y_discrete(labels=c("Elev","Temp", "Precip", "Mass", "Gen. (Ser-Ser)","Gen. (Gly-Ser)")) +
            theme(plot.title = element_text(family = "Helvetica", color="black", size=9)) + # font = panel header size
            theme_bw() +  #theme with white background
            #annotate('text', x = -1.38, y = 0.78, label='bold("*")', size=4, variablee=TRUE) + # ASTERISK 4 top model sig
            geom_point(aes(x=0.775, y=1), color="red", size=1.35) + # Add red circle 4 important predictor (elev)
             theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
            theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
            theme(axis.text=element_text(size=5), axis.title=element_text(size=5)) + # y-axis text, x-axis text
            #ggtitle("A") + # Assign panel number/header
            theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
            plot.title = element_text(face="bold"))  # This makes panel header bold
            # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
            #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
fig3a_hb_inset
ggsave(fig3a_hb_inset, filename = "fig3a_hb_inset_SpeciesMean.pdf", bg="transparent", height=7, width=9, units="in")

# HB PANEL A WITH INSET
fig3a_hb_panel <- fig3a_hb_main + inset_element(fig3a_hb_inset,
                                                       right = 1.03, # THIS IS LEFT 
                                                       bottom = 0.58, 
                                                       left = 0.60, # THIS IS RIGHT 
                                                       top = 1.2)
fig3a_hb_panel

###

# HCT - Panel B
color_scheme_set("viridisD")
fig3b_hct_main <- mcmc_plot(hct.m4, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer","b_elev.pos.z",  "b_intravar.temp.z", "b_intravar.precip.z", "b_intravar.mass.z"),
            prob_outer=0.95, # 95% outer CI 
            prob=0.89, # 50% inner CI 
            point_est="mean") + # mean point est; default is median 
            labs(x="Effect on Hct (within species)", title = " ") + 
            # scale_y_discrete(labels=c("Elevation","Elevation Position", "Mass", "Temp", "Precip", "Genotype (Gly-Ser)", 
            # "Genotype (Ser-Ser)", "Intra Var: Mass", "Intra Var: Temp", "Intra Var: Precip")) +  
            scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel fig
            theme(plot.title = element_text(family = "Helvetica", color="black", size=12)) + # font = panel header size
            theme_bw() +  
            geom_point(aes(x=-.437, y=10), color="red", size=3.35) + # red circle: intravar mass 
            geom_point(aes(x=.671, y=7), color="red", size=3.35) + # red circle: elev pos
            geom_point(aes(x=1.873, y=1), color="red", size=3.35) + # red circle: elev
            theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
            theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
            theme(text = element_text(family="Helvetica", size=10)) + # y axis labels
            ggtitle("B") + # Assign panel number/header
            theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
            plot.title = element_text(face="bold"))  # This makes panel header bold 
            # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
            #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left 
fig3b_hct_main
ggsave(fig3b_hct_main, filename = "fig3b_hct_main_IndivLevel.pdf", bg="transparent", height=7, width=9, units="in")


#HCT Inset plot 
color_scheme_set("brightblue")
fig3b_hct_inset <- mcmc_plot(spm.hct.m2, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer"),
            prob_outer=0.95, # 95% outer CI
            prob=0.89, # 50% inner CI
            point_est="mean", 
              point_size=2, inner_size=1) +
            labs(x="Effect on Hct\n(among species)", title = " ") +
            scale_y_discrete(labels=c("Elev","Temp", "Precip", "Mass", "Gen. (Ser-Ser)","Gen. (Gly-Ser)")) +
            # scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel fig
            theme(plot.title = element_text(family = "Helvetica", color="black", size=9)) + # font = panel header size
            theme_bw() +  #theme with white background
            #annotate('text', x = -5.58, y = 0.78, label='bold("*")', size=4, parse=TRUE) + # ASTERISK 
            geom_point(aes(x=1.829, y=1), color="red", size=1.35) + # Add red (elev)
            theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
            theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
            theme(axis.text=element_text(size=5), axis.title=element_text(size=5)) + # y-axis text, x-axis text
            #ggtitle("B") + # Assign panel number/header
            theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
            plot.title = element_text(face="bold"))  # This makes panel header bold
            # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
            #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
fig3b_hct_inset
ggsave(fig3b_hct_inset, filename = "fig3b_hct_inset_SpeciesMean.pdf", bg="transparent", height=7, width=9, units="in")


# HB PANEL A WITH INSET
fig3b_hct_panel <- fig3b_hct_main + inset_element(fig3b_hct_inset,
                                                       right = 1.03, # THIS IS LEFT
                                                       bottom = 0.58,
                                                       left = 0.60, # THIS IS RIGHT
                                                       top = 1.2)
# RIGHT IS LEFT AND LEFT IS RIGHT; THIS IS VERY CONFUSING
fig3b_hct_panel


###


# TRBC - Panel C
color_scheme_set("viridisD")
fig3c_trbc_main <- mcmc_plot(trbc.m4, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer","b_elev.pos.z",  "b_intravar.temp.z", "b_intravar.precip.z", "b_intravar.mass.z"),
              prob_outer=0.95, # 95% outer CI 
              prob=0.89, # 50% inner CI 
              point_est="mean") + # mean point est; default is median 
              labs(x="Effect on TRBC (within species)", title = " ") + 
              #scale_y_discrete(labels=c("Elev", "Elev^2", "Temp", "Precip", "Mass", "Genotype\n(Ser-Ser)", 
              # "Genotype\n(Gly-Ser)")) +
              scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel
              theme(plot.title = element_text(family = "Helvetica", color="black", size=12)) + # font = panel header size
              theme_bw() +
              geom_point(aes(x=0.408, y=1), color="red", size=3.35) + # Add red circle (elev)
              theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                    panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
              theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
              theme(text = element_text(family="Helvetica", size=10)) + # y axis labels
              ggtitle("C") + # Assign panel number/header
              theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
              plot.title = element_text(face="bold"))  # This makes panel header bold 
              # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
              #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left 
fig3c_trbc_main
ggsave(fig3c_trbc_main, filename = "fig3c_trbc_main_IndivLevel.pdf", bg="transparent", height=7, width=9, units="in")


# TRBC Inset plot (Null model, M1)
color_scheme_set("brightblue")
fig3c_trbc_inset <- mcmc_plot(spm.trbc.m4, variable=c("b_elev.z", "b_Ielev.zE2","b_temp.z", "b_precip.z", "b_mass.z", 
              "b_genotypeSerMSer", "b_genotypeGlyMSer"),
              prob_outer=0.95, # 95% outer CI
              prob=0.89, # 50% inner CI
              point_est="mean", 
              point_size=2, inner_size=1) +
              labs(x="Effect on TRBC\n(among species)", title = " ") +
              scale_y_discrete(labels=c("Elev"," Elev^2", "Mass", "Temp", "Precip","Gen. (Gly-Ser)",
              "Gen. (Ser-Ser)")) +
              # scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel 
              theme(plot.title = element_text(family = "Helvetica", color="black", size=9)) + # font = panel header size
              theme_bw() +  
              #geom_point(aes(x=0.502, y=1), color="red", size=1.35) + # Add red (elev) # no sig; null is top
              theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                    panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
              theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
              theme(axis.text=element_text(size=5), axis.title=element_text(size=5)) + # y-axis text, x-axis text
              # ggtitle("C") + # Assign panel number/header
              theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
              plot.title = element_text(face="bold"))  # This makes panel header bold
              # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
              #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
fig3c_trbc_inset
ggsave(fig3c_trbc_inset, filename = "fig3c_trbc_inset_SpeciesMean.pdf", bg="transparent", height=7, width=9, units="in")


# TRBC PANEL A WITH INSET
fig3c_trbc_panel <- fig3c_trbc_main + inset_element(fig3c_trbc_inset,
                                                       right = 1.03, # THIS IS LEFT 
                                                       bottom = 0.58, 
                                                       left = 0.60, # THIS IS RIGHT 
                                                       top = 1.2)
fig3c_trbc_panel

###

# MCV - panel D
color_scheme_set("viridisD")
fig3d_mcv_main <- mcmc_plot(mcv.m4, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer","b_elev.pos.z", "b_intravar.temp.z", "b_intravar.precip.z", "b_intravar.mass.z"),
             prob_outer=0.95, # 95% outer CI 
             prob=0.89, # 50% inner CI 
             point_est="mean") + # mean point est; default is median 
             labs(x="Effect on MCV (within species)", title = " ") + 
             scale_y_discrete(labels=c("Elev", "Temp", "Precip", "Mass", "Genotype (Ser-Ser)", "Genotype (Gly-Ser)", 
             "Elev Position", "Intra Var: Temp", "Intra Var: Precip", "Intra Var: Mass")) +
             theme(plot.title = element_text(family = "Helvetica", color="black", size=12)) + # font = panel header size
             theme_bw() +  #theme with white background
             theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
             theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
             theme(text = element_text(family="Helvetica", size=10)) + # y axis labels 
             ggtitle("D") + # Assign panel number/header
             theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
             plot.title = element_text(face="bold"))  # This makes panel header bold 
             # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
             #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left 
fig3d_mcv_main # + inset_element(mcv.m6.fig3d.inset, 0.6, 0.6, 1, 1)
ggsave(fig3d_mcv_main, filename = "fig3d_mcv_main_IndivLevel.pdf", bg="transparent", height=7, width=9, units="in")

# MCV Inset plot (Final model, M6)
color_scheme_set("brightblue")
fig3d_mcv_inset <- mcmc_plot(spm.mcv.m4, variable=c("b_elev.z", "b_Ielev.zE2","b_temp.z", "b_precip.z", "b_mass.z", 
            "b_genotypeSerMSer", "b_genotypeGlyMSer"),
             prob_outer=0.95, # 95% outer CI
             prob=0.89, # 50% inner CI
             point_est="mean", 
              point_size=2, inner_size=1) +
             labs(x="Effect on MCV\n(among species)", title = " ") +
             scale_y_discrete(labels=c("Elev", "Elev^2", "Temp", "Precip", "Mass", "Gen. (Ser-Ser)", 
              "Gen. (Gly-Ser)")) +
             theme(plot.title = element_text(family = "Helvetica", color="black", size=9)) + # font = panel header size
             theme_bw() +  #theme with white background
             theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
             theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
             theme(axis.text=element_text(size=5), axis.title=element_text(size=5)) + # y-axis text, x-axis text
             # ggtitle("D") + # Assign panel number/header
             theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
             plot.title = element_text(face="bold"))  # This makes panel header bold
             # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
             #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
fig3d_mcv_inset # + inset_element(mcv.m6.fig3d.inset, 0.6, 0.6, 1, 1)
ggsave(fig3d_mcv_inset, filename = "fig3d_mcv_inset_SpeciesMean.pdf", bg="transparent", height=7, width=9, units="in")

# MCV PANEL D WITH INSET 
fig3d_mcv_panel <- fig3d_mcv_main + inset_element(fig3d_mcv_inset, 
                                                       right = 1.03, 
                                                       bottom = 0.58, 
                                                       left = 0.60, 
                                                       top = 1.2)
# RIGHT IS LEFT AND LEFT IS RIGHT; THIS IS VERY CONFUSING
fig3d_mcv_panel

###

# MCH - panel E
color_scheme_set("viridisD")
fig3e_mch_main <- mcmc_plot(mch.m4, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer","b_elev.pos.z",  "b_intravar.temp.z", "b_intravar.precip.z", "b_intravar.mass.z"),
             prob_outer=0.95, # 95% outer CI 
             prob=0.89, # 50% inner CI 
             point_est="mean") + # mean point est; default is median 
             labs(x="Effect on MCH (within species)", title = " ") + 
              # scale_y_discrete(labels=c("Elevation","Elevation Position","Mass", "Temp", "Precip", "Genotype (Gly-Ser)",
              #"Genotype (Ser-Ser)", "Intra Var: Mass", "Intra Var: Temp", "Intra Var: Precip")) +
             scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel fig
             theme(plot.title = element_text(family = "Helvetica", color="black", size=12)) + # font = panel header size
             theme_bw() +  #theme with white background
             theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
             theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
             theme(text = element_text(family="Helvetica", size=10)) + # y axis labels 
             ggtitle("E") + # Assign panel number/header
             theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
             plot.title = element_text(face="bold"))  # This makes panel header bold 
             # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
             #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
fig3e_mch_main
ggsave(fig3e_mch_main, filename = "fig3e_mch_main_IndivLevel.pdf", bg="transparent", height=7, width=9, units="in")

# MCH INSET PLOT 
color_scheme_set("brightblue")
fig3e_mch_inset <- mcmc_plot(spm.mch.m4, variable=c("b_elev.z", "b_Ielev.zE2","b_temp.z", "b_precip.z", "b_mass.z", 
            "b_genotypeSerMSer", "b_genotypeGlyMSer"),
             prob_outer=0.95, # 95% outer CI
             prob=0.89, # 50% inner CI
             point_est="mean", 
             point_size=2, inner_size=1) + 
             labs(x="Effect on MCH\n(among species)", title = " ") +
              scale_y_discrete(labels=c("Elev", "Elev^2", "Temp", "Precip", "Mass", "Gen. (Ser-Ser)", 
              "Gen. (Gly-Ser)")) +
             # scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel fig
             theme(plot.title = element_text(family = "Helvetica", color="black", size=9)) + # font = panel header size
             theme_bw() + 
             geom_point(aes(x=-1.485, y=2), color="red", size=1.35) + # Add red (elev quadratic)
             #annotate('text', x=-3.98, y=0.78, label='bold("*")', size=4, parse=TRUE) + # ASTERISK 4 top model sig
             theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
             theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
             theme(axis.text=element_text(size=5), axis.title=element_text(size=5)) + # y-axis text, x-axis text
             # ggtitle("E") + # Assign panel number/header
             theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
             plot.title = element_text(face="bold"))  # This makes panel header bold
             # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
             #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm"))  # top, right, bottom, left
fig3e_mch_inset
ggsave(fig3e_mch_inset, filename = "fig3e_mch_inset_SpeciesMean.pdf", bg="transparent", height=7, width=9, units="in")

# MCH PANEL E WITH INSET 
fig3e_mch_panel <- fig3e_mch_main + inset_element(fig3e_mch_inset, 
                                                  right = 1.03, # RIGHT IS LEFT 
                                                  bottom = 0.58, 
                                                  left = 0.60, # LEFT IS RIGHT
                                                  top = 1.2)
fig3e_mch_panel

###

# MCHC - panel F
color_scheme_set("viridisD")
fig3f_mchc_main <- mcmc_plot(mchc.m4, variable=c("b_elev.z", "b_temp.z", "b_precip.z", "b_mass.z", "b_genotypeSerMSer", 
            "b_genotypeGlyMSer","b_elev.pos.z",  "b_intravar.temp.z", "b_intravar.precip.z", "b_intravar.mass.z"),
              prob_outer=0.95, # 95% outer CI 
              prob=0.89, # 50% inner CI 
              point_est="mean") + # mean point est; default is median 
              labs(x="Effect on MCHC (within species)", title = "") + 
              # scale_y_discrete(labels=c("Elevation","Elevation Position", "Mass", "Temp", "Precip","Genotype (Gly-Ser)",
              #"Genotype (Ser-Ser)", "Intra Var: Mass", "Intra Var: Temp", "Intra Var: Precip")) +
              scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel fig
              theme(plot.title = element_text(family = "Helvetica", color="black", size=12)) + # font = panel header size
              theme_bw() +
              geom_point(aes(x=0.2015, y=10), color="red", size=3.35) + # Add red circle (intravar: mass)
              geom_point(aes(x=-0.2755, y=8), color="red", size=3.35) + # Add red circle (intravar: temp)
              #annotate('text', x = -1.42, y = 9.878, label='bold("*")', size=4, parse=TRUE) + # top asterisk
              #annotate('text', x = -1.42, y = 7.878, label='bold("*")', size=4, variablee=TRUE) + # bottom asterisk
              theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                    panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
              theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
              theme(text = element_text(family="Helvetica", size=10)) + # y axis labels
              ggtitle("F") + # Assign panel number/header
              theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
              plot.title = element_text(face="bold")) + # This makes panel header bold 
              # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
              theme(plot.margin = unit(c(0.4,0.4,0.0,0.4), "cm")) # top, right, bottom, left
fig3f_mchc_main
ggsave(fig3f_mchc_main, filename = "fig3f_mchc_main_IndivLevel.pdf", bg="transparent", height=7, width=9, units="in")

# MCHC inset element 
color_scheme_set("brightblue")
fig3f_mchc_inset <- mcmc_plot(spm.mchc.m4, variable=c("b_elev.z", "b_Ielev.zE2","b_temp.z", "b_precip.z", "b_mass.z", 
            "b_genotypeSerMSer", "b_genotypeGlyMSer"),
              prob_outer=0.95, # 95% outer CI
              prob=0.89, # 50% inner CI
              point_est="mean", 
              point_size=2, inner_size=1) + 
              labs(x="Effect on MCHC\n(among species)", title = "") +
              scale_y_discrete(labels=c("Elev", "Elev^2", "Temp", "Precip", "Mass", "Gen. (Ser-Ser)", 
              "Gen. (Gly-Ser)")) +
              # scale_y_discrete(labels=c("","", "", "", "", "", "", "", "", "")) + # Replace  w/ blanks for multipanel 
              theme(plot.title = element_text(family = "Helvetica", color="black", size=9)) + # font = panel header size
              theme_bw() +  
              theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                    panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
              theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
              theme(axis.text=element_text(size=5), axis.title=element_text(size=5)) + # y-axis text, x-axis text
              # theme(text = element_text(family="Helvetica", size=7), 
              #       axis.title=element_text(size=7)) + # y axis labels
              # ggtitle("F") + # Assign panel number/header
              theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
              plot.title = element_text(face="bold")) + # This makes panel header bold
              # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
              theme(plot.margin = unit(c(top=0.4, right=0.4, bottom=0.4,left=0.4), "cm"))   # top, right, bottom, left
fig3f_mchc_inset
ggsave(fig3f_mchc_inset, filename = "fig3f_mchc_inset_SpeciesMean.pdf", bg="transparent", height=3, width=6, units="in")

# MCHC PANEL E WITH INSET 
fig3f_mchc_panel <- fig3f_mchc_main + inset_element(fig3f_mchc_inset, 
                                                    right = 1.03, # RIGHT IS LEFT 
                                                    bottom = 0.58, 
                                                    left = 0.60, # LEFT IS RIGHT 
                                                    top = 1.2)
fig3f_mchc_panel


#### FINAL PLOT #####

fig3_v2_Indiv_and_SpeciesMean <- (fig3a_hb_panel + 
                                  fig3b_hct_panel + 
                                  fig3c_trbc_panel + 
                                  fig3d_mcv_panel + 
                                  fig3e_mch_panel + 
                                  fig3f_mchc_panel +
                                  plot_layout(guides = "collect"))
plot(fig3_v2_Indiv_and_SpeciesMean)
ggsave(fig3_v2_Indiv_and_SpeciesMean, filename = "Fig3_PosteriorProbPlots_Indiv_and_SpeciesMean__Plus89PercentCred_2022-01-22.pdf", bg="transparent", height=6.8, width=12.5, units="in")
```




# FIGURE 4 (6-panel fig): CELL NUMBER-SIZE TRADEOFF + CENS RESULTS

Quick workup of predictors: 
```{r}
# full dataset
# p <- ggpairs(subset(final, select = c(elev, mass, hb, hct, trbc, mcv, mch, mchc))) 
# print(p)

# Subset final dataset to just have hb, mcv, trbc, mch
final.hmt <- subset(final, select=c(species, rowID, nk, elev, hb, mcv, trbc, mch)) 
final.hmt <- na.omit(final.hmt)

# For plotting blood-O2 carrying capacity as background of plot, make variable called o2cc
# test <- final %>% mutate(o2cc = mcv * trbc * (1/3))
# As it turns out, we don't need this; see plot below

# Isolate points above and below 4,200 m
# Loop through each data subset and assign "below" to birds sampled < the breakpoint and "above" to > break point
final.hmt$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final.hmt)){
    if((final.hmt$elev[i] < 4200) ){
        final.hmt$elev_cat[i] <- "< 4,200 m"
    }else if((final.hmt$elev[i] > 4200)){ 
      final.hmt$elev_cat[i] <- ">= 4,200 m"
    }
}

# Make data subsets in case you need them 
above.hmt <- final.hmt[which(final.hmt$elev_cat == "> 4,200 m"),] # 399 individuals 
below.hmt <- final.hmt[which(final.hmt$elev_cat == "< 4,200 m"),] # 567 individuals 
```

CeNS results.
```{r}
# Load in data 
cens.ss.mean <- read.csv("cens.ss.mean_ForCeNSmodeling_06-14-21.csv")

# Load in data required to create plot 
load("Rdata_files/cens.m3_CeNS.RData") # top model  
load("Rdata_files/cens.m5_CeNS_ReducedModel.RData") # essentially tied w/ top model

# No need to load cens.ss.mean if it's already loaded
```


Calculate color gradients needed for Fig 4A (banana plot shaded by predicted O2 CC)
```{r}
# PANEL A: PREDICTIONS OF BLOOD-O2 CARRYING CAPACITY, BANANA-GRADIENT-PLOT. 
# Creating a heat map/color gradient background in ggplot isn't super straightforward. Through a back-channel google images search, I found this advice, which is exactly what we want: 
# https://stackoverflow.com/questions/46354513/diagonal-gradient-in-a-ggplot-background
# (Believe it or not, google image searching "ggplot panel.background color gradient" was more helpful than web searching)

# TRBC v. MCV w/ background shaded by heat map of blood O2 carrying capacity 
# To get a heat map of blood-O2 carrying capacity, we need to make a contour of circles to plot behind our data 
# Formula for O2 carrying capacity is: oc22 = MCV x TRBC x 1/3
# Simplest way to do this is to use plot limits (i.e. data range) to determine min and max of color gradient:
o2cc.min <- 55 * 3 * (1/3); o2cc.min # This is min MCV * min TRBC * 1/3
o2cc.max <- 152 * 11 * (1/3); o2cc.max # Max MCV * max TBBC * 1/3
# Can also do this (because no NAs, but this will be calibrated to data range vs plot range; subtly different)
# o2cc.min <- min(final.hmt$mcv) * min(final.hmt$trbc) * (1/3); o2cc.min # This is min MCV * min TRBC * 1/3
# o2cc.max <- max(final.hmt$mcv) * max(final.hmt$trbc) * (1/3); o2cc.max # Max MCV * max TBBC * 1/3

# Generate heat map color gradient of values: 
xColor <- seq(from=o2cc.min, to=o2cc.max, length.out=100) # scale of color for x; length.out = desired length of sequence
yColor <- seq(from=o2cc.min, to=o2cc.max,length.out=100) # scale of color for y 

# For both, min value min(mcv) * min(trbc) * 1/3; max = max(mcv) * max(trbc) * 1/3
# But that just gives us min/max values for our data; we want min/max for plot space 
# so multiply min(ylim)*min(xlim)*1/3 and max(ylim)*max(xlim)*1/3 to get plot space dimensions
# For xColor: 55*3*1/3; For yColor: 152*11*1/3
x <- seq(from=55, to=152, length.out=100) # on x axis; min MCV: 58.48877; max MCV: 147.6209
y <- seq(from=3, to=11, length.out=100) # on y axis; min TRBC: 3.58; max TRBC: 10.130
# This is adjusting the plot space that the background will fill 
# Essentially, this is xlim and ylim + buffer to expand beyond min and max values to fill entire plot space 
df <- cbind(expand.grid(x=xColor, y=yColor), expand.grid(x=x, y=y)) #grid for colors
colnames(df)<-c("xColor", "yColor", "x", "y")
# df$zColor <- (df$xColor+df$yColor) #femtoliter per microliter
# df$zColor <- ((df$xColor+df$yColor)/3) # Intermediate conversion step...
df$zColor <- (((df$xColor+df$yColor)/3)/10) # Scale in g/dl; /10 = shorthand to force legend; scale remains unchanged

# NOTES ON SCALE (these are kind of confusing because I moved fast in meeting w/ Chris)
# Raw  calculations of O2  carrying capacity are in femtoliters per microliter (fl/ul), but we want to  view the plot 
# in g/dl to be consistent w/ hemoglobin panel. 
# We'll use the same conversion for for Hb as for H2O; assume that one mililiter is a gram (this is mass to volume)
# one femtoliter is one trillion miiliters (1e+12) = wow, femtoliter really, really small!
# to get deciliters from microliters, divide by 100,000
# value of 600 * 100,000 = 60 million, 60 million femtoliters per deciliter
# now convert femtoliters to grams
# If we assume mass of Hb is mass of water, we're converting femtoliters to mililiters
# divide 60 million by 1 trillion 
# THEN, to account for TRBC scale, multiply by a million. Correct scale: g/dl! 
# Since the math is right and we just want to visualize the scale in g/dl, let's do shorthand
# divide df$zColor by 10 (after dividing by 3) and scale will be n correct units. Data & scale remain unchanged. 
```


PRINT FIGURE 4. 
```{r}
# FIGURE 3, PANEL A 
(Fig4_PanelA <- ggplot(final.hmt, aes(x=mcv, y=trbc)) +
    geom_raster(data=df, aes(x, y, fill=zColor))+ # can add alpha to make this lighter, but doesn't shade well
    scale_fill_viridis(option="rocket", discrete=FALSE, direction=1, name="Predicted\n[Hb]\n(g/dl)")+ 
      # direction=1 goes cool to warm; direcdtion=-1 goes warm to cool
    # scale_colour_viridis(option="rocket", discrete=FALSE, name="[Hb] (g/dl)") +
    geom_point(shape=21, colour="black", size=2.0, stroke=1, alpha=1.0) + # aliceblue, ivory1 & snow1 also nice, slightly warmer # If adding fill="black', reduce alpha to 0.3 or 0.4
    labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
         x="") + # Removed x-axis label because we don't want Panel A to have label
    theme_classic() +
   # xlim(56, 150) + ylim(2.58, 11)+
    coord_cartesian(expand = FALSE) + # FALSE = no padding for border
    theme(legend.position = "right") + 
    theme(legend.title=element_text(size=10)) +
    ggtitle("A") + # Assign panel number/header
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + #makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.2,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
)
ggsave(Fig4_PanelA, filename="Fig4_PanelA_BananaHeatGradient_2022-01-24.pdf", height=5, width=6, units="in") 


# PANELS B AND C
# See Hawkey et al. 1991 for inspiration.


#THIS ONE
# Panel B: RBC and MCV - redo of plot in Wintrobe 1933 and Hawkey et al. 1991
(Fig4_PanelB <- ggplot(final.hmt, aes(x=mcv, y=trbc, colour=hb)) +
      scale_colour_viridis(option="rocket", discrete=FALSE, name="Measured\n[Hb]\n(g/dl)") +
     # p <- p + geom_boxplot(alpha = 0.5, outlier.size=0.1) # Add this back if categorical x-axis variable 
    geom_point(size=2, alpha=1.0) +
    # scale_color_manual(values=hum_colors) +
    # stat_smooth(data=final, aes(y=trbc), size=1.2, method="lm") +
    #geom_jitter(width=0.8, height=0.8, alpha=0.8, size=2) + # jitter size should match geom_point size or it looks weird
    labs(y=expression(Total~Red~Blood~Cell~Count~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     x="Mean Cell Volume (fl)") + # Removed x-axis label because we don't want Panel A to have label
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    #scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
    theme_classic() +
    theme(legend.position = "right") + 
    theme(legend.title=element_text(size=10)) +
    ggtitle("B") + # Assign panel number/header
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + #makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.2,0.2,0.0,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
)
ggsave(Fig4_PanelB, filename="Fig4_PanelB_TRBC_vs_MCV_Scatterplot_ColoredByElev_2022-11-23_4x5.pdf", height=4, width=5, units="in") 
# Wow, this is amazing. See Hawkey et al. 1991 for some dynamite interpretation
# Essentially this means: species have either evolved many small erythrocytes or few large erythrocytes 
# Seems to be a sweet spot of a moderate number of a moderate sized erythrocyte



# Panel C: MCH and MCV - redo of plot in Hawksey et al. 1991
(Fig4_PanelC <- ggplot(final.hmt, aes(x=mcv, y=mch, colour=trbc)) +
    scale_colour_viridis(option="rocket", discrete=FALSE, name="TRBC") +
    # p <- p + geom_boxplot(alpha = 0.5, outlier.size=0.1) # Add this back if categorical x-axis variable 
    geom_point(size=2, alpha=1.0) +
    # stat_smooth(data=full1, aes(y=trbc), size=1.2, method="lm") +
    # geom_jitter(width=0.8, height=0.8, alpha=0.8, size=2) + # jitter size should match geom_point size or it looks weird
    labs(y="MCH (pg)", 
         x="MCV (fl)") + # fix RBC title/notation
    #scale_color_manual(values=c(neither="black", generalist="burlywood2", specialist="#853FCD")) + #336B87
    #scale_color_manual(values=c(neither="#8C847B", generalist="#1CD2D8", specialist="#6E1EDC")) + #336B87
    theme_classic() +
    theme(legend.position = "right") + 
    theme(legend.title=element_text(size=10)) +
    ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.0,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
)
ggsave(Fig4_PanelC, filename="Fig4_PanelC_RBC_vs_MCH_Scatterplot_ColoredByElev_2021-07-16.pdf", height=7, width=9, units="in") 


####

# REDUCED CeNS model 
# Main: Best full CeNS model (M3)
library(bayesplot)
library(brms)
color_scheme_set("viridisC") # Used viridisD for other posterior prob plots (figure 4)
fig4d_reduced <- mcmc_plot(cens.m5, variable=c("b_mass.z", "b_elev.z", "b_Ielev.zE2"),
              prob_outer=0.95, # 95% outer CI
              prob=0.89, # 50% inner CI
              point_est="mean", # mean point est; default is median
              point_size=4, inner_size=3) + # Adjust line and point thickness for inset formatting
              labs(x=expression(paste("Effect on"~italic(CeNS)~"(reduced model)"), title = "")) + 
              #labs(x="Effect on CeNS (reduced model)", title = "") + 
              scale_y_discrete(labels=c("Mass", "Elev", "Elev^2")) + 
              theme(plot.title = element_text(family = "Helvetica", color="black", size=10)) + # font = panel header size
             theme_bw() +  #theme with white background
             theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
             theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
             theme(text = element_text(family="Helvetica", size=10)) + # y axis labels 
             ggtitle("D") + # Assign panel number/header
             theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
             plot.title = element_text(face="bold"))  # This makes panel header bold 
             # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
             #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
             theme(plot.margin = unit(c(0.0,0.0,0.0,0.0), "cm"))    # top, right, bottom, left
           # theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
fig4d_reduced
ggsave(fig4d_reduced, filename = "fig4d_reduced_CensIndex_PosteriorEstimates&CredibleIntervals.pdf", bg="transparent", height=7, width=9, units="in")

####

# PANEL E: CENS V. ELEVATION 
# Get R^2 for Panel D
fig4.panelE.r2 <- glm(cens.index ~ elev + I(elev^2), data=cens.ss.mean, family=gaussian())
round((1 - (fig4.panelE.r2$deviance/fig4.panelE.r2$null.deviance)),2) # R^2 = 0.13

(fig4e <- ggplot(cens.ss.mean, aes(x=elev, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.8, alpha=0.8) + # cens.ss.mean$mcv/10 = to size by MCV but they don't look super variable
#  geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
 # scale_colour_viridis(discrete=F) + 
 # scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
   geom_smooth(method = "lm", formula = y ~ x + I(x^2), colour="black", size=0.7) + # quadratic 
  # geom_smooth(method = "lm", size = 1,colour="darkturquoise", size=1.0) + # linear 
  theme_classic() + 
  labs(x="Elevation (m)", 
       #y="Cell Number-Size Index (0-1)") + 
       y=expression(italic(CeNS))) + # CeNS = (TRBC/MCV+TRBC)
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=4000, y=0.72, size=3.2, label = "italic(R) ^ 2 == 0.13", parse = TRUE) + # Add R^2 value
  ggtitle("E") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.0,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
 )
ggsave(fig4e, filename="fig4e_CeNSindex_by_MeanElev_2023-04-19_4x5.pdf", height=4, width=5, units="in")


##### 


# PANEL F: CENS V. MASS 
 
# Get R^2 for Panel E
fig4.panelF.r2 <- glm(cens.index ~ mass.log, data=cens.ss.mean, family=gaussian())
round((1 - (fig4.panelF.r2$deviance/fig4.panelF.r2$null.deviance)),2) # R^2 = 0.07

# Scatterplot: CeNS index v. Mass 
(fig4f <- ggplot(cens.ss.mean, aes(x=mass.log, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.8, alpha=0.8) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
 # scale_colour_viridis(discrete=F) + 
 # scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
 # geom_smooth(method = "lm", formula = y ~ x + I(x^2), size = 1, colour="black", size=1.0) + # quadratic 
  geom_smooth(method = "lm", colour="black", size=0.7) + # linear 
  theme_classic() + 
  labs(x="Log Mass (g)", 
       y=expression(italic(CeNS))) + # CeNS = (TRBC/MCV+TRBC)
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=1.25, y=0.72, size=3.2, label = "italic(R) ^ 2 == 0.07", parse = TRUE) + # Add R^2 value
  ggtitle("F") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.0,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
 )
ggsave(fig4f, filename="fig4f_CeNSindex_by_Mass_2022-01-22.pdf", height=7, width=9, units="in")

# fig4EandF <- (fig4e / fig4f)
# fig4EandF

##### 

library(patchwork)
# Alternative layout option = BETTER, USE THIS ONE:
Figure4_TradeoffAndCeNS_Results <- (
                                    (Fig4_PanelA / Fig4_PanelB / Fig4_PanelC) |
                                    (fig4d_reduced / fig4e / fig4f)) + # note using full model of fig4d
                                    plot_layout(widths = c(1.5, 1.2))
                                  #  plot_layout(widths=c(2,1,1), heights = unit(c(3,5,2.5), c('cm', 'cm', 'cm'))) 
                                # widths = says make col 1 have 2x the width of column 2
                                # heights = controlling size in cm, order: 1st row, 2nd row, 3rd row
plot(Figure4_TradeoffAndCeNS_Results)
ggsave(Figure4_TradeoffAndCeNS_Results, filename = "Figure4_TradeoffAndCeNSResults_2023-04-19.pdf", bg="transparent", height=7.4, width=7.6, units="in")

# Another way to format layout (though doesn't allow for adjustment of panels E and F separately, as I wanted): 
# Figure4_TradeoffAndCeNS_Results <- (
#                                     (Fig4_PanelA / Fig4_PanelB / Fig4_PanelC) |
#                                     ((fig4d_reduced) /
#                                      (fig4e / fig4f))
#                                     ) # note using full model of fig4d
#                                   #  plot_layout(widths=c(2,1,1), heights = unit(c(3,5,2.5), c('cm', 'cm', 'cm'))) 
#                                 # widths = says make col 1 have 2x the width of column 2
#                                 # heights = controlling size in cm, order: 1st row, 2nd row, 3rd row
# plot(Figure4_TradeoffAndCeNS_Results)
```





######## SUPPLEMENTAL FIGS ######## 


# FIGURE S1: In `HumBlood_Modeling.Rmd` script. Contmaps for each blood trait. Produced in the script, hand-combined in Illustrator. 


# FIGURE S2: In `HumBlood_CeNS.Rmd` script. Single-species posterior probability plots.


# FIGURE S3: Supplementary CeNS results. 
```{r}
# PANEL A: Full model results
# Main: Best full CeNS model (M3) 
color_scheme_set("viridisC") # Used viridisD for other posterior prob plots (figure 4)
FigS3_A <- mcmc_plot(cens.m3, variable=c("b_wl.z", "b_mass.z", "b_elev.z", "b_Ielev.zE2"),
              prob_outer=0.95, # 95% outer CI
              prob=0.89, # 50% inner CI
              point_est="mean", # mean point est; default is median
              point_size=4, inner_size=3) + # Adjust line and point thickness for inset formatting
              labs(x=expression(paste("Effect on"~italic(CeNS)~"(full model)"), title = "")) + 
              #labs(x="Effect on CeNS (full model)", title = "") + 
              scale_y_discrete(labels=c("Wing\nLoading", "Mass", "Elev", "Elev^2")) + 
              theme(plot.title = element_text(family = "Helvetica", color="black", size=10)) + # font = panel header size
             theme_bw() +  #theme with white background
             theme(plot.background=element_blank(),  # eliminates background, grid lines, and chart border
                  panel.grid.major=element_blank(),  panel.grid.minor=element_blank(), panel.border=element_blank()) +
             theme(axis.line = element_line(color = 'black')) + # draws back x and y axis lines
             theme(text = element_text(family="Helvetica", size=10)) + # y axis labels 
             ggtitle("A") + # Assign panel number/header
             theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
             plot.title = element_text(face="bold"))  # This makes panel header bold 
             # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
             #theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
FigS3_A
ggsave(FigS3_A, filename = "FigS2A_CensIndex_PosteriorEstimates&CredibleIntervals.pdf", bg="transparent", height=7, width=9, units="in")


# PANEL B - CENS V. WING LOADING SCATTERPLOT
#cens.ss.mean$wl.rounded <- round(cens.ss.mean$wl, 2)
(FigS3_B <- ggplot(cens.ss.mean, aes(x=round(wl,2), y=cens.index)) + # Add shape=Clade here to get shapes to work 
  geom_point(size=1.8, alpha=0.8) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
 # scale_colour_viridis(discrete=F) + 
 # scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
 # geom_smooth(method = "lm", formula = y ~ x + I(x^2), size = 1, colour="black", size=1.0) + # quadratic 
 #  geom_smooth(method = "lm", colour="black", size=1.0) + # linear 
  theme_classic() + 
  labs(x=expression(Wing~Loading~"("~"g/cm"^{2}~")"), 
         y=expression(italic(CeNS))) + 
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.0,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
 )
ggsave(FigS3_B, filename="FigS2_B_CeNSindex_by_WingLoading_2022-01-22.pdf", height=7, width=9, units="in")


# PANEL C: Scatterplot of CeNS index v. MCV  
# Correlation matrix 
cor(cens.ss.mean[,c("cens.index", "mcv")]) # r = 0.51

# Quick linear model to get R^2
cens.mcv.link.model <- glm(cens.index ~ mcv, data=cens.ss.mean, family=gaussian(), na.action=na.fail)
(1 - (cens.mcv.link.model$deviance/cens.mcv.link.model$null.deviance)) # R^2 = 0.26

# plot
(FigS3_C <- ggplot(cens.ss.mean, aes(x=mcv, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.8, alpha=0.8) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
  scale_colour_viridis(option="rocket", discrete=FALSE, name="TRBC") + 
 # scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
 # geom_smooth(method = "lm", formula = y ~ x + I(x^2), size = 1, colour="black", size=1.0) + # quadratic 
  geom_smooth(method = "lm", colour="black", size=0.7) + # linear 
  theme_classic() + 
  #labs(x="MCV (fl)", y="CeNS") + # CeNS index = (TRBC/MCV+TRBC)
  labs(x="MCV (fl)", 
       y=expression(italic(CeNS))) + # CeNS = (TRBC/MCV+TRBC)
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=105, y=0.76, size=3.2, label = "italic(R) ^ 2 == 0.26", parse = TRUE) + # Add R^2 value
  #annotate("text", x=105, y=0.71, size=2.4, label = "italic(r) == 0.51", parse = TRUE) + # Add R^2 value
  ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.0,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
 )
ggsave(FigS3_C, filename="FigS2_C _CeNSindex_by_MCV_2022-01-22.pdf", height=7, width=9, units="in")


# PANEL D: Scatterplot: CeNS index v. TRBC  
# Correlation matrix 
cor(cens.ss.mean[,c("cens.index", "trbc")]) # r = -.44

# Quick linear model to get R^2
cens.trbc.link.model <- glm(cens.index ~ trbc, data=cens.ss.mean, family=gaussian(), na.action=na.fail)
(1 - (cens.trbc.link.model$deviance/cens.trbc.link.model$null.deviance)) # R^2 = 0.19

# PLot
(FigS3_D <- ggplot(cens.ss.mean, aes(x=trbc, y=cens.index)) + # Had previously colored by MCV
  geom_point(size=1.8, alpha=0.8) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
 scale_colour_viridis(option="rocket", discrete=FALSE, name="MCV (fl)") +  
 # scale_color_manual(values = hum_colors) +
 # geom_jitter(width=0.5, height=0.5, alpha=0.8, size=1.1) + # jitter to fill plot space 
 # geom_smooth(method = "lm", formula = y ~ x + I(x^2), size = 1, colour="black", size=1.0) + # quadratic 
  geom_smooth(method = "lm", colour="black", size=0.7) + # linear 
  theme_classic() + 
  labs(x=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     y="") + # Removed x-axis label because we don't want Panel A to have label
 # labs(x="TRBC", y="CeNS Index\n(TRBC/MCV + TRBC)") + # CeNS index = (TRBC/MCV+TRBC)
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=7.3, y=0.76, size=3.3, label = "italic(R) ^ 2 == 0.19", parse = TRUE) + # Add R^2 value 
  #annotate("text", x=7.3, y=0.71, size=2.4, label = "italic(r) == -0.44", parse = TRUE) + # Add R^2 value
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.0,0.0,0.0), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=10))
 )
ggsave(FigS3_D, filename="FigS2_D_CeNSindex_by_TRBC_2022-01-22.pdf", height=7, width=9, units="in")


#### FINAL PLOT #####

library(patchwork)
# Alternative layout option = BETTER, USE THIS ONE:
# Alternative layout option = BETTER, USE THIS ONE:
FigureS3_CensSupp <- ((FigS3_A + FigS3_B) / (FigS3_C + FigS3_D) ) #+
                                  #  plot_layout(widths=c(1,1), heights = unit(c(3,3), c('cm', 'cm', 'cm'))) 
                                # widths = says make col 1 have 2x the width of column 2
                                # heights = controlling size in cm, order: 1st row, 2nd row, 3rd row
plot(FigureS3_CensSupp)
ggsave(FigureS3_CensSupp, filename = "FigureS3_CensSupp_2022-01-22.pdf", bg="transparent", height=6.8, width=7.4, units="in")
```



# FIGURE S4: SPECIES BEHAVE DIFFERENTLY AT EXTREME ELEVS
Looking at Figure 2, there's a noticeable difference in TRBC, MCV, and MCH for hummingbird sampled below 4,200 m, and those sampled above 4,200 m. Namely: We see a marked increase in TRBC, and marked decreases in MCV and MCH. We wanted to explore this - Do we find significant differences in cell size, cell volume, and MCH between species sampled <4,200 m and >4,200 m? 

**Note, to create this figure, you must run all code for Figure 3 first (variables and settings are applied there that must be in effect in order to make this. )
```{r}
# Replicate panel 3A but highlight 'extreme' values. 

# filter dataframe to get highlight values >4,200 meters in elev 
highlight_df <- final.hmt %>% filter(elev > 4200)

# Quick plot 
# final.hmt %>% 
#   ggplot(aes(x=mcv, y=trbc)) + 
#   geom_point(alpha=0.3) +
#   geom_point(data=highlight_df, 
#              aes(x=mcv,y=trbc), 
#              color='red',
#              size=3)

# Work with the final, individual-level dataset (called 'final'). 
# Loop through dataset and assign "below4k" to birds sampled <4,000m and "above4k" to birds sampled >4,000m
final$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final)){
    if((final$elev[i] <= 4200) ){
        final$elev_cat[i] <- "<= 4,200"
    }else if((final$elev[i] > 4200)){ 
      final$elev_cat[i] <- "> 4,200"
    }
}

# Make data subsets in case you need them 
above4ksub <- final[which(final$elev_cat == "> 4,200"),] # 29 individuals of 5 species -- the high-high
below4ksub <- final[which(final$elev_cat == "<= 4,200"),] # 1,122 individuals 
# Named this way, they will be clear in plots and you don't need to sort descending

# Make data subsets for plotting to elminate weird NAs being dragged along 
final.abovebelow.trbc <- subset(final, select = c(species, rowID, nk, elev, elev_cat, trbc))
final.abovebelow.mcv <- subset(final, select = c(species, rowID, nk, elev, elev_cat, mcv))
final.abovebelow.mch <- subset(final, select = c(species, rowID, nk, elev, elev_cat, mch))

# Omit NAs to plots are clean/don't have weird missing values 
final.abovebelow.trbc <- na.omit(final.abovebelow.trbc)
final.abovebelow.mcv <- na.omit(final.abovebelow.mcv)
final.abovebelow.mch <- na.omit(final.abovebelow.mch)

# Compare differences between birds sampled < 4,200 m and > 4,200 m in elevation w/ one-way ANOVAS 
aov1 <- aov(trbc ~ elev_cat, data=final.abovebelow.trbc); summary(aov1)
aov2 <- aov(mcv ~ elev_cat, data=final.abovebelow.mcv); summary(aov2)
aov3 <- aov(mch ~ elev_cat, data=final.abovebelow.mch); summary(aov3)

# TRBC is significantly different below and >= 4,200 m (p = 0.000258 ***)
# MCV is significantly different below and >= 4,200 m (p = 0.00569 **)
# MCH is significantly different below and >= 4,200 m (p= 0.0164 *)

# And, what are rates of increase/decrease w/ elevation? 

# TRBC
above4k.trbc <- glm(trbc ~ elev, data=above4ksub, family=gaussian())
summary(above4k.trbc) # hb decreases w/ elev above 4k
Anova(above4k.trbc, type=3) # yes, TRBC increases significantly with elev > 4k
lm_diag_plots(above4k.trbc)
(1 - (above4k.trbc$deviance/above4k.trbc$null.deviance)) # R^2 = 0.224

round(above4k.trbc$coefficients[2]*100, 2) # For every 100-m increase above 4,200 m, TRBC decreases by 0.53 
round(above4k.trbc$coefficients[2]*100, 2)*(10^6) # Convert to raw # cells per microliter (can also *1000000)

# MCV
above4k.mcv <- glm(mcv ~ elev, data=above4ksub, family=gaussian())
summary(above4k.mcv) # MCV decreases w/ elev above 4k
Anova(above4k.mcv, type=3) # yes, MCV decreases significantly with elev > 4k
lm_diag_plots(above4k.mcv)
(1 - (above4k.mcv$deviance/above4k.mcv$null.deviance)) # R^2 = 0.30

round(above4k.mcv$coefficients[2]*100, 2) 

# MCH
above4k.mch <- glm(mch ~ elev, data=above4ksub, family=gaussian())
summary(above4k.mch) # MCH decreases w/ elev above 4k
Anova(above4k.mch, type=3) # yes, MCV decreases significantly with elev > 4k
lm_diag_plots(above4k.mch)
(1 - (above4k.mch$deviance/above4k.mch$null.deviance)) # R^2 = 0.355

round(above4k.mch$coefficients[2]*100, 2) 

# For every 100-m increase above 4,200 m, TRBC increases by 530,000 cells per ul
# For every 100-m increase above 4,200 m, MCV decreases by 5.52 fl 
# For every 100-m increase above 4,200 m, MCH decreases by 2.37 pg
```


FIGURE S4 PLOTS. 
```{r}
# PLOTS
# FIGURE S6, PANEL A  
(FigureS4_PanelA <- final.hmt %>% 
    ggplot(aes(x=mcv, y=trbc)) +
    geom_raster(data=df, aes(x, y, fill=zColor))+ # can add alpha to make this lighter, but doesn't shade well
    scale_fill_viridis(option="rocket", discrete=FALSE, direction=1, name="Predicted\n[Hb]\n(g/dl)")+ 
      # direction=1 goes cool to warm; direcdtion=-1 goes warm to cool
    # scale_colour_viridis(option="rocket", discrete=FALSE, name="[Hb] (g/dl)") +
    geom_point(shape=21, colour="black", size=2.0, stroke=1, alpha=1.0) + 
    # aliceblue, ivory1 & snow1 also nice, slightly warmer # If adding fill="black', reduce alpha to 0.3 or 0.4
    geom_point(data=highlight_df, aes(x=mcv,y=trbc), color="gold", size=2.0, stroke=1, alpha=1.0) + # Pts >4,200må
    labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
         x="MCV (fl)") + # Removed x-axis label because we don't want Panel A to have label
    theme_classic() +
   # xlim(56, 150) + ylim(2.58, 11)+
    coord_cartesian(expand = FALSE) + # FALSE = no padding for border
    theme(legend.position = "right") + 
    ggtitle("A") + # Assign panel number/header
    theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + #makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
    theme(plot.margin = unit(c(0.2,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left
    theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
)
ggsave(FigureS4_PanelA, filename="FigureS4_PanelA_BananaHeatGradient_ExtremeValues_2022-01-11.pdf", height=5, width=6, units="in")

# Box plot: Elev cat vs. TRBC 
(FigureS4_PanelB <- ggplot(final.abovebelow.trbc, aes(x=elev_cat, y=trbc)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.8, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("<= 4,200", "> 4,200")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs 
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=3.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        # scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold")) + # orange, blue 
        labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # Hella confusing formatting for x and y axis labels 
             x="") +
        ggtitle("B") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(plot.margin = unit(c(0.2,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=8), axis.title=element_text(size=12))
)
ggsave(FigureS4_PanelB, filename="AboveBelow_Boxplot_ElevCat-vs-TRBC_2022_01-22.pdf", height=7, width=9, units="in")

# Box plot: Elev cat vs. MCV
(FigureS4_PanelC <- ggplot(final.abovebelow.mcv, aes(x=elev_cat, y=mcv)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.8, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("<= 4,200", "> 4,200")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs    
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=3.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        #scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold")) + # orange, blue 
        labs(y="MCV (fl)", # Hella confusing formatting for x and y axis labels 
             x="Elevation (m)") +
        ggtitle("C") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(plot.margin = unit(c(0.2,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left    
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=8), axis.title=element_text(size=12))
)
ggsave(FigureS4_PanelC, filename="AboveBelow_Boxplot_ElevCat-vs-MCV_2022_01-22.pdf", height=7, width=9, units="in")

# Box plot: Elev cat vs. MCH
(FigureS4_PanelD <- ggplot(final.abovebelow.mch, aes(x=elev_cat, y=mch)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.8, alpha=0.7) +
       # geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("<= 4,200", "> 4,200")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=3.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        #scale_fill_viridis(option="cividis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold")) + # orange, blue 
        labs(y="MCH (pg)", # Hella confusing formatting for x and y axis labels 
             x="") +
        ggtitle("D") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(plot.margin = unit(c(0.0,0.2,0.2,0.2), "cm")) +  # top, right, bottom, left
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=8), axis.title=element_text(size=12))
)
ggsave(FigureS4_PanelD, filename="AboveBelow_Boxplot_ElevCat-vs-MCH_2022_01-22.pdf", height=7, width=9, units="in")


#### FINAL PLOT #####

library(patchwork)
# Alternative layout option = BETTER, USE THIS ONE:
FigureS4_4panel_ExtremeSpecies <- (FigureS4_PanelA / (FigureS4_PanelB + FigureS4_PanelC + FigureS4_PanelD) 
                                    )
plot(FigureS4_4panel_ExtremeSpecies)
ggsave(FigureS4_4panel_ExtremeSpecies, filename = "FigureS4_4panel_ExtremeSpecies_2022-01_22.pdf", bg="transparent", height=8.0, width=5.8, units="in")
# I used 8 height and 5 width for Figure 3 blood banana plots 

# You know, on second thought, I hate the way the bottom panel on this figure is squished - plots are long and skinny 
# when they could better fill plot space. I'm going to plot the top and  bottom panels separately and then modify them
# to the dimensions we want in Illustrator: 

# Remake Figure6_PanelA to dimensions we want:
ggsave(FigureS4_PanelA, filename="FigS4_PanelA_BananaHeatGradient_ExtremeValues_ForFig6Merge_2022-01-22.pdf", height=4, width=6, units="in")

# Join with: 
FigureS4_BottomPanel <- (FigureS4_PanelB + FigureS4_PanelC + FigureS4_PanelD)
plot(FigureS4_BottomPanel)
ggsave(FigureS4_BottomPanel, filename = "FigureS4_BottomPanel_2022-01_22.pdf", bg="transparent", height=4.4, width=6, units="in")

# Remember that I manually adjusted width of top and bottom panels in Illustrator to prevent bottom from looking stretchy
```

Good post on adding significance bars to plots: https://stackoverflow.com/questions/17084566/put-stars-on-ggplot-barplots-and-boxplots-to-indicate-the-level-of-significanc




# FIGURE S5: BREAK POINT ANALYSIS and mondo figure 
Our data show evidence of structure and clear patterns (i.e., slope appears to change in all plots >4,200 m, and there are no points in the bottom right-hand quadrant, for example). These observations made us wonder: Do we see evidence of a break point "threshold" in the data? And, does fitting a segmented or piecewise regression model improve model fits? 

# Break point analysis - load in the data 
Break point tutorial: https://rpubs.com/MarkusLoew/12164
```{R}
library(segmented)
library(strucchange)
library(changepoint)
library(ggsignif)

# Read in the data 
final.hb <- read.csv("HumBlood_final.hb.csv", stringsAsFactors = TRUE)
final.hct <- read.csv("HumBlood_final.hct.csv", stringsAsFactors = TRUE)
final.trbc <- read.csv("HumBlood_final.trbc.csv", stringsAsFactors = TRUE)
final.mcv <- read.csv("HumBlood_final.mcv.csv", stringsAsFactors = TRUE)
final.mch <- read.csv("HumBlood_final.mch.csv", stringsAsFactors = TRUE)
final.mchc <- read.csv("HumBlood_final.mchc.csv", stringsAsFactors = TRUE)
```


# Estimation of break points and analysis of data above and below estimated thresholds 
Based on extensive prelminary analysis (see CUT script), auto-estimating break points in the 'segmented' package doesn't work well for our data because, depending on the dataset, is not typically one glaringly obvious breakpoint. We're taking an informed manual approach described in Crawley 2007 (R book) to iteratively search for a best breakpoint(s) given our data; we'll evaluate their fit with MSE, AIC comparison, and other model diagnostics.

Piecewise regression tutorial RBloggers post (consistent w/ guidance from Crawley 2007 R book): https://www.r-bloggers.com/2012/08/r-for-ecologists-putting-together-a-piecewise-regression/

Since we don't want our break points to be heavily influenced by the tail ends of our distribution (i.e., small sample sizes at extreme high and low elevations; as well as anomalous species at both ends), we'll constrain our analysis to 400 m from the bottom and top ends of our distribution; that is, between 400 meters to 4,178 meters (versus 0 m to 4,578 m).

Must use 'elev.z', standardized elev, in all models because Hb and elev are on vastly different scales! 
Lower bound: 395 m in elev: -1.413011025 (elev.z from 'final'; standardized)
Upper bound: 4,178 m in elev: 1.421487375 (elev.z from 'final'; standardized)

# HB BREAKPOINT ANALYSIS
```{r}
# Fit basic linear model 
hb.elev.basic <- lm(hb ~ elev.z, data=final.hb); summary(hb.elev.basic)

# Estimate breaks (remember to specify standardized data)
# It's important to specify unique breaks only since there may be >1 individual sampled at the same elevation
# hb.breaks <- unique(final.hb$elev.z[which(final.hb$elev.z >= min(final.hb$elev.z)  & final.hb$elev.z <= max(final.hb$elev.z))]) # from entire distribution; don't want this
hb.breaks <- unique(final.hb$elev.z[which(final.hb$elev.z >= -1.413011025  & final.hb$elev.z <= 1.421487375)])

# Iteratively search for the best break point based on lowest residual MSE; create an empty container for MSE values from
# each model and use a forloop to run a linear regression on each possible break point. 
# This is the approach recommended by Crawley 2007 textbook and the RBloggers post. 
hb.mse <- numeric(length(hb.breaks))
for(i in 1:length(hb.breaks)){
 hb.piecewise1 <- lm(hb ~ elev.z*(elev.z < hb.breaks[i]) + elev.z*(elev.z >= hb.breaks[i]), data=final.hb)
 hb.mse[i] <- summary(hb.piecewise1)[6] # crawley has double  brackets, tutorial online has single
}
hb.mse <- as.numeric(hb.mse) 

# Plot and Examine suite of possible break points; lowest MSE is best 
# Each point represents a single unique possible break point
plot(hb.mse)
#plot(hb.mse, type="line") 

# Pick the break point(s) with the lowest error:
hb.breaks[which(hb.mse==min(hb.mse))] # literally: At which break is there the lowest residual standard error?
hb.best.break <- 0.5178646
# There is a second close value, but I'm not sure what elevation is corresponds to
 
# Run model with best break point estimate:
hb.piecewise2 <- lm(hb ~ elev.z*(elev.z < hb.best.break) + elev.z*(elev.z >= hb.best.break), data=final.hb)
summary(hb.piecewise2)

# Exract model fits 
hb.fits <- fitted(hb.piecewise2)
hb.piecewise.final <- data.frame(Elevation = final.hb$elev, Hb = hb.fits)

# Compare basic linear regression model to breakpoint model: 
summary(hb.elev.basic) 
summary(hb.piecewise2) 
# Breakpoint model has lower standard error, fewer df, and higher R^2

lm_diag_plots(hb.piecewise2)

aic(hb.elev.basic)
aic(hb.piecewise2) # top 

anova(hb.elev.basic, hb.piecewise2)
# piecewise model fits significantly better than basic linear model, so we want to use this in our plots. 

# Plot our breakpoint estimates over the distribution of the data, colored by clade (Fig. 2) 
(bp.hb1 <- ggplot(final, aes(x=elev, y=hb, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  scale_color_manual(values = hum_colors) +
  #geom_smooth(method = "loess", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  geom_line(color='black', data = hb.piecewise.final, aes(x=Elevation, y=Hb), size=0.8) + 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="[Hb] (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(bp.hb1, filename="Breakpoint_Original_Hb_2021-11-24.pdf", height=7, width=9, units="in")

# But we want to get the model plotted WITHOUT the weird connector segment and with confidence intervals. 
# Create separate data subsets for above and below the breakpoint, then try to estimate linear models separately for each

# Make a subset w/ just variables you need to avoid dragging along mysterious NAs
final.hb.sub <- subset(final.hb, select = c(species, rowID, nk, elev, elev.z, hb, Clade))
final.hb.sub <- na.omit(final.hb.sub)

# Set breakpoint
breakpoint.hb <- 2972

# Loop through each data subset and assign "below" to birds sampled < the breakpoint and "above" to > break point
final.hb.sub$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final.hb.sub)){
    if((final.hb.sub$elev[i] < breakpoint.hb) ){
        final.hb.sub$elev_cat[i] <- "< 2,972 m"
    }else if((final.hb.sub$elev[i] > breakpoint.hb)){ 
      final.hb.sub$elev_cat[i] <- ">= 2,972 m"
    }
}

# Make above and below data subsets
above.hb <- final.hb.sub[which(final.hb.sub$elev_cat == ">= 2,972 m"),] # 399 individuals 
below.hb <- final.hb.sub[which(final.hb.sub$elev_cat == "< 2,972 m"),] # 567 individuals 
# Named this way, they will be clear in plots and you don't need to sort descending

# Compare differences between birds sampled below and above break point w/ one-way ANOVAs
aov.hb <- aov(hb ~ elev_cat, data=final.hb.sub); summary(aov1) # sig 
# Hb is significantly different below and above the break point (p = 0.000455 ***)

# Now separately model below and above 
hb.above.m1 <- lm(hb ~ elev.z, data=above.hb); summary(hb.above.m1)
hb.below.m1 <- lm(hb ~ elev.z, data=below.hb); summary(hb.below.m1)
# We will use estimates from the main piecewise model but plot these pieces separately with their CIs, since they match

Anova(hb.above.m1, type=3) # Hb increases significantly with Hb > break
Anova(hb.below.m1, type=3)

# Now run model with raw, unstandardized elev to get real world estimates of change 
hb.below.raw <- glm(hb ~ elev, data=below.hb, family=gaussian())
hb.above.raw <- glm(hb ~ elev, data=above.hb, family=gaussian()) # need raw elev for real-world estimate
round(hb.below.raw$coefficients[2]*100, 2) # For every 100-m increase < 2,972 m, Hb increases by 0.06 g/dl
round(hb.above.raw$coefficients[2]*100, 2) # For every 100-m increase > 2,972 m, Hb increases by 0.09 g/dl
# So higher elevation species have lower intercepts but steeper rates of increase in Hb. 
```


# HCT BREAKPOINT ANALYSIS
```{r}
# Fit basic linear model 
hct.elev.basic <- lm(hct ~ elev.z, data=final.hct); summary(hct.elev.basic)

# Estimate breaks (remember to specify standardized data)
# It's important to specify unique breaks only since there may be >1 individual sampled at the same elevation
# hct.breaks <- unique(final.hct$elev.z[which(final.hct$elev.z >= min(final.hct$elev.z)  & final.hct$elev.z <= max(final.hct$elev.z))]) # from entire distribution; don't want this
hct.breaks <- unique(final.hct$elev.z[which(final.hct$elev.z >= -1.413011025  & final.hct$elev.z <= 1.421487375)])

# Iteratively search for the best break point based on lowest residual MSE; create an empty container for MSE values from
# each model and use a forloop to run a linear regression on each possible break point. 
# This is the approach recommended by Crawley 2007 textbook and the RBloggers post. 
hct.mse <- numeric(length(hct.breaks))
for(i in 1:length(hct.breaks)){
 hct.piecewise1 <- lm(hct ~ elev.z*(elev.z < hct.breaks[i]) + elev.z*(elev.z >= hct.breaks[i]), data=final.hct)
 hct.mse[i] <- summary(hct.piecewise1)[6] # crawley has double  brackets, tutorial online has single
}
hct.mse <- as.numeric(hct.mse) 

# Plot and Examine suite of possible break points; lowest MSE is best 
# Each point represents a single unique possible break point
plot(hct.mse)
#plot(hct.mse, type="line") 

# Pick the break point(s) with the lowest error:
hct.breaks[which(hct.mse==min(hct.mse))] # literally: At which break is there the lowest residual standard error?
hct.best.break <- -0.6142864
# There is a second close value, but I'm not sure what elevation is corresponds to
 
# Run model with best break point estimate:
hct.piecewise2 <- lm(hct ~ elev.z*(elev.z < hct.best.break) + elev.z*(elev.z >= hct.best.break), data=final.hct)
summary(hct.piecewise2)

# Exract model fits 
hct.fits <- fitted(hct.piecewise2)
hct.piecewise.final <- data.frame(Elevation = final.hct$elev, hct = hct.fits)

# Compare basic linear regression model to breakpoint model: 
summary(hct.elev.basic) 
summary(hct.piecewise2) 
# Breakpoint model has lower standard error, fewer df, and higher R^2

lm_diag_plots(hct.piecewise2)

aic(hct.elev.basic)
aic(hct.piecewise2) # top 

anova(hct.elev.basic, hct.piecewise2)
# piecewise model fits significantly better than basic linear model, so we want to use this in our plots. 

# Plot our breakpoint estimates over the distribution of the data, colored by clade (Fig. 2) 
(p1.2 <- ggplot(final, aes(x=elev, y=hct, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  scale_color_manual(values = hum_colors) +
  #geom_smooth(method = "loess", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  geom_line(color='black', data = hct.piecewise.final, aes(x=Elevation, y=hct), size=0.8) + 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Hct %") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p1.2, filename="Breakpoint_Best_hct_2021-11-24.pdf", height=7, width=9, units="in")

# But we want to get the model plotted WITHOUT the weird connector segment and with confidence intervals. 
# Create separate data subsets for above and below the breakpoint, then try to estimate linear models separately for each

# Make a subset w/ just variables you need to avoid dragging along mysterious NAs
final.hct.sub <- subset(final.hct, select = c(species, rowID, nk, elev, elev.z, hct, Clade))
final.hct.sub <- na.omit(final.hct.sub)

# Set breakpoint
breakpoint.hct <- 1461

# Loop through each data subset and assign "below" to birds sampled < the breakpoint and "above" to > break point
final.hct.sub$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final.hct.sub)){
    if((final.hct.sub$elev[i] < breakpoint.hct) ){
        final.hct.sub$elev_cat[i] <- "< 1,461 m"
    }else if((final.hct.sub$elev[i] > breakpoint.hct)){ 
      final.hct.sub$elev_cat[i] <- ">= 1,461 m"
    }
}

# Make above and below data subsets
above.hct <- final.hct.sub[which(final.hct.sub$elev_cat == ">= 1,461 m"),] # 727 individuals 
below.hct <- final.hct.sub[which(final.hct.sub$elev_cat == "< 1,461 m"),] # 358 individuals 
# Named this way, they will be clear in plots and you don't need to sort descending

# Compare differences between birds sampled below and above break point w/ one-way ANOVAs
aov.hct <- aov(hct ~ elev_cat, data=final.hct.sub); summary(aov1) # sig 
# hct is significantly different below and above the break point (p = 0.000455 ***)

# Now separately model below and above 
hct.above.m1 <- lm(hct ~ elev.z, data=above.hct); summary(hct.above.m1)
hct.below.m1 <- lm(hct ~ elev.z, data=below.hct); summary(hct.below.m1)
# We will use estimates from the main piecewise model but plot these pieces separately with their CIs, since they match

Anova(hct.above.m1, type=3) # Not significant
Anova(hct.below.m1, type=3)

# Now run model with raw, unstandardized elev to get real world estimates of change 
hct.below.raw <- glm(hct ~ elev, data=below.hct, family=gaussian())
hct.above.raw <- glm(hct ~ elev, data=above.hct, family=gaussian()) # need raw elev for real-world estimate
round(hct.below.raw$coefficients[2]*100, 2) # For every 100-m increase < 1,461 m, hct decreases by 0.06%
round(hct.above.raw$coefficients[2]*100, 2) # For every 100-m increase > 1,461 m, hct increases by 0.01%
# So higher elevation species have lower intercepts but steeper rates of increase in hct. 
```


# TRBC BREAKPOINT ANALYSIS
```{r}
# Fit basic linear model 
trbc.elev.basic <- lm(trbc ~ elev.z, data=final.trbc); summary(trbc.elev.basic)

# Estimate breaks (remember to specify standardized data)
# It's important to specify unique breaks only since there may be >1 individual sampled at the same elevation
# trbc.breaks <- unique(final.trbc$elev.z[which(final.trbc$elev.z >= min(final.trbc$elev.z)  & final.trbc$elev.z <= max(final.trbc$elev.z))]) # from entire distribution; don't want this
trbc.breaks <- unique(final.trbc$elev.z[which(final.trbc$elev.z >= -1.413011025  & final.trbc$elev.z <= 1.421487375)])

# Iteratively search for the best break point based on lowest residual MSE; create an empty container for MSE values from
# each model and use a forloop to run a linear regression on each possible break point. 
# This is the approach recommended by Crawley 2007 textbook and the RBloggers post. 
trbc.mse <- numeric(length(trbc.breaks))
for(i in 1:length(trbc.breaks)){
 trbc.piecewise1 <- lm(trbc ~ elev.z*(elev.z < trbc.breaks[i]) + elev.z*(elev.z >= trbc.breaks[i]), data=final.trbc)
 trbc.mse[i] <- summary(trbc.piecewise1)[6] # crawley has double  brackets, tutorial online has single
}
trbc.mse <- as.numeric(trbc.mse) 

# Plot and Examine suite of possible break points; lowest MSE is best 
# Each point represents a single unique possible break point
plot(trbc.mse)
#plot(trbc.mse, type="line") 

# Pick the break point(s) with the lowest error:
trbc.breaks[which(trbc.mse==min(trbc.mse))] # literally: At which break is there the lowest residual standard error?
trbc.best.break <- 0.928466
# There is a second close value, but I'm not sure what elevation is corresponds to
 
# Run model with best break point estimate:
trbc.piecewise2 <- lm(trbc ~ elev.z*(elev.z < trbc.best.break) + elev.z*(elev.z >= trbc.best.break), data=final.trbc)
summary(trbc.piecewise2)

# Exract model fits 
trbc.fits <- fitted(trbc.piecewise2)
trbc.piecewise.final <- data.frame(Elevation = final.trbc$elev, trbc = trbc.fits)

# Compare basic linear regression model to breakpoint model: 
summary(trbc.elev.basic) 
summary(trbc.piecewise2) 
# Breakpoint model has lower standard error, fewer df, and higher R^2

lm_diag_plots(trbc.piecewise2)

aic(trbc.elev.basic)
aic(trbc.piecewise2) # top 

anova(trbc.elev.basic, trbc.piecewise2)
# piecewise model fits better, but not significantly so, than the linear model

# Plot our breakpoint estimates over the distribution of the data, colored by clade (Fig. 2) 
(p1.2 <- ggplot(final, aes(x=elev, y=trbc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  scale_color_manual(values = hum_colors) +
  #geom_smooth(method = "loess", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  geom_line(color='black', data = trbc.piecewise.final, aes(x=Elevation, y=trbc), size=0.8) + 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="trbc") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p1.2, filename="Breakpoint_Best_trbc_2021-11-24.pdf", height=7, width=9, units="in")

# But we want to get the model plotted WITHOUT the weird connector segment and with confidence intervals. 
# Create separate data subsets for above and below the breakpoint, then try to estimate linear models separately for each

# Make a subset w/ just variables you need to avoid dragging along mysterious NAs
final.trbc.sub <- subset(final.trbc, select = c(species, rowID, nk, elev, elev.z, trbc, Clade))
final.trbc.sub <- na.omit(final.trbc.sub)

# Set breakpoint
breakpoint.trbc <- 3520

# Loop through each data subset and assign "below" to birds sampled < the breakpoint and "above" to > break point
final.trbc.sub$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final.trbc.sub)){
    if((final.trbc.sub$elev[i] < breakpoint.trbc) ){
        final.trbc.sub$elev_cat[i] <- "< 3,520 m"
    }else if((final.trbc.sub$elev[i] > breakpoint.trbc)){ 
      final.trbc.sub$elev_cat[i] <- ">= 3,520 m"
    }
}

# Make above and below data subsets
above.trbc <- final.trbc.sub[which(final.trbc.sub$elev_cat == ">= 3,520 m"),] # 155 individuals 
below.trbc <- final.trbc.sub[which(final.trbc.sub$elev_cat == "< 3,520 m"),] # 568 individuals 
# Named this way, they will be clear in plots and you don't need to sort descending

# Compare differences between birds sampled below and above break point w/ one-way ANOVAs
aov.trbc <- aov(trbc ~ elev_cat, data=final.trbc.sub); summary(aov1) # sig 
# trbc is significantly different below and above the break point (p = 0.000455 ***)

# Now separately model below and above 
trbc.above.m1 <- lm(trbc ~ elev.z, data=above.trbc); summary(trbc.above.m1)
trbc.below.m1 <- lm(trbc ~ elev.z, data=below.trbc); summary(trbc.below.m1)
# We will use estimates from the main piecewise model but plot these pieces separately with their CIs, since they match

Anova(trbc.above.m1, type=3) # Not significant
Anova(trbc.below.m1, type=3) # significant

# Now run model with raw, unstandardized elev to get real world estimates of change 
trbc.below.raw <- glm(trbc ~ elev, data=below.trbc, family=gaussian())
trbc.above.raw <- glm(trbc ~ elev, data=above.trbc, family=gaussian()) # need raw elev for real-world estimate
round(trbc.below.raw$coefficients[2]*100, 2) # For every 100-m increase < 1,461 m, trbc decreases by 0.01 units
round(trbc.above.raw$coefficients[2]*100, 2) # For every 100-m increase > 1,461 m, trbc increases by 0.07 units

# Convert to raw # cells per microliter (can also *1000000)
round(trbc.below.raw$coefficients[2]*100, 2)*(10^6) 
round(trbc.above.raw$coefficients[2]*100, 2)*(10^6) 
# For every 100-m increase <3,520 m, TRBC increases by 10,000 cells per ul
# For every 100-m increase >3,520 m, TRBC increases by 70,000 cells per ul
```


# MCV BREAKPOINT ANALYSIS
```{r}
# Fit basic linear model 
mcv.elev.basic <- lm(mcv ~ elev.z, data=final.mcv); summary(mcv.elev.basic)

# Estimate breaks (remember to specify standardized data)
# It's important to specify unique breaks only since there may be >1 individual sampled at the same elevation
# mcv.breaks <- unique(final.mcv$elev.z[which(final.mcv$elev.z >= min(final.mcv$elev.z)  & final.mcv$elev.z <= max(final.mcv$elev.z))]) # from entire distribution; don't want this
mcv.breaks <- unique(final.mcv$elev.z[which(final.mcv$elev.z >= -1.413011025  & final.mcv$elev.z <= 1.421487375)])

# Iteratively search for the best break point based on lowest residual MSE; create an empty container for MSE values from
# each model and use a forloop to run a linear regression on each possible break point. 
# This is the approach recommended by Crawley 2007 textbook and the RBloggers post. 
mcv.mse <- numeric(length(mcv.breaks))
for(i in 1:length(mcv.breaks)){
 mcv.piecewise1 <- lm(mcv ~ elev.z*(elev.z < mcv.breaks[i]) + elev.z*(elev.z >= mcv.breaks[i]), data=final.mcv)
 mcv.mse[i] <- summary(mcv.piecewise1)[6] # crawley has double  brackets, tutorial online has single
}
mcv.mse <- as.numeric(mcv.mse) 

# Plot and Examine suite of possible break points; lowest MSE is best 
# Each point represents a single unique possible break point
plot(mcv.mse)
#plot(mcv.mse, type="line") 

# Pick the break point(s) with the lowest error:
mcv.breaks[which(mcv.mse==min(mcv.mse))] # literally: At which break is there the lowest residual standard error?
mcv.best.break <- -0.187201
# There is a second close value, but I'm not sure what elevation is corresponds to
 
# Run model with best break point estimate:
mcv.piecewise2 <- lm(mcv ~ elev.z*(elev.z < mcv.best.break) + elev.z*(elev.z >= mcv.best.break), data=final.mcv)
summary(mcv.piecewise2)

# Exract model fits 
mcv.fits <- fitted(mcv.piecewise2)
mcv.piecewise.final <- data.frame(Elevation = final.mcv$elev, mcv = mcv.fits)

# Compare basic linear regression model to breakpoint model: 
summary(mcv.elev.basic) 
summary(mcv.piecewise2) 
# Breakpoint model has lower standard error, fewer df, and higher R^2

lm_diag_plots(mcv.piecewise2)

aic(mcv.elev.basic)
aic(mcv.piecewise2) # top 

anova(mcv.elev.basic, mcv.piecewise2)
# piecewise model fits significantly better than basic linear model, so we want to use this in our plots. 

# Plot our breakpoint estimates over the distribution of the data, colored by clade (Fig. 2) 
(p1.2 <- ggplot(final, aes(x=elev, y=mcv, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  scale_color_manual(values = hum_colors) +
  #geom_smooth(method = "loess", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  geom_line(color='black', data = mcv.piecewise.final, aes(x=Elevation, y=mcv), size=0.8) + 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="mcv") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p1.2, filename="Breakpoint_Best_mcv_2021-11-24.pdf", height=7, width=9, units="in")

# But we want to get the model plotted WITHOUT the weird connector segment and with confidence intervals. 
# Create separate data subsets for above and below the breakpoint, then try to estimate linear models separately for each

# Make a subset w/ just variables you need to avoid dragging along mysterious NAs
final.mcv.sub <- subset(final.mcv, select = c(species, rowID, nk, elev, elev.z, mcv, Clade))
final.mcv.sub <- na.omit(final.mcv.sub)

# Set breakpoint
breakpoint.mcv <- 2031

# Loop through each data subset and assign "below" to birds sampled < the breakpoint and "above" to > break point
final.mcv.sub$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final.mcv.sub)){
    if((final.mcv.sub$elev[i] < breakpoint.mcv) ){
        final.mcv.sub$elev_cat[i] <- "< 2,031 m"
    }else if((final.mcv.sub$elev[i] > breakpoint.mcv)){ 
      final.mcv.sub$elev_cat[i] <- ">= 2,031 m"
    }
}

# Make above and below data subsets
above.mcv <- final.mcv.sub[which(final.mcv.sub$elev_cat == ">= 2,031 m"),] # 727 individuals 
below.mcv <- final.mcv.sub[which(final.mcv.sub$elev_cat == "< 2,031 m"),] # 358 individuals 
# Named this way, they will be clear in plots and you don't need to sort descending

# Compare differences between birds sampled below and above break point w/ one-way ANOVAs
aov.mcv <- aov(mcv ~ elev_cat, data=final.mcv.sub); summary(aov1) # sig 
# mcv is significantly different below and above the break point (p = 0.000455 ***)

# Now separately model below and above 
mcv.above.m1 <- lm(mcv ~ elev.z, data=above.mcv); summary(mcv.above.m1)
mcv.below.m1 <- lm(mcv ~ elev.z, data=below.mcv); summary(mcv.below.m1)
# We will use estimates from the main piecewise model but plot these pieces separately with their CIs, since they match

Anova(mcv.above.m1, type=3) # Sig
Anova(mcv.below.m1, type=3) # Not sig

# Now run model with raw, unstandardized elev to get real world estimates of change 
mcv.below.raw <- glm(mcv ~ elev, data=below.mcv, family=gaussian())
mcv.above.raw <- glm(mcv ~ elev, data=above.mcv, family=gaussian()) # need raw elev for real-world estimate
round(mcv.below.raw$coefficients[2]*100, 2) # For every 100-m increase < 2,031 m, mcv decreases by 0.09 fl
round(mcv.above.raw$coefficients[2]*100, 2) # For every 100-m increase > 2,031 m, mcv decreases by 0.44 fl
# So higher elevation species have lower intercepts but steeper rates of increase in mcv. 
```


# MCH BREAKPOINT ANALYSIS
```{r}
# Fit basic linear model 
mch.elev.basic <- lm(mch ~ elev.z, data=final.mch); summary(mch.elev.basic)

# Estimate breaks (remember to specify standardized data)
# It's important to specify unique breaks only since there may be >1 individual sampled at the same elevation
# mch.breaks <- unique(final.mch$elev.z[which(final.mch$elev.z >= min(final.mch$elev.z)  & final.mch$elev.z <= max(final.mch$elev.z))]) # from entire distribution; don't want this
mch.breaks <- unique(final.mch$elev.z[which(final.mch$elev.z >= -1.413011025  & final.mch$elev.z <= 1.421487375)])

# Iteratively search for the best break point based on lowest residual MSE; create an empty container for MSE values from
# each model and use a forloop to run a linear regression on each possible break point. 
# This is the approach recommended by Crawley 2007 textbook and the RBloggers post. 
mch.mse <- numeric(length(mch.breaks))
for(i in 1:length(mch.breaks)){
 mch.piecewise1 <- lm(mch ~ elev.z*(elev.z < mch.breaks[i]) + elev.z*(elev.z >= mch.breaks[i]), data=final.mch)
 mch.mse[i] <- summary(mch.piecewise1)[6] # crawley has double  brackets, tutorial online has single
}
mch.mse <- as.numeric(mch.mse) 

# Plot and Examine suite of possible break points; lowest MSE is best 
# Each point represents a single unique possible break point
plot(mch.mse)
#plot(mch.mse, type="line") 

# Pick the break point(s) with the lowest error:
mch.breaks[which(mch.mse==min(mch.mse))] # literally: At which break is there the lowest residual standard error?
mch.best.break <- 0.6467395
# There is a second close value, but I'm not sure what elevation is corresponds to
 
# Run model with best break point estimate:
mch.piecewise2 <- lm(mch ~ elev.z*(elev.z < mch.best.break) + elev.z*(elev.z >= mch.best.break), data=final.mch)
summary(mch.piecewise2)

# Exract model fits 
mch.fits <- fitted(mch.piecewise2)
mch.piecewise.final <- data.frame(Elevation = final.mch$elev, mch = mch.fits)

# Compare basic linear regression model to breakpoint model: 
summary(mch.elev.basic) 
summary(mch.piecewise2) 
# Breakpoint model has lower standard error, fewer df, and higher R^2

lm_diag_plots(mch.piecewise2)

aic(mch.elev.basic)
aic(mch.piecewise2) # top 

anova(mch.elev.basic, mch.piecewise2)
# piecewise model fits significantly better than basic linear model, so we want to use this in our plots. 

# Plot our breakpoint estimates over the distribution of the data, colored by clade (Fig. 2) 
(p1.2 <- ggplot(final, aes(x=elev, y=mch, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  scale_color_manual(values = hum_colors) +
  #geom_smooth(method = "loess", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  geom_line(color='black', data = mch.piecewise.final, aes(x=Elevation, y=mch), size=0.8) + 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="mch") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p1.2, filename="Breakpoint_Best_mch_2021-11-24.pdf", height=7, width=9, units="in")

# But we want to get the model plotted WITHOUT the weird connector segment and with confidence intervals. 
# Create separate data subsets for above and below the breakpoint, then try to estimate linear models separately for each

# Make a subset w/ just variables you need to avoid dragging along mysterious NAs
final.mch.sub <- subset(final.mch, select = c(species, rowID, nk, elev, elev.z, mch, Clade))
final.mch.sub <- na.omit(final.mch.sub)

# Set breakpoint
breakpoint.mch <- 3144

# Loop through each data subset and assign "below" to birds sampled < the breakpoint and "above" to > break point
final.mch.sub$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final.mch.sub)){
    if((final.mch.sub$elev[i] < breakpoint.mch) ){
        final.mch.sub$elev_cat[i] <- "< 3,144 m"
    }else if((final.mch.sub$elev[i] > breakpoint.mch)){ 
      final.mch.sub$elev_cat[i] <- ">= 3,144 m"
    }
}

# Make above and below data subsets
above.mch <- final.mch.sub[which(final.mch.sub$elev_cat == ">= 3,144 m"),] # 727 individuals 
below.mch <- final.mch.sub[which(final.mch.sub$elev_cat == "< 3,144 m"),] # 358 individuals 
# Named this way, they will be clear in plots and you don't need to sort descending

# Compare differences between birds sampled below and above break point w/ one-way ANOVAs
aov.mch <- aov(mch ~ elev_cat, data=final.mch.sub); summary(aov1) # sig 
# mch is significantly different below and above the break point (p = 0.000455 ***)

# Now separately model below and above 
mch.above.m1 <- lm(mch ~ elev.z, data=above.mch); summary(mch.above.m1)
mch.below.m1 <- lm(mch ~ elev.z, data=below.mch); summary(mch.below.m1)
# We will use estimates from the main piecewise model but plot these pieces separately with their CIs, since they match

Anova(mch.above.m1, type=3) # Not significant
Anova(mch.below.m1, type=3) # Significant

# Now run model with raw, unstandardized elev to get real world estimates of change 
mch.below.raw <- glm(mch ~ elev, data=below.mch, family=gaussian())
mch.above.raw <- glm(mch ~ elev, data=above.mch, family=gaussian()) # need raw elev for real-world estimate
round(mch.below.raw$coefficients[2]*100, 2) # For every 100-m increase < 3,144 m, mch increases by 0.15 pg
round(mch.above.raw$coefficients[2]*100, 2) # For every 100-m increase > 3,144 m, mch increases by 0.16 pg
# So higher elevation species have lower intercepts but steeper rates of increase in mch. 
```


# MCHC BREAKPOINT ANALYSIS
```{r}
# Fit basic linear model 
mchc.elev.basic <- lm(mchc ~ elev.z, data=final.mchc); summary(mchc.elev.basic)

# Estimate breaks (remember to specify standardized data)
# It's important to specify unique breaks only since there may be >1 individual sampled at the same elevation
# mchc.breaks <- unique(final.mchc$elev.z[which(final.mchc$elev.z >= min(final.mchc$elev.z)  & final.mchc$elev.z <= max(final.mchc$elev.z))]) # from entire distribution; don't want this
mchc.breaks <- unique(final.mchc$elev.z[which(final.mchc$elev.z >= -1.413011025  & final.mchc$elev.z <= 1.421487375)])

# Iteratively search for the best break point based on lowest residual MSE; create an empty container for MSE values from
# each model and use a forloop to run a linear regression on each possible break point. 
# This is the approach recommended by Crawley 2007 textbook and the RBloggers post. 
mchc.mse <- numeric(length(mchc.breaks))
for(i in 1:length(mchc.breaks)){
 mchc.piecewise1 <- lm(mchc ~ elev.z*(elev.z < mchc.breaks[i]) + elev.z*(elev.z >= mchc.breaks[i]), data=final.mchc)
 mchc.mse[i] <- summary(mchc.piecewise1)[6] # crawley has double  brackets, tutorial online has single
}
mchc.mse <- as.numeric(mchc.mse) 

# Plot and Examine suite of possible break points; lowest MSE is best 
# Each point represents a single unique possible break point
plot(mchc.mse)
#plot(mchc.mse, type="line") 

# Pick the break point(s) with the lowest error:
mchc.breaks[which(mchc.mse==min(mchc.mse))] # literally: At which break is there the lowest residual standard error?
mchc.best.break <- 1.240163
# There is a second close value, but I'm not sure what elevation is corresponds to
 
# Run model with best break point estimate:
mchc.piecewise2 <- lm(mchc ~ elev.z*(elev.z < mchc.best.break) + elev.z*(elev.z >= mchc.best.break), data=final.mchc)
summary(mchc.piecewise2)

# Exract model fits 
mchc.fits <- fitted(mchc.piecewise2)
mchc.piecewise.final <- data.frame(Elevation = final.mchc$elev, mchc = mchc.fits)

# Compare basic linear regression model to breakpoint model: 
summary(mchc.elev.basic) 
summary(mchc.piecewise2) 
# Breakpoint model has lower standard error, fewer df, and higher R^2

lm_diag_plots(mchc.piecewise2)

aic(mchc.elev.basic)
aic(mchc.piecewise2) # top 

anova(mchc.elev.basic, mchc.piecewise2)
# piecewise model fits significantly better than basic linear model, so we want to use this in our plots. 

# Plot our breakpoint estimates over the distribution of the data, colored by clade (Fig. 2) 
(p1.2 <- ggplot(final, aes(x=elev, y=mchc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=1.1, alpha=0.8) +
  scale_color_manual(values = hum_colors) +
  #geom_smooth(method = "loess", formula = y ~ x, colour="black", size=1.0) + # line used to be dark turquoise 
  geom_line(color='black', data = mchc.piecewise.final, aes(x=Elevation, y=mchc), size=0.8) + 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="mchc") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(p1.2, filename="Breakpoint_Best_mchc_2021-11-24.pdf", height=7, width=9, units="in")

# But we want to get the model plotted WITHOUT the weird connector segment and with confidence intervals. 
# Create separate data subsets for above and below the breakpoint, then try to estimate linear models separately for each

# Make a subset w/ just variables you need to avoid dragging along mysterious NAs
final.mchc.sub <- subset(final.mchc, select = c(species, rowID, nk, elev, elev.z, mchc, Clade))
final.mchc.sub <- na.omit(final.mchc.sub)

# Set breakpoint
breakpoint.mchc <- 3936

# Loop through each data subset and assign "below" to birds sampled < the breakpoint and "above" to > break point
final.mchc.sub$elev_cat <- NA # instantiate new column
for(i in 1:nrow(final.mchc.sub)){
    if((final.mchc.sub$elev[i] < breakpoint.mchc) ){
        final.mchc.sub$elev_cat[i] <- "< 3,936 m"
    }else if((final.mchc.sub$elev[i] > breakpoint.mchc)){ 
      final.mchc.sub$elev_cat[i] <- ">= 3,936 m"
    }
}

# Make above and below data subsets
above.mchc <- final.mchc.sub[which(final.mchc.sub$elev_cat == ">= 3,936 m"),] # 79individuals 
below.mchc <- final.mchc.sub[which(final.mchc.sub$elev_cat == "< 3,936 m"),] # 843 individuals 
# Named this way, they will be clear in plots and you don't need to sort descending

# Compare differences between birds sampled below and above break point w/ one-way ANOVAs
aov.mchc <- aov(mchc ~ elev_cat, data=final.mchc.sub); summary(aov1) # sig 
# mchc is significantly different below and above the break point (p = 0.000455 ***)

# Now separately model below and above 
mchc.above.m1 <- lm(mchc ~ elev.z, data=above.mchc); summary(mchc.above.m1)
mchc.below.m1 <- lm(mchc ~ elev.z, data=below.mchc); summary(mchc.below.m1)
# We will use estimates from the main piecewise model but plot these pieces separately with their CIs, since they match

Anova(mchc.above.m1, type=3) # Sig
Anova(mchc.below.m1, type=3) # Not sig

# Now run model with raw, unstandardized elev to get real world estimates of change 
mchc.below.raw <- glm(mchc ~ elev, data=below.mchc, family=gaussian())
mchc.above.raw <- glm(mchc ~ elev, data=above.mchc, family=gaussian()) # need raw elev for real-world estimate
round(mchc.below.raw$coefficients[2]*100, 2) # For every 100-m increase < 3,936 m, mchc decreases by 0.01 g/dl
round(mchc.above.raw$coefficients[2]*100, 2) # For every 100-m increase > 3,936 m, mchc decreases by 0.29 g/dl
```


# FIGURE S5: COMBINED INDIVIDUAL-LEVEL AND SPECIES MEAN DATA IN ONE MASTER PLOT 
Figure S5 - A 6-panel fig w/ relationship between predictors and elevation 
```{r}
# IMPORTANT! Read in 'final' dataset, set levels and colors for hum_colors above

# Read in species mean datasets (made in HumBlood_Modeling_SpeciesMeans.Rmd)
read.csv("AllHummingbirdBlood_Wrangled_withBioClim_andPCA_03-01-21.csv")
spm.hb <- read.csv("HumBlood_SpeciesMeans_spm.hb_06-14-21.csv")
spm.hct <- read.csv("HumBlood_SpeciesMeans_spm.hct_06-14-21.csv")
spm.trbc <- read.csv("HumBlood_SpeciesMeans_spm.trbc_06-14-21.csv")
spm.mcv <- read.csv("HumBlood_SpeciesMeans_spm.mcv_06-14-21.csv")
spm.mch <- read.csv("HumBlood_SpeciesMeans_spm.mch_06-14-21.csv")
spm.mchc <- read.csv("HumBlood_SpeciesMeans_spm.mchc_06-14-21.csv")

# Set levels. I'm sure there's a FAR less tedious and ultra-repetitive way to do this, but I'm in the mood to make plots, 
# not struggle w/ syntax challenges. 
# NO MOUNTAIN-GEM IN SPECIES MEAN DATA (because of n=1)
spm.hb$Clade <- factor(spm.hb$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.hct$Clade <- factor(spm.hct$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.trbc$Clade <- factor(spm.trbc$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.mcv$Clade <- factor(spm.mcv$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.mch$Clade <- factor(spm.mch$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))
spm.mchc$Clade <- factor(spm.mchc$Clade, levels = c("Coquette", "Brilliant", "Patagona", "Bee", 
                                                "Emerald", "Mango", "Topaz", "Hermit"))

# species_mean_colors
spec_mean_colors <- c( "#04A5A5", # Turquoise (Coquette)
                 "#9E9285", # brown (Brilliant)
                 "#EE4024", # red (Patagona)
                 "#F4CD0B", # Gold (Bee)
                # "#E786B7", # pink (Mountain-gem)
                 "#97B25B", # green (Emerald)  
                 "#F79E0F", # orange (Mango)
                 "#03446B", # navy blue (Topaz)
                 "#916EF9" # purple (Hermit)
                    )

# Make sure you have no NAs in your data - for some reason this keeps NOT running above where it should; 
# messes up the box plots 
final.hb.sub <- na.omit(final.hb.sub)
final.hct.sub <- na.omit(final.hct.sub)
final.trbc.sub <- na.omit(final.trbc.sub)
final.mcv.sub <- na.omit(final.mcv.sub)
final.mch.sub <- na.omit(final.mch.sub)
final.mchc.sub <- na.omit(final.mchc.sub)
```


# Plot master Figure S5
```{r}
# Scatterplots of each of the 6 blood characteristics with elevation, colored by clade
# Code to set shapes by clade, but this was totally overwhelming
  # scale_shape_manual(values=c("Bee"=4, "Brilliant"=17, "Coquette"=15, # Shapes are a little overwhelming, honestly
  #                             "Emerald"=18, "Hermit"=25, "Mango"=8, "Mountain-gem"=3,
  #                             "Patagona"=16, "Topaz"=9)) +
 #   scale_shape_manual(values=shapes) + # Another way to do this is set colors above 

# PANEL A: Scatterplot: Hb v. elev 
(p1 <- ggplot(final, aes(x=elev, y=hb, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  # geom_vline(xintercept = 2972, linetype="dashed", color = "black", size=0.5) + # Break point; too busy, interferes
  geom_point(size=1.1, alpha=0.8) +
 # facet_wrap(~Clade) + 
  #scale_colour_viridis(discrete=TRUE) + 
  scale_color_manual(values = hum_colors) +
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above.hb) + # Segment 1  
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=below.hb) + # Segment 2 
  theme_classic() + 
  labs(x=" ", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="[Hb] (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") + 
  annotate("text", x=4000, y=24.2, size=3, label = "italic(R) ^ 2==0.09", parse = TRUE) + # Add R^2 value from glm
  ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
   theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
 )
ggsave(p1, filename="p1_ScatterplotBasic_Elev_by_Hb.pdf", height=7, width=9, units="in")
# The 166 rows of missing values come from many NA Hb values; don't worry about these 

# Adjusted R-squared from piecewise model:  0.09381

##### 

# PANEL B: Scatterplot: Hct v. elev
(p2 <- ggplot(final, aes(x=elev, y=hct, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  # scale_colour_viridis(discrete=TRUE) + 
  #facet_wrap(~Clade) + 
  scale_color_manual(values = hum_colors) +
#  geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above.hct) + # Segment 1  
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=below.hct) + # Segment 2 
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4000, y=78, size=3, label = "italic(R) ^ 2==0.07", parse = TRUE) + # Add R^2 value from glm
  ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
   theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
 )
ggsave(p2, filename="p3_ScatterplotBasic_Elev_by_Hct_2021-11-24.pdf", height=7, width=9, units="in")

summary(hct.piecewise2)
# Adjusted R-squared from piecewise model:  0.07 

#####

# PANEL B: Scatterplot: trbc v. elev
(p3 <- ggplot(final, aes(x=elev, y=trbc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  #facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
  #geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.8) + # jitter to fill plot space 
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above.trbc) + # Segment 1  
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=below.trbc) + # Segment 2  turquoise
  theme_classic() + 
   labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # horrible expression language for getting formatted y axis 
     x="") + # Removed "Elevation (m)" because we don't want Panel A to have label
    # in expression world, ~ is space and mathematical symbols like parenthesis need to be set of w/ " " 
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4000, y=9.8, size=3, label = "italic(R) ^ 2==0.02", parse = TRUE) + # Add R^2 value from glm
  ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,right=5,0.0,0.2), "cm")) +  # top, right, bottom, left
   theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
 )
ggsave(p3, filename="p3_ScatterplotBasic_Elev_by_TRBC_2021-11-24.pdf", height=7, width=9, units="in")

summary(trbc.piecewise2)
# Adjusted R-squared from piecewise model:  0.02 

#####

# PANEL B: Scatterplot: mcv v. elev
(p4 <- ggplot(final, aes(x=elev, y=mcv, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
  # facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above.mcv) + # Segment 1  
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=below.mcv) + # Segment 2  turquoise 
  theme_classic() + 
  labs(x="", # Removed "Elevation (m)" because we don't want Panel A to have label
       y="MCV (fl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4000, y=155, size=3, label = "italic(R) ^ 2==0.02", parse = TRUE) + # Add R^2 value from glm
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  #theme(legend.box.spacing = unit(c(0.2,5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(plot.margin = unit(c(0.0,0.2,0.0,0.1), "cm")) +  # top, right, bottom, left
   theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
 )
ggsave(p4, filename="p4_ScatterplotBasic_Elev_by_MCV_2021-11-24.pdf", height=7, width=9, units="in")

summary(mcv.piecewise2)
# Adjusted R-squared from piecewise model:  0.02

#######

# PANEL E: Scatterplot: MCH v. elev
(p5 <- ggplot(final, aes(x=elev, y=mch, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
 # facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above.mch) + # Segment 1  
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=below.mch) + # Segment 2  turquoise 
  theme_classic() + 
  labs(x="", y="MCH (pg)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  annotate("text", x=4000, y=52, size=3, label = "italic(R) ^ 2==0.05", parse = TRUE) + # Add R^2 value from glm
  ggtitle("E") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,0.1,0.2,0.2), "cm")) +  # top, right, bottom, left
   theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
 )
ggsave(p5, filename="p5_ScatterplotBasic_Elev_by_MCH_2021-11-24.pdf", height=7, width=9, units="in")

summary(mch.piecewise2)
# Adjusted R-squared from piecewise model:  0.05 

#####

# PANEL F: Scatterplot: mchc v. elev
(p6 <- ggplot(final, aes(x=elev, y=mchc, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final, size=1.1, alpha=0.8) +
  #scale_colour_viridis(discrete=TRUE) + 
 #  facet_wrap(~Clade) +
  scale_color_manual(values = hum_colors) +
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=above.mchc) + # Segment 1  
  geom_smooth(method = "lm", formula = y ~ x, colour="black", size=1.0, data=below.mchc) + # Segment 2   
  theme_classic() + 
  labs(x="", y="MCHC (g/dl)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  ggtitle("F") + # Assign panel number/header; this will be (a) because first in series of 3
  annotate("text", x=4000, y=40, size=3, label = "italic(R) ^ 2==0.02", parse = TRUE) + # Add R^2 value from glm
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.0,right=5,0.2,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
 )
ggsave(p6, filename="p6_ScatterplotBasic_Elev_by_MCHC_2021-11-24.pdf", height=7, width=9, units="in")

summary(mchc.piecewise2)
# Adjusted R-squared from piecewise model:  0.02 

#####

# BOX PLOTS 

# Box plot: Elev cat vs. Hb (above/below differences)
(box.hb <- ggplot(final.hb.sub, aes(x=elev_cat, y=hb)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.3, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("< 2,972 m", ">= 2,972 m")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs 
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=2.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        # scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold1")) + # orange, blue 
        labs(y="[Hb] (g/dl)", # Hella confusing formatting for x and y axis labels 
             x="") +
        ggtitle("G") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
)
ggsave(box.hb, filename="AboveBelow_Boxplot_ElevCat-vs-Hb_2021_11-24.pdf", height=2, width=2, units="in")

# Box plot: Elev cat vs. hct (above/below differences)
(box.hct <- ggplot(final.hct.sub, aes(x=elev_cat, y=hct)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.3, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("< 1,461 m", ">= 1,461 m")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs 
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=2.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        # scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold1")) + # orange, blue 
        labs(y="Hct (%)", # Hella confusing formatting for x and y axis labels 
             x="") +
        ggtitle("H") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
)
ggsave(box.hct, filename="AboveBelow_Boxplot_ElevCat-vs-hct_2021_11-24.pdf", height=2, width=2, units="in")



# Box plot: Elev cat vs. trbc (above/below differences)
(box.trbc <- ggplot(final.trbc.sub, aes(x=elev_cat, y=trbc)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.3, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("< 3,520 m", ">= 3,520 m")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs 
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=2.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        # scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold1")) + # orange, blue 
        labs(y=expression(TRBC~"("~RBC~x~10^{6}~"/"~mm^{3}~")"), # Hella confusing formatting for x and y axis labels 
             x="") +
        ggtitle("I") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=8))
)
ggsave(box.trbc, filename="AboveBelow_Boxplot_ElevCat-vs-trbc_2021_11-24.pdf", height=2, width=2, units="in")

# Box plot: Elev cat vs. mcv (above/below differences)
(box.mcv <- ggplot(final.mcv.sub, aes(x=elev_cat, y=mcv)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.3, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("< 2,031 m", ">= 2,031 m")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs 
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=2.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        # scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold1")) + # orange, blue 
        labs(y="MCV (fl)", # Hella confusing formatting for x and y axis labels 
             x="Elevation (m)") +
        ggtitle("J") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
)
ggsave(box.mcv, filename="AboveBelow_Boxplot_ElevCat-vs-mcv_2021_11-24.pdf", height=2, width=2, units="in")


# Box plot: Elev cat vs. mch (above/below differences)
(box.mch <- ggplot(final.mch.sub, aes(x=elev_cat, y=mch)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.3, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("< 3,144 m", ">= 3,144 m")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs 
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=2.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        # scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold1")) + # orange, blue 
        labs(y="MCH (pg)", # Hella confusing formatting for x and y axis labels 
             x="Elevation (m)") +
        ggtitle("K") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
)
ggsave(box.mch, filename="AboveBelow_Boxplot_ElevCat-vs-mch_2021_11-24.pdf", height=2, width=2, units="in")


# Box plot: Elev cat vs. mchc (above/below differences)
(box.mchc <- ggplot(final.mchc.sub, aes(x=elev_cat, y=mchc)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(elev_cat)), outlier.size=0.3, alpha=0.7) +
        #geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        geom_signif(comparisons = list(c("< 3,936 m", ">= 3,936 m")), 
                    map_signif_level=TRUE, tip_length=0) + # Note sig diffs 
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=2.5) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
        # scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        scale_fill_manual(values = c("cornsilk4", "gold1")) + # orange, blue 
        labs(y="MCHC (g/dl)", # Hella confusing formatting for x and y axis labels 
             x="Elevation (m)") +
        ggtitle("L") + # Assign panel number/header
        theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
          plot.title = element_text(face="bold")) + # This makes panel header bold 
        # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=8), axis.title=element_text(size=10))
)
ggsave(box.mchc, filename="AboveBelow_Boxplot_ElevCat-vs-mchc_2021_11-24.pdf", height=2, width=2, units="in")

####

# COMBINE ALL PLOTS TO MAKE 12-PANEL FIGURE: 
# library(patchwork)
# FigS6BreaksnInsetBoxes <- (
#           p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(widths = c(0.8,0.8,0.8))
#           + inset_element(box.hb / box.hct / box.trbc / box.mcv / box.mch / box.mchc, 
#                           left = 1.0, # Increased #s move left edge to the right; decreased to the left
#                           bottom = -0.19, 
#                           right = 1.72, # Increased #s move right edge right; decreased to the left
#                           top = 2.38) 
#           )  
# print(FigS6BreaksnInsetBoxes)
# ggsave(FigS6BreaksnInsetBoxes, filename = "FigS6BreaksnInsetBoxes_2021-11-29.pdf", bg="transparent", height=9, width=12, units="in")
# A little janky, but I set up the species mean panel as an inset in order to get the formatting to work
# SYNTAX VERY TRICKY! 
# Only big problem with this is that I can't figure out how to get guides="collect" to put common A-F legend on RIGHT
# side of my giant inset - legend wants to appear over the top of the inset. 

# Since I need to export this to Illustrator to add the legend anyways, seems easier to add in a bottom panel of box plots there, rather than endlessly fiddle with aesthetics affected by resizing and adding in more panels. dd


# WITH BOX PLOTS
FigS5_BreaksnBox <- (
          (p1 + p2 + p3 + p4 + p5 + p6) /
          (box.hb + box.hct + box.trbc + box.mcv + box.mch + box.mchc) 
          + plot_layout(guides = "collect")
          )  
print(FigS5_BreaksnBox)
ggsave(FigS5_BreaksnBox, filename = "FigS5_BreaksnBox_BreakpointPlotsAndBoxPlots_2022-01-22.pdf", bg="transparent", height=9, width=9, units="in")
# Note that this file used to be FigS6, so it's saved under this name on hard drive

# plot layout guides collect was being super janky, so I'm writing this out with a lot of white space on the side and 
# will manually add legend in Word or Illustrator. 
# I ended up needing to open in Illustrator anyways since significance stars got cut off; needed to unveil from behind 
# the clipping masks
```



# FIGURE S6: in `HumBlood_CeNS.Rmd` script, contmap showing CeNS across the phylogeny


# FIGURE S7: "SIGNIFICANT" CLADE*ELEV INTERACTION FOR [HB]
```{R}
# Load in individual-level models for plots 
load("Rdata_files/hb.m4.ce_TopModel_CladeElevInteraction.RData") 

# Note: if you use font_family "Arial" ggsave will have trouble writing this to a pdf because it makes the plot w/ Arial 
# Narrow and pdf is slightly different; either use family "Helvetica" (or any other), or follow instructions here: 
# https://github.com/thomasp85/ggraph/issues/152

# Hemoglobin model posterior probability plot 
color_scheme_set("viridisD")
FigureS7 <- mcmc_plot(hb.m4.ce, variable=c("b_elev.z", "b_elev.pos.z","b_mass.z","b_temp.z","b_precip.z",
                        "b_genotypeGlyMSer", "b_genotypeSerMSer", "b_intravar.mass.z", "b_intravar.temp.z", 
                        "b_intravar.precip.z", "b_elev.z:CladeBrilliant", "b_elev.z:CladeCoquette", 
                        "b_elev.z:CladeEmerald", "b_elev.z:CladeHermit", "b_elev.z:CladeMango", "b_elev.z:CladePatagona", 
                        "b_elev.z:CladeTopaz"),
                         prob_outer=0.95, # 95% outer CI 
                         prob=0.89, # 50% inner CI 
                         point_est="mean") + # mean point est; default is median 
labs(x="Effect on [Hb]", title = "") + 
scale_y_discrete(labels=c("Elevation","Elevation Position","Mass", "Temp", "Precip", "Genotype (Gly-Ser)", "Genotype (Ser-Ser)", "Intra Var - Mass", "Intra Var - Temp", "Intra Var - Precip", "Elev*Brilliant", "Elev*Coquette", "Elev*Emerald", "Elev*Hermit", "Elev*Mango", "Elev*Patagona", "Elev*Topaz"))   
FigureS7
ggsave(FigureS7, filename = "FigureS7_HbModel_CladeVElevInteraction_PosteriorProbs_2022-01-22.pdf", bg="transparent", height=7, width=9, units="in")
summary(hb.m4.ce)
```



# FIGURE S8: CENS DIAGNOSTIC PLOT - correlations between R^2, TRBC estimates, and MCV estimates
Read in data if necessary: 
```{r}
cens.ss.mean <- read.csv("cens.ss.mean_ForCeNSmodeling_06-14-21.csv", stringsAsFactors = FALSE)
cens.ss.mean <- cens.ss.mean[ , !(colnames(cens.ss.mean) %in% c("X"))] # drop weird X column 1 if reading in from read.csv
cens.unst.trbc <- read.csv("cens.unst.trbc_Values_06-14-21.csv")
cens.unst.mcv <- read.csv("cens.unst.mcv_Values_06-14-21.csv")
```

```{r}
# Firstly, Let's assess possible correlation between the CeNS index and better fitting models - do better or worse fitting # models consistently have certain CeNS values? 

# Scatterplot: CeNS Index v. R^2 
(s8.1 <- ggplot(cens.ss.mean, aes(x=cens.R2, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=2.4, alpha=1.0) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
 # geom_smooth(method = "lm", colour="black", size=1.0) + # linear 
  theme_classic() + 
  labs(x=expression(CeNS~Model~italic(R)^{2}), y="CeNS Index (0-1)") + # CeNS index = (TRBC/MCV+TRBC)
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=0.8, y=0.70, label = "italic(R) ^ 2 == 0.06", parse = TRUE) + # Add R^2 value from lmer model
  annotate("text", x=0.8, y=0.68, label = "italic(r) == -0.23", parse = TRUE) + # Add r correlation coefficient
  ggtitle("A") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(s8.1, filename="CensIndex_Versus_R^2.pdf", height=7, width=9, units="in")

# Correlation matrix 
cor(cens.ss.mean[,c("cens.index", "cens.R2")])
# No correlation between R^2 values and the CeNS index: value = -0.23

# Quick linear model 
cens.R2.index.model <- glm(cens.index ~ cens.R2, data=cens.ss.mean, family=gaussian(), na.action=na.fail)
summary(cens.R2.index.model)
Anova(cens.R2.index.model, type=3)
lm_diag_plots(cens.R2.index.model)
(1 - (cens.R2.index.model$deviance/cens.R2.index.model$null.deviance)) # R^2 = 0.06

# Answer: No! Cens index is *not* correlated with R^2 values from models. This is good, suggests there isn't an 
# artifactual pattern of better or worse fitting models producing similar CeNS values. 


####

# Is cens.index correlated with STANDARDIZED TRBC estimates? 

# Scatterplot: CeNS Index v. unstandardized cens trbc estimate   
(s8.2 <- ggplot(cens.ss.mean, aes(x=cens.trbc, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=2.4, alpha=1.0) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
 # geom_smooth(method = "lm", colour="black", size=1.0) + # linear 
  theme_classic() + 
  labs(x="CeNS TRBC Estimate\n(Standardized)", y="") + # Remove CeNS index y-axis label for shared plot
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=3.2, y=0.70, label = "italic(R) ^ 2 == 0.07", parse = TRUE) + # Add R^2 value from lmer model
  annotate("text", x=3.2, y=0.66, label = "italic(r) == -0.27", parse = TRUE) + # Add r correlation coefficient
  ggtitle("B") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(s8.2, filename="CeNSIndex_Versus_StandardizedTRBCEstimate.pdf", height=7, width=9, units="in")
# Coeligena coeligena is the super weird outlier point 

# Correlation matrix 
cor(cens.ss.mean[,c("cens.index", "cens.trbc")])
# No correlation between R^2 values and the CeNS index: value = -0.27

# Quick linear model 
cens.index.sttrbc.model <- glm(cens.index ~ cens.trbc, data=cens.ss.mean, family=gaussian(), na.action=na.fail)
summary(cens.index.sttrbc.model)
Anova(cens.index.sttrbc.model, type=3)
lm_diag_plots(cens.index.sttrbc.model)
(1 - (cens.index.sttrbc.model$deviance/cens.index.sttrbc.model$null.deviance)) # R^2 = 0.07


# Is cens.index correlated with UNSTNDARDIZED TRBC estimates? 

# Scatterplot: CeNS Index v. unstandardized cens trbc estimate   
(s8.3 <- ggplot(cens.ss.mean, aes(x=trbc.unst, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=2.4, alpha=1.0) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
 # geom_smooth(method = "lm", colour="black", size=1.0) + # linear 
  theme_classic() + 
  labs(x="CeNS TRBC Estimate\n(Unstandardized)", y="") + # Remove CeNS index y-axis label for shared plot
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=3.2, y=0.70, label = "italic(R) ^ 2 == 0.07", parse = TRUE) + # Add R^2 value from lmer model
  annotate("text", x=3.2, y=0.66, label = "italic(r) == -0.26", parse = TRUE) + # Add r correlation coefficient
  ggtitle("C") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(s8.3, filename="CeNSIndex_Versus_UnstandardizedTRBCEstimate.pdf", height=7, width=9, units="in")
# Coeligena coeligena is the super weird outlier point 

# Correlation matrix 
cor(cens.ss.mean[,c("cens.index", "cens.unst.trbc")])
# No correlation between R^2 values and the CeNS index: value = -0.26

# Quick linear model 
cens.index.unsttrbc.model <- glm(cens.index ~ cens.unst.trbc, data=cens.ss.mean, family=gaussian(), na.action=na.fail)
summary(cens.index.unsttrbc.model)
Anova(cens.index.unsttrbc.model, type=3)
lm_diag_plots(cens.index.unsttrbc.model)
(1 - (cens.index.unsttrbc.model$deviance/cens.index.unsttrbc.model$null.deviance)) # R^2 = 0.07

# Answer: No! Cens index is *not* correlated with unstandardized TRBC values from models. 


##### 

# Is cens.index correlated with STANDARDIZED MCV estimates? 

# Scatterplot: CeNS Index v. unstandardized cens MCV estimate   
(s8.4 <- ggplot(cens.ss.mean, aes(x=cens.mcv, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=2.4, alpha=1.0) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
  geom_smooth(method = "lm", colour="black", size=1.0) + # linear 
  theme_classic() + 
  labs(x="CeNS MCV Model Estimate\n(Standardized)", y="") + # Remove CeNS index y-axis label for shared plot
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=3.2, y=0.70, label = "italic(R) ^ 2 == 0.51", parse = TRUE) + # Add R^2 value from lmer model
  annotate("text", x=3.2, y=0.66, label = "italic(r) == -0.71", parse = TRUE) + # Add r correlation coefficient
  ggtitle("D") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(s8.4, filename="CeNSIndex_Versus_MCVEstimateStandardized.pdf", height=7, width=9, units="in")


# Correlation matrix 
cor(cens.ss.mean[,c("cens.index", "cens.mcv")])
# Fairly strong negative correlation between CeNS index and unstandardized MCV values: value = -0.7

# Quick linear model 
cens.index.stmcv.model <- glm(cens.index ~ cens.mcv, data=cens.ss.mean, family=gaussian(), na.action=na.fail)
summary(cens.index.stmcv.model)
Anova(cens.index.stmcv.model, type=3) # yes, unstandardized MCV is a significant predictor of CeNS index
lm_diag_plots(cens.index.stmcv.model)
(1 - (cens.index.stmcv.model$deviance/cens.index.stmcv.model$null.deviance)) # R^2 = 0.51


# Is cens.index correlated with UNSTANDARDIZED MCV estimates? 

# Scatterplot: CeNS Index v. unstandardized cens MCV estimate   
(s8.5 <- ggplot(cens.ss.mean, aes(x=mcv.unst, y=cens.index)) + # Add shape=Clade here to get shapes to work properly
  geom_point(size=2.4, alpha=1.0) +
 # geom_text(aes(label=species),hjust=0, vjust=0) + # UNCOMMENT TO ADD SPECIES LABELS TO POINTS
  geom_smooth(method = "lm", colour="black", size=1.0) + # linear 
  theme_classic() + 
  labs(x="CeNS MCV Model Estimate\n(Unstandardized)", y="") + # Remove CeNS index y-axis label for shared plot
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "right") + 
  annotate("text", x=0.25, y=0.70, label = "italic(R) ^ 2 == 0.45", parse = TRUE) + # Add R^2 value from lmer model
    annotate("text", x=0.25, y=0.66, label = "italic(r) == -0.67", parse = TRUE) + # Add r correlation coefficient
  ggtitle("E") + # Assign panel number/header; this will be (a) because first in series of 3
  theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
         plot.title = element_text(face="bold")) + # This makes panel header bold 
       # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(plot.margin = unit(c(0.2,0.1,0.0,0.2), "cm")) +  # top, right, bottom, left
  theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=10), axis.title=element_text(size=12))
 )
ggsave(s8.5, filename="CeNSIndex_Versus_MCVEstimateUnstandardized.pdf", height=7, width=9, units="in")


# Correlation matrix 
cor(cens.ss.mean[,c("cens.index", "cens.unst.mcv")])
# Fairly strong negative correlation between CeNS index and unstandardized MCV values: value = -0.67

# Quick linear model 
cens.index.unstmcv.model <- glm(cens.index ~ cens.unst.mcv, data=cens.ss.mean, family=gaussian(), na.action=na.fail)
summary(cens.index.unstmcv.model)
Anova(cens.index.unstmcv.model, type=3) # yes, unstandardized MCV is a significant predictor of CeNS index
lm_diag_plots(cens.index.unstmcv.model)
(1 - (cens.index.unstmcv.model$deviance/cens.index.unstmcv.model$null.deviance)) # R^2 = 0.45

# Answer: Yes, cens.index is strongly negatively correlated with unstandardized MCV estimates 
# The higher the unstandardized MCV coefficient, the lower the CeNS index (aka, greater contribution of)


#### FINAL PLOT #####

library(patchwork)
# Alternative layout option = BETTER, USE THIS ONE:
FigureS8_5panel_CeNSDiagnostics <- (s8.1 | s8.2 + s8.3 + s8.4 + s8.5 +
                                    plot_layout(guides = "collect"))
plot(FigureS8_5panel_CeNSDiagnostics)
ggsave(FigureS8_5panel_CeNSDiagnostics, filename = "FigureS8_5panel_CeNSDiagnostics_2022-01-22.pdf", bg="transparent", height=6.8, width=12, units="in")
```



# FIGURE S9: Boxplot of Hb genotype and CeNS index
Figure S9 in paper. The version of this plot in the HumBlood_CeNS script is for diagnostics. 
```{r}
# HB GENOTYPE DOES NOT AFFECT CENS INDEX
(FigureS9 <- ggplot(cens.ss.mean, aes(x=genotype, y=cens.index)) +
       # facet_wrap(~sex, nrow=2) + 
        geom_boxplot(aes(fill = factor(genotype)), outlier.size=2.0, alpha=0.7) +
        #geom_signif(comparisons = list(c("Gly-Gly", "Gly-Ser", "Ser-Ser")), map_signif_level=TRUE) + # Note sig diffs    
        geom_jitter(position = position_jitter(width = 0.01), alpha = 0.5, size = 2.0) +
        stat_summary(fun=mean, colour="firebrick1", geom="point", shape=18, size=6) + # Red diamonds at mean 
        theme_classic() +   # theme_classic() removes all gridlines; theme_bw() retains subtle nice gridlines           
        theme(panel.grid.minor = element_blank()) +
        #scale_fill_brewer(palette="RdYlGn") + 
         scale_fill_viridis(option="viridis", discrete=TRUE) + # for Viridis palettes
        labs(y="CeNS Index", # Hella confusing formatting for x and y axis labels 
             x="Hb Genotype") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        theme(legend.position = "none") +
        theme(axis.text.y=element_text(size=10), axis.text.x=element_text(size=12), axis.title=element_text(size=14))
)
ggsave(FigureS9, filename="FigureS9_CensIndex_vs_Genotype_2022_01-22.pdf", height=5, width=7, units="in")
```



# end of paper plots


-----


# Plot for Chris Brown Bag: 
Not in paper! This was for a spring 2021 (I think) UNM Bio Brown Bag talk.
If you want to plot this, read in final.hct from modeling file
```{r}
# Scatterplot: Hct v. elev
(bbag_plot <- ggplot(final.hct, aes(x=elev_position, y=hct, colour=Clade)) + # Add shape=Clade here to get shapes to work properly
  geom_point(data=final.hct, size=1.5, alpha=0.8) +
  scale_colour_viridis(discrete=TRUE) + 
  #scale_color_manual(values = peru_chimborazo) +
 # geom_jitter(width=1.5, height=1.5, alpha=0.8, size=1.5) + # jitter to fill plot space 
  geom_smooth(method = "lm", colour="firebrick3", size=1.4) + # line used to be dark turquoise 
  theme_classic() + 
  labs(x="Elevation Position", y="Hct (%)") +
       # theme(axis.text.x = element_text(hjust = 1)) +
  theme(legend.position = "none") +
  # ggtitle("(b)") + # Assign panel number/header; this will be (a) because first in series of 3
  # theme(plot.title.position = "plot", # parameter "plot" specifies that you want "title" flush with y-axis
  #        plot.title = element_text(face="bold")) + # This makes panel header bold 
  #      # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
  theme(axis.text.y=element_text(size=12), axis.text.x=element_text(size=12), axis.title=element_text(size=16))
 )
ggsave(bbag_plot, filename="bbag_plot_ScatterplotBasic_ElevPosition_by_Hct.pdf", height=3, width=4, units="in")
```



-----

Helpful tutorials: 
Helpful tutorials on piecewise or segmented regression: 

RBloggers (describes Crawley 2007 iterative method but sucks because code chunks and plots are missing): https://www.r-bloggers.com/2012/08/r-for-ecologists-putting-together-a-piecewise-regression/

This describes change points and structual breaks; slightly different method: https://kevin-kotze.gitlab.io/tsm/ts-2-tut/

Segmented regression on world running times (not helpful for code, but explanations and some visuals are a good starting point): https://blog.revolutionanalytics.com/2015/12/using-segmented-regression-to-analyse-world-record-running-times.html

RPubs tutorial...not super helpful: https://rpubs.com/MarkusLoew/12164

Good post on adding significance bars to plots: https://stackoverflow.com/questions/17084566/put-stars-on-ggplot-barplots-and-boxplots-to-indicate-the-level-of-significanc

Always include lower-order polynomial when modeling a quadratic: 
https://stats.stackexchange.com/questions/28730/does-it-make-sense-to-add-a-quadratic-term-but-not-the-linear-term-to-a-model

---


# Print environment for reproducibility
```{r}
sessionInfo() # List of packages and versions in use 
```

###########

## END 
