---
title: "Hummingbird blood data from an interspecific comparative standpoint"
author: "Jessie Williamson"
date: "7/2/2020"
output:
  pdf_document: default
  html_document: default
---

#### 

This script includes: processing raw data, evaluating and eliminating outliers, creating new variables, combining data w/ Stotz, making individual elevational range adjustments, processing spatial data and generating sampling map, BioClim processing, reading out of final data file for modeling, plotting, etc.

####

# QUESTIONS 

**1) How do blood parameters vary with abiotic aspects of the environment?** 
Predictors: 
1) elevation 
2) elevation position (position relative to the width of the elevational range; scale from 0-1)
3) temperature index (PC1 of 19 BioClim variables)
4) aridity index (PC1 of 19 BioClim variables)
5) latitude

**2) How to blood parameters vary with aspects of species biology?** 
Predictors: 
5) body mass
6) hemoglobin genotype (beta13-beta83 genotype) 

**3) Are patterns in blood phylogenetically constrained?**

**4) What relationships exist among blood parameters?**
6 total: [Hb], Hct, TRBC, MCV, MCH, MCHC


# DATASET 
Dataset includes blood data for 77 species of hummingbirds from Chile and Peru. Sampling was conducted from 2006-2020 by students, affiliates, researchers, and professors at the Museum of Southwestern Biology at the University of New Mexico. All data are linked to vouchered specimens*.
*Note: Because some tissues and specimens haven't been exported from Peru, and because Chile collection isn't yet catalogued, not all are archived in Arctos. 

# APPROACH 
We will build 6 sets of phylogenetic mixed models in brms (x 10 models per set)

Response variables: 
1) Hb
2) Hct
3) TRBC
4) MCV
5) MCH
6) MCHC

Predictor set for each set of models: 
1) elevation 
2) elevation position (position relative to the width of the elevational range; scale from 0-1)
3) temperature index
4) precip index
5) latitude 
6) mass
7) Hb genotype    
8) intraspecific variation: elevation
9) intraspecific variation: elevation position
10) intraspecific variation: latitude
11) intraspecific variation: mass
12) intraspecific vatiation: precip
13) intraspecific variation: temp 
+ (1|species)
+ (1|phylogeny)

For each response, we will include: 
1) Null model (intercept-only)
2) Random effects-only model (intercept + phylogenetic random effect + species random effect)
3) Full model with no random effects (just predictors) 
4) Full model with just phylo random effect
5) Full model with just species random effect 
6) Full model with phylo and species random effects 
7) Reduced model with no random effects (just predictors whose CIs do NOT overlap zero)
8) Reduced model with just phylo random effect 
9) Reduced model with just sepcies random effect 
10) Reduced model with phylogenetic and species random effects 

We'll use LOOIC and WAIC to compare model sets.


```{R, echo=FALSE}
# I set some GLOBAL R chunk options here.
#   (to hide this message add "echo=FALSE" to the code chunk options)

knitr::opts_chunk$set(comment = NA, message = FALSE, warning = FALSE, width = 100)
knitr::opts_chunk$set(fig.align = "center", fig.height = 4, fig.width = 6)

#knitr::opts_chunk$set(cache = TRUE, autodep=TRUE)
knitr::opts_chunk$set(cache = TRUE, autodep=TRUE)
```


# Load packages
```{R}
library(reshape)
library(reshape2)
library(plyr)
library(dplyr)
library(car)
library(GGally)
library(Hmisc)
library(gridExtra)
library(stats)
library(gplots)
library(ggplot2)
library(stats4) # Forces knitr to work when it's being wonky
library(PMCMR) #Allows Kruskal-Wallis post-hocs
library(effects)
library(gridExtra)
library(lattice)
library(survival)
library(fmsb)
library(faraway)
library(ape)

# Mapping 
library(raster)
library(sp)
library(rgdal)
library(RStoolbox)
library(prettymapr)
library(viridis)
library(rasterVis)

# Modeling packages 
library(nlme)
library(lme4)
library(AICcmodavg)
library(MuMIn)
library(glmulti)
library(lsmeans)
library(rsq) # get r-squared values from GLM
library(r2glmm) # for R^2 values from lmer() and glmer()
library(multcompView) # related to multiple comparisons?
library(jtools) # interaction plots 
library(interactions) # interaction plots 
library(broom)
library(stargazer) # model output tables
library(ggeffects) # for estimating model predictions from mixed effects models
library(brms)
library(MCMCglmm)
library(rstan)
library(bayesplot)
library(ggrepel)

# Phylo packages 
library(phytools)
```


---

# Clear workspace and set WD
```{R}
rm(list=ls(all=TRUE)) # clear workspace 
setwd("/Users/Jessie/Dropbox (MSBbirds)/Rdirectory/ComparativeHummingbirdBlood")
```



# load in data 
```{R}
# Load functions 
source("ada_functions.R") # Erik's ADA functions for clean & collated lm diag plots

# All Patagona blood data
# now combined into one working master blood file 
# patagona <- read.csv("Patagona_Blood_Morpho_Tissue_MASTER_working_June2020.csv", na.strings = c("", "NA"))

# Note: all variables read in with "X0_" (just "0_" in Excel) are variables that will be dropped once you run code for 
# blood calculations below. This prefix indicates DROP ME, NOT PART OF FINAL ANYTHING.

# All hummingbird blood data
blood <- read.csv("All_hummingbird_blood_data_forJLWcomparison_07-02-20.csv")

# Assign row IDs 
#patagona$rowID <- 1:nrow(patagona)
blood$rowID <- 1:nrow(blood)
blood$species <- as.factor(blood$species)

# Cleaned Patagona data (written in Patagona_blood_analysis_2020.Rmd file for Patagona analysis)
pdat <- read.csv("Patagona_subset_cleaned_for_merge_with_AllHummingbirdComparative_10-27-20.csv")  
# Make sure this is most updated version; 10-27-20 version updated from 07-08-20 w/ museum cat num added 
pdat <- pdat[ , !(colnames(pdat) %in% c("X"))] # drop weird X column 1 if reading in from read.csv

# Stotz elev data 
stotz <- read.csv("StotzEtAl_ElevRangeData.csv")

# Wing loading data from MSB records (59 spp only)
wload <- read.csv("Hummingbird_WingLoading_Summary_59spp_FromChris.csv") 

# Phylogeny pruned from McGuire and adjusted (see script HumBlood_Phylogeny.Rmd)
tree <- read.tree("McGuirePruned_AllHummingbirdComparativeTree_FINAL.tre") # a list 
tree$tip.label # Check 77 tips
plotTree(tree, ftype="i") # Check this pre-final tree w/ adjusted branch lengths 
```


# First trim of dataset 
NOTE: This is a super quick and dirty first pass. You WILL need to revise this. Goal is to remove a bunch of things you won't need so this is a little easier to manipulate. 
```{R}
# Drop clunky notes variables you don't want for analysis
blood <- subset(blood, select = -c(pe6_no,
                                   order, # since all same order 
                                   family, # since all same family 
                                 #  cat_owner,
                                   cat_owner_initials,
                                   pers_cat_no,
                                   preparator,
                                   gonads,
                                   skull,
                                   bursa, # consider keeping this and using it to make an age variable
                                   molt,
                                   fat,
                                   prep_catalog_notes,
                                   soft_part_colors,
                                   voucher_specimen_disposition,
                                   skin_0.1,
                                   skel_0.1,
                                   skel_0.p.c,
                                   spreadwing_0.1,
                                   stomach_content,
                                   coll_method,
                                   collector1,
                                   collector2,
                                   mass_prep_cat, # Take out all these masses cause you'll use mass_for_analyses only
                                   mass_blood_cat,
                                   mass_at_death,
                                   mass_at_capture,
                                   mass_remarks,
                                   time_of_capture1_needs_merging,
                                   individual,
                                   higher_geog,
                                   locality_description,
                                   hct_saved_0.1,
                                   hb_corr, # Remove this because you'll calculate below 
                                   uno_corr_factor, # shouldn't need this
                                   uno_corr_factor2, # shouldn't need this either 
                                   blood_smear_0.1,
                                   smear_temp_0.1,
                                   smearID,
                                   smear_notes,
                                   time_in_captivity_hrs_note_if_fed,
                                   time_of_bleeding,
                                   blood_notes,
                                   RBC_notes,
                                   RBC2_notes
                                        ) )
```



# Calculate blood metrics and make variables needed for analysis 
```{r}
library(dplyr)

# Convert integers to numeric for the following code chunk to work properly:
str(blood)
blood$hb <- as.numeric(blood$hb) # NAS intro-ed after coercion is because NK161316 has value of "infer" for Hb
blood$mass_for_analyses <- as.numeric(blood$mass_for_analyses)
blood$hct1_column <- as.numeric(blood$hct1_column)
blood$hct2_column <- as.numeric(blood$hct2_column)
blood$hct2_top <- as.numeric(blood$hct2_top)
blood$hct2_bottom <- as.numeric(blood$hct2_bottom)
blood$hct_best <- as.numeric(blood$hct_best)
# Lots of NAs introduced by coercion is because there are some wonky values in cells 

# See Patagona script for code if you need to merge masses and/or calculate flight muscle or organ mass variables 

# MASS
# HEMOGLOBIN - corrected
names(blood)[names(blood) == "mass_for_analyses"] <- "mass_final" # Change name for consistent naming
# NOTE: mass "converted from dwt" under "mass_for_analyses" means converted from pennyweight (dwt) 

# HEMOGLOBIN - corrected
blood <- blood %>% mutate(hb_corr = (hb - 1)) # correct all Hb values 
names(blood)[names(blood) == "hb_corr"] <- "hb_final" # Change name for consistent naming

# HEMATOCRIT
blood <- 
  blood %>% mutate(hct1_middle_calculated = (hct1_top + hct1_bottom)/2, # Make Hct middle calculation
                      hct1_ratio = (hct1_middle_calculated/hct1_column), # Hct1 ratio
                      hct2_middle_calculated = (hct2_top + hct2_bottom)/2, # Make Hct 2 middle calculation
                      hct2_ratio = (hct2_middle_calculated/hct2_column), # Hct2 ratio
                      wtavg_hct = (hct1_ratio*(hct1_column/(hct1_column + hct2_column))) + (hct2_ratio*(hct2_column/(hct2_column + hct1_column))), # Weighted average of Hct measurements
                      hct_final = coalesce(hct1_ratio, wtavg_hct), # Take Hct1_ratio when only 1 measurement exists
                      hct_final_percent = (hct_final*100) # display HCT in percent; necessary for proper format
                      )
# Weighted average of Hct measurements is equivalent to Packed Cell Volume (PCVs %)
# Note that last line is important! In instances where you can't take weighted avg because only 1 Hct measurement, 
# use Hct1_ratio; else use wtavg_hct measurements for calculations 

              
# MEAN CELL VOLUME (MCV)
# Formula for MCV (fl) from Campbell & Ellis (2007): (HCT%/TRBC)*10
# This can also be calculated with the following: (hct_final/TRBC)*1000 (note that hct_final isn't in %; it's decimals)
# NOTE: Our data doesn't have the x5, x10000, and x200_dilution variables that my Patagona dataset does becuase no mean 
# MCV cell counts; just finalized RBCx10^6mm^3 counts
names(blood)[names(blood) == "RBC_best_estimate"] <- "TRBC" # Change name for use in formula below
blood <- blood %>% mutate(MCV_calculated = (hct_final_percent/TRBC*10)) # , # Mean Cell Volume 
names(blood)[names(blood) == "MCV_calculated"] <- "MCV_final" # Change name for use in formula below


# MCHC (mean cellular hemoglobin concentration)
# Formula for MCHC (gm/dl) from Campbell & ellis (2007): (Hb/HCT %)*100
blood <- blood %>% mutate(MCHC_calculated = (hb_final/hct_final_percent)*100) # Calculated MCHC
names(blood)[names(blood) == "MCHC_calculated"] <- "MCHC_final" # Change name 


# MCH (mean cell hemoglobin)
# Formula for MCH (pg) from Campbell & ellis (2007): (Hb/TRBC)*10
blood <- blood %>% mutate(MCH_calculated = (hb_final/TRBC*10)) # Calculated MCH
names(blood)[names(blood) == "MCH_calculated"] <- "MCH_final" # Change name

```


# convert lats and lons
```{r}
# Function to convert latitude from degrees decimal minutes to decimal degrees 
convert_lat <- function(dataframe){
  lat_degrees_S <- dataframe$lat_degrees_S 
  lat_mins <- dataframe$lat_mins 
  lat <- lat_degrees_S + (lat_mins/60)
  return(-lat) # ask to return negative to indicate southern hemisphere 
}

# Function to convert longitude from degrees decimal minutes to decimal degrees 
convert_lon <- function(dataframe){
  lon_degrees_W <- dataframe$lon_degrees_W
  lon_mins <- dataframe$lon_mins
  lon <- lon_degrees_W + (lon_mins/60)
  return(-lon) # ask to return negative to indicate southern hemisphere 
}

blood$lat_convert <- convert_lat(blood) # Convert all lat values and add to data frame 
blood$lon_convert <- convert_lon(blood) # Convert all lon values and add to data frame 

# Convert decimal lat and lon to negative for southern hemisphere (currently these are NOT negative)
blood$lat_dec_deg_S <- -blood$lat_dec_deg_S
blood$lon_dec_deg_W <- -blood$lon_dec_deg_W

# Merge raw decimal degree lats and lons w/ converted lats and lons to create one column for each
# Take raw over converted values when possible, as we assume these are more precise from GPS readings
blood <- blood %>% mutate(lat = coalesce(lat_dec_deg_S, lat_convert), 
                          lon = coalesce(lon_dec_deg_W, lon_convert))
# specifying lat_dec_deg_S & lon_dec_deg_W means these get inserted OVER lat_convert and lon_convert values

# Check positions of lat and lon with simple world map (all should be in southern hemisphere)
library(maptools)
data(wrld_simpl)
plot(wrld_simpl, xlim=c(-100,-10), ylim=c(-57,17), axes=TRUE, col="snow2") # plots gray world map 
box() # restore the box around the map
points(blood$lon, blood$lat, col='cyan3', pch=20, cex=1) # plot lat/lon points
# Looks good; all points appear in their proper places 

# Drop the one instance of no lat/lon, which will cause problems for spatial data below
blood <- blood %>% filter(!is.na(lat)) # keep only records with lat 
blood <- blood %>% filter(!is.na(lon)) # keep only records with lon
# Note: this is a Phaethornis malaris observation with only Hct, so we don't lose much by dropping it

#write.csv(blood, "blood.latlontest.csv")
```

Note about lat/lons: A bunch of lat/lons were missing from July 2007 Tres Lagunas, Lambayeque birds at 3,200 m. Chris said there was an error in processing all of these birds, which are all problematic cases, so we don't have coords for them. However, one bird, NK162815 (C. iris) from 12 July 2007 does have coords for this locality, elev, and date range. With Chris' permission, I used NK162815's lat/lon to fill in coordinates for all missing-coord NKS.  

I added coords from NK162815 to the following NKs: 162691, 162758, 162720, 162705, 162757, 162690, 162725, 162592, 162759, 162776, 162782, 162689, 162591, 162783, 162704



# Subset the data by just values you'll use in analysis 
Previous code chunks made the dataset bloated and w/ many values you don't need; after comparing w/ raw Excel values, subset to have something more manageable; also helps initiate removal of NAs, which is prudent for modeling. 
```{r}
# The order that you write these in here is the order they'll appear in once subsetted
bsub <- subset(blood, select = c(rowID,
                                 nk,
                                 msb_cat_no, 
                                 prep_num,
                                 cat_owner,
                                 species,
                                 sex,
                                 age_est, # See notes in Excel metadata about estimating age from bursa; check w/ Chris
                               # day,
                                 month,
                                 year,
                                 locality,
                                 elev,
                                 department,
                               # provincia,
                               # distrito,
                                 lat,
                                 lon,
                                 mass_final,
                                 hb_final,
                               # hct_final,
                                 hct_final_percent,
                                 TRBC,
                                 MCV_final,
                                 MCHC_final,
                                 MCH_final
                                  ))
# Keep in mind minimizing the NAs in the dataset by only selecting things you'll need 
# Why does row 1189 keep ending up blank??

# Rename cumbersome variables to more workable names: 
colnames(bsub)
# library(plyr) # dplyr masked by functions in reshape and reshape 2
# So annoying: for some reason rename() in plyr won't work, so need to add this line by line
names(bsub)[names(bsub) == "age_est"] <- "age"
names(bsub)[names(bsub) == "mass_final"] <- "mass"
names(bsub)[names(bsub) == "hb_final"] <- "hb"
names(bsub)[names(bsub) == "hct_final"] <- "hct"
names(bsub)[names(bsub) == "hct_final_percent"] <- "hct_percent"
names(bsub)[names(bsub) == "MCV_final"] <- "mcv"
names(bsub)[names(bsub) == "MCHC_final"] <- "mchc"
names(bsub)[names(bsub) == "MCH_final"] <- "mch"

# bsub <- rename(bsub, c("mass_final"="mass", 
#                        "hb_final"="hb", 
#                        "hct_final"="hct",
#                        "hct_final_percent"="hct_percent",
#                        "MCV_final"="mcv", 
#                        "MCHC_final"="mchc", 
#                        "MCH_final"="mch")) 

# Instantiate column for wing chord length (mm); necessary for combining with Patagona and Skandalis data 
bsub$wing <- as.numeric(paste(""))
```

Note: NK219789 oddly has no locality or collection info associated w/ it. It does have some blood data, but this may end up being dropped unless I can find other info (it doesn't come up in a quick Arctos search, so may require more digging).



# Prep for merge w/ Patagona data 
```{r}
# Drop Patagona data in pdat that doesn't have any blood information associated with it
pdat$blood_data_YN <- as.factor(pdat$blood_data_YN)
pdat <- pdat[-which(pdat$blood_data_YN =="no"),] 
table(pdat$blood_data_YN) # Confirm that there are zero instances of "no" (should be 166 of "yes")
#test <- subset(pdat, (!is.na(pdat[,2]))) # This is a way to remove all NA vals from only column 2, but not sure I want this

# Split Patagona' museum_cat_num (in MSB:BIRD:xxxxx format) into three variables so you can extract just catalog number 
pdat <- separate(pdat, museum_cat_num, into=c("museum", "filler", "msb_cat_no"), sep = "[^[:alnum:]]+",
  remove = TRUE, # If TRUE, remove original "Identifier" column you split 
  convert = FALSE,
  extra = "warn",
  fill = "warn"
)

# subset pdat to include all the same columns that bsub has 
# subset pdat to prepare for merge w/ bsub
pdat <- subset(pdat, select = c(rowID,
                                 nk,
                                 msb_cat_no, # Any NA values are from uncatalogued birds 
                                 pers_cat_no,
                                 cat_owner,
                                 species,
                                 sex,
                                 age, 
                               # day, # did I not write "day" out to this file from Patagona_blood_analysis_2020?
                                 month,
                                 year,
                                 locality,
                                 elev,
                                 department,
                                 lat,
                                 lon,
                                 mass,
                                 hb,
                                 hct, # This is in percent
                                 TRBC,
                                 mcv,
                                 mchc,
                                 mch,
                                 wing
                                  ))

# Rename column headers to match bsub
names(pdat)[names(pdat) == "pers_cat_no"] <- "prep_num"
names(pdat)[names(pdat) == "hct"] <- "hct_percent"

# Set levels of pdat, aka change name of Patagona_gigas to Patagona gigas
pdat$species <- as.factor(pdat$species) # Make sure this is coded as a factor 
levels(pdat$species) # Verify level 
levels(pdat$species) <- c("Patagona gigas")

# Now fill in cat_owner NAs with my name as the collector of blood data: 
pdat$cat_owner[is.na(pdat$cat_owner)] <- paste("Jessie L. Williamson") 

# Drop Patagona records from bsub - Patagona dataset is more stringently curated, so you want to replace Patagona
# data in bsub with Patagona data from pdat 
bsub <- bsub[-which(bsub$species =="Patagona gigas"),] 

# Compare column names of bsub and pdat to confirm order before merge 
colnames(pdat)
colnames(bsub)
```


# Merge all hummingbird blood data w/ curated Patagona data 
```{r}
# Merge the two datasets together: marry the all hummingbird dataset minus Patagona WITH the full Patagona dataset 
# This includes Chilean Patagona and Peru Patagona collected by JLW 
full <- rbind(bsub, pdat) # Should be 1317 total rows 
# Keep in mind that many 'wing' measurements are zero at this point; this will be fixed once we combine w/ Skandalis & impute

# Now change name of "hct_percent" to "hct" to standardize
names(full)[names(full) == "hct_percent"] <- "hct"
names(full)[names(full) == "TRBC"] <- "trbc"

# Calculate number of species in the dataset:
length(unique(full$species)) # number of unique species
# 77 total species 

# Write out "full" dataset for HumBlood_PreparatorContributions.rmd
#write.csv(full, "humblood_full.csv")
```


# Transform some variables
```{r}
# Run the above predictor scatterplot matrix, transform the data as needed, re-run scatterplot matrix
full$mass.log <- log10(full$mass) # Helped a bit, but still pretty off at tails, definitely not normal
hist(full$mass.log) 
qqPlot(full$mass.log) 

full$wing.log <- log10(full$wing) # Did almost nothing, though we aren't *very* interested in wing measurements
hist(full$wing.log)   
qqPlot(full$wing.log)  

# Mass looks like it might need to be transformed because Patagona throws things off...
# distributions of Hb, hct, trbc, and mchc look fantastic
# MCV and MCH might need transformation 
# This looks a LOT better after dropping unrealistic values (see code chunk below)
```


# Predictor scatterplot matrix
Predictors/possible predictors for models for models
```{R, fig.height = 6, fig.width = 6}
# full dataset
p <- ggpairs(subset(full, select = c(elev, mass, mass.log, hb, hct, trbc, mcv, mch, mchc))) 
print(p)

# Initial outlier/distribution assessment 

# Hb
hist(full$hb)
abline(v=c(14,25), col=c("red", "red"), lty=c(2,2), lwd=c(2, 3)) # <24 and > 25 outliers 
full[which(full$hb < 14),] # 3 outliers 
full[which(full$hb > 25),] # 0 outliers 

# Hct
hist(full$hct)
abline(v=c(35,75), col=c("red", "red"), lty=c(2,2), lwd=c(3, 3)) # <35 and > 75 outliers 
full[which(full$hct < 35),] # 2 outliers 
full[which(full$hct > 75),] # 3 outliers 

# trbc
hist(full$trbc)
abline(v=c(2,10), col=c("red", "red"), lty=c(2,2), lwd=c(3, 3)) # <2 and > 10 outliers 
full[which(full$trbc < 2),] # 4 outliers 
full[which(full$trbc > 10),] # 6 outliers
# Note: some crossover in TRBC and MCV outliers

# MCV
hist(full$mcv)
abline(v=c(50,175), col=c("red", "red"), lty=c(2,2), lwd=c(3, 3)) # <50 and > 200 outliers 
full[which(full$mcv < 50),] # 0 outliers 
full[which(full$mcv > 150),] # 10 outliers 
# Note: some crossover in TRBC and MCV outliers

# MCHC
hist(full$mchc)
abline(v=c(20,42), col=c("red", "red"), lty=c(2,2), lwd=c(3, 3)) # <20 and > 42 outliers 
full[which(full$mchc < 20),] # 0 outliers 
full[which(full$mchc > 42),] # 3 outliers

# MCH
hist(full$mch)
abline(v=c(18,50), col=c("red", "red"), lty=c(2,2), lwd=c(3, 3)) # <50 and > 200 outliers 
full[which(full$mch < 18),] # 0 outliers 
full[which(full$mch > 50),] # 5 outliers


# Generally, these look really good. I don't think any need to be transformed, but that could change. 
qqPlot(full$mass) # looks like shit (but would look ok w/out Patagona...)
qqPlot(full$elev) # not great but maybe hard to fix because of uneven sampling?
qqPlot(full$hb) # sexy AF 
qqPlot(full$hct) # damn near perfect
qqPlot(full$trbc) # some *slight* curvature tails, but within normal range 
qqPlot(full$mcv) # Upper limit of 150 MCV brings curvature at tails within liveable/normal-ish range = good 
qqPlot(full$mch) # off at the right tail, ok
qqPlot(full$mchc) # Still a little off at the tails, but very livable 

hist(full$mass)
hist(full$elev)
hist(full$hb) # So beautiful I want to cry 
hist(full$hct)
hist(full$mcv) # Right skew apparent
hist(full$mch) # right skew apparent
hist(full$mchc) # pretty good 

# Categorical predictors
p <- ggpairs(subset(full, select = c(sex, month, age))) # dpto, provincia, & distrito have too many categories
print(p)
```


# Drop outlier values for blood parameters
note: you CANNOT run na.omit on your subsetted data or you'll be left with only 18 observations. 
I think the best approach is to create separate subsets that remove outliers for each possible response variable of interest; that way, you'll retain as many values as possible in the subsetted dataset
```{r}
# FULL DATASET OUTLIERS 
# For [Hb]: drop any value <14 or >25
# DuBay and Witt range from 13.2-21.4 g/dl
full <- full[-which(full$hb < 14 | full$hb > 25),] # drop Hb <14 # -3 observations 

# for Hct: drop any value <35 or >75
# (we originally had this set to 72 but I bumped it to 75 after examining distribution)
# DuBay and Witt range from 46.3-58.1 for Hct
full <- full[-which(full$hct < 35 | full$hct > 75),] # 5 observations 
# Campbell and Ellis report "normal" range for birds 35-55
# <35% suggests anemia; >55% suggests dehyrdation or erythrocytocis (polycythemia; increase in RBCs)
# They note that PCV > 70% is usually indicative of chronic plumonary disease 

# For TRBC: drop any value <2 or >10
# I read somewhere online that bird 'normal' range is 2.5-4.5, with lower indicating anemia
# Full Campbell and Ellis range from Table B3 is 1.2-4.7
# We expect these values to be higher because of altitude effects
# Based on inspection of the distribution, we pick an informed range of 2-10
full <- full[-which(full$trbc < 2 | full$trbc > 10),] # -10 observations 


# Note: Campbell and Ellis 2007 say: 
# "assessment of abnormal values of MCV, MCH, and MCHC has not been properly evaluted in birds" - so outliers are up to us

# for MCV: drop any value <50 or >200
full <- full[-which(full$mcv < 50 | full$mcv > 150),] # >30 observations 
# Cambpell & Ellis note that 308 MCV is "large" and indicative of younger cells while 128 is smaller, suggests older cells
# I made the upper limit cut-off 150 instead of 200 after looking at distributions again; makes hist and qqPlots look better

# for MCHC: drop any value <20 or >42
# Campbell and Ellis values range from 20-38.5 (Table B3); Dubay & Witt values range from 
# DuBay and Witt range from 26.5-36.9 g/dl
full <- full[-which(full$mchc < 20 | full$mchc > 42),] # -3 observations 
# Leave for now; consider making this a more stringent 22-40

# for MCH: drop any value 
# Elarabany 2018 values for two duck species range from 24.6-31.1 PG; normal range for humans is 32-36 gm/dl 
full <- full[-which(full$mch < 18 | full$mch > 50),] # 4 observations 

# Total outlier points dropped: 68 of 1,317
# 1,249 total data points for analysis

# The plots/distributions above certainly look a LOT better after removing outliers
```

Note about outlier removal: Since dropping extreme values for e.g. MCV leads to a LOT of lost data, another thing I could do is JUST remove outliers for individual model datasets. I could then check the distributions of individual model data sets before working with them, and this would allow me to ditch 8 weird MCV values but keep the associated perfectly regular hemoglobin values, for example. 


# Now drop the levels of species for which you have < 3 observations 
NOTE: This is a relict code chunk - we decided that we don't want to drop any species. On 7/8/20 Chris said: 
"I say use all individuals, don’t have a threshold -- any such sample size threshold can be applied post hoc for specific purpose or question, but in general, individual-based analyses using all the data are the way to go (however, prudent elimination of errant/outlier datapoints is also essential)"
--> If you do decide to stick this back in, make sure to drop these AFTER removing outliers above. If not, you'll drop poorly sampled species, possibly remove outliers in the process, and then possibly have to drop additional species on the basis of sparse sampling (revealed after outlier analysis).
```{r}
# Get a sense for what species will need to be dropped 
# table(full$species)
# 
# # Identify species with fewer than 3 records 
# weakly.sampled <- c()
# for(i in full$species){
#    thresh <- full[full$species==i,]
#    if(nrow(thresh)>3){weakly.sampled[i] <- as.character(thresh$species[1])}
#  }
#  weakly.sampled <- as.vector(weakly.sampled)
 
# Now subset the data by species with fewer than 3 records 
# full <- full[full$species %in% weakly.sampled, ]

# Take a look at the unique number of species in the dataset and observations per species:
sp.count.full <- full %>% group_by(species) %>% summarise(count = length(nk))
# Should be 77 total species 
write.csv(sp.count.full, "HumBlood_species_list.csv")
```


# Assess age-based outliers
Blood data is anomalous in in nestlings and fledglings and normalizes quickly within a few months of fledging. 
Bursa lasts longer than weird juvenile blood characteristics, so presence/absence of bursa isn't 100% correlative with ages that would be important for blood analyses. Me, Chris, and Ethan decided that looking at a combination of characteristics is the best approach: 
- First, eliminate anomalous blood values 
- Then, analyze distribution of blood values. If these look normal and fit in with others, don't remove on the basis of having a bursa. But, if blood values look weird AND a bird has a "large" bursa, then remove value. 
- "Large" bursa is ~2x2 mm or greater for hummingbirds - CCW says this is a bird recently out of the nest.  
- This approach allows us to identify and possibly drop any errant values caused by age and retain all values of juveniles that fall within the "normal" adult range. 
- Also NOTE: Any non-vouchered (tracked) Patagona whose age was scored by plumage shouldn't be eliminated, as juvenile plumage lasts much longer than anomalous juvenile blood values. 
```{r}
# GGpairs comparison plot colored by age 
p <- ggpairs(full[ ,c("elev", "mass", "hb", "hct", "trbc", "mcv", "mch", "mchc", "age")], 
            mapping=ggplot2::aes(colour=full$age), lower=list(continuous="smooth"), 
            diag = list(continuous = "density")) #, upper = list(params = list(corSize = 6)) )
print(p) 

# pink = adult
# green = juvenile
# blue = unknown (because bursa was undetermined, not found, not recorded, etc.)

# Honestly, this looks pretty good. If age-based outliers remained, we'd expect lots of errant green points
# Instead, these fall in line with pink clusters and show similar spread in boxplots. 
```



-------

# Wrangling MSB wing loading data 
```{r}
library(tidyr)

# Need to merge with full dataset
# and copy down species mean values for all the species records 

# Now change column names 
names(wload)[names(wload) == "mass_g"] <- "wing.mass" # units = grams 
names(wload)[names(wload) == "wing_area_mm2"] <- "wing.area" # units = mm2
names(wload)[names(wload) == "wing_loading"] <- "wing.loading" 

# Merge full data w/ MSB wing loading data 
full.and.wload <- merge(full, wload, by.x="species", by.y="species", all.x=TRUE) 
# by.x corresponds to column to merge from full; by.y corresponds to column to merge from wload
# Because both data frames have same col name, by.x and by.y are the same; in Stotz example below, col names differ 
# all.x=TRUE argument means don't drop data if there isn't an exact match w/ wload 
# What I want: number of rows should be 1,249 (same as full), wing loading data should exist for wload species = SUCCESS! 
# GOOD NEWS: Merging in this way copies each species mean value down for all representatives of that species = GOOD. 

# Now look for correlations between wing loading and the 6 blood parameters of interest 

hist(full.and.wload$wing.loading.log) # Neither log nor sqrt transforming helps; best w/out transformation? 
qqPlot(full.and.wload$wing.loading.log)
#full.and.wload$wing.loading.log <- sqrt(full.and.wload$wing.loading)
p <- ggpairs(subset(full.and.wload, select = c(elev, mass.log, wing.loading, hb, hct, trbc, mcv, mch, mchc))) 
print(p)
# Remember that not all species have wing loading data, so this won't print for all 
```


# Wrangle Stotz
See Ethan Linck's code here for smooth Stotz merging: https://github.com/elinck/andean_range_limits/blob/master/scripts/01_blood_data_exploration.Rmd
```{R}
# Subset Stotz by just variables you want 
stotz.sub <- cbind.data.frame(stotz$GENUS, stotz$SPECIES, stotz$MIN, stotz$MAX, stotz$MIDPT.ELEV) 
colnames(stotz.sub) <- c("genus", "species", "elev_min", "elev_max", "elev_midpt") # assign column names 
stotz.sub$binomial <- paste0(stotz.sub$genus, " ", stotz.sub$species) # Combine genus and species cols into one  
full1 <- merge(full, stotz.sub, by.x = "species", by.y = "binomial", all.x=TRUE) # Merge full data w/ Stotz 
# by.x corresponds to column to merge from full; by.y corresponds to column to merge from stotz.sub
# The data frames have different col name, so by.x and by.y are the same; in Skandalis example above, col names are the same
# all.x=TRUE argument means don't drop data if there isn't an exact match w/ Stotz 
head(full1)

# How many unique species and records are in this dataset?  
length(unique(full1$species)) # number of unique species
nrow(full1) # number of unique records
# We should have all 77 species 
# Number of rows should match 'full'; aka be n=1283

# ID which species failed to pick up elev range data w/ the Stotz data merge:
full1[is.na(full1$elev_min),]$species %>% unique() %>% length() # 11 total

# Identify which ones: 
missing <- full1[is.na(full1$elev_min),]$species %>% unique(); missing
# All are related to typos and/or taxonomy changes  

# Eyeball Stotz data, fill in gaps tediously
# AKA: take names of 'missing' df, use these to manually fill in gaps by inputting Stotz Excel sheet values
# elev_min = col 28
# elev_max = col 29
# elev_midpt = col 30

full1[full1$species==missing[1],][28] <- 1300 #Aglaiocercus kingii # kingi in Stotz
full1[full1$species==missing[1],][29] <- 2600 #Aglaiocercus kingii
full1[full1$species==missing[1],][30] <- 1300 #Aglaiocercus kingii

full1[full1$species==missing[2],][28] <- 900 # Chaetocercus mulsant # Acestrura mulsant in Stotz
full1[full1$species==missing[2],][29] <- 2800 # Chaetocercus mulsant
full1[full1$species==missing[2],][30] <- 1900 # Chaetocercus mulsant

full1[full1$species==missing[3],][28] <- 1200 # Doryfera ludovicae # ludoviciae in Stotz
full1[full1$species==missing[3],][29] <- 2800 # Doryfera ludovicae
full1[full1$species==missing[3],][30] <- 1600 # Doryfera ludovicae

full1[full1$species==missing[4],][28] <- 2000 # Eriocnemis aline # alinae in Stotz
full1[full1$species==missing[4],][29] <- 2800 # Eriocnemis aline
full1[full1$species==missing[4],][30] <- 800  # Eriocnemis aline

full1[full1$species==missing[5],][28] <- 2250 # Eriocnemis vestita # vestitus in Stotz; but only 3 so dropped 
full1[full1$species==missing[5],][29] <- 3850 # Eriocnemis vestita
full1[full1$species==missing[5],][30] <- 1600 # Eriocnemis vestita

full1[full1$species==missing[6],][28] <- 0    # Glaucis hirsutus # hirsutas in Stotz
full1[full1$species==missing[6],][29] <- 1100 # Glaucis hirsutus
full1[full1$species==missing[6],][30] <- 1100 # Glaucis hirsutus

full1[full1$species==missing[7],][28] <- 2400 # Heliangelus micraster # spelling issue in Stotz
full1[full1$species==missing[7],][29] <- 2900 # Heliangelus micraster
full1[full1$species==missing[7],][30] <- 500  # Heliangelus micraster

full1[full1$species==missing[8],][28] <- 0    # Heliodoxa aurescens # Polyplancta aurescens in Stotz; only 3 so dropped
full1[full1$species==missing[8],][29] <- 1050 # Heliodoxa aurescens
full1[full1$species==missing[8],][30] <- 1050 # Heliodoxa aurescens

full1[full1$species==missing[9],][28] <- 3500 # Oreotrochilus estella # (estella) estella & (estella) chimborazo in Stotz
full1[full1$species==missing[9],][29] <- 4600 # Oreotrochilus estella
full1[full1$species==missing[9],][30] <- 1100 # Oreotrochilus estella

full1[full1$species==missing[10],][28] <- 0 # Phaethornis atrimentalis # Not in Stotz; adding Schulenberg 2007 ranges
full1[full1$species==missing[10],][29] <- 1100 # Phaethornis atrimentalis
full1[full1$species==missing[10],][30] <- 1100 # Phaethornis atrimentalis

full1[full1$species==missing[11],][28] <- 0 # Sephanoides sephaniodes # Sephanoides sephanOIdes in Stotz...of course.
full1[full1$species==missing[11],][29] <- 2000 # Sephanoides sephaniodes
full1[full1$species==missing[11],][30] <- 2000 # Sephanoides sephaniodes

# Now missing should be 0
```


# Species elevational range adjustments 
Stotz (Parker et al. 1996) is a good baseline for elevational ranges but is frequently inaccurate, and in some cases, outdated. Adjust species elevational ranges based on Schulenberg et al. 2007 Birds of Peru, museum records, other field guides (if necessary), and working knowledge. 
```{r}
# Note: this process could be simplified with a nice function like the one Ethan wrote for Linck et al. all-species
# blood comparison. *But* I committed to this clunky, "species-specific-personalized" method, so sticking with it for now.

# First, figure out which species MSB collection elevations exceed max and which exceed min
# Species collection elevs > elev_max 
max_exceed <- full1[full1$elev > full1$elev_max,]$species %>% unique(); max_exceed
# After adjusting, 0. 

# Species collection elevs > elev_max 
min_exceed <- full1[full1$elev < full1$elev_min,]$species %>% unique(); min_exceed
# Before adjusting, 4: Amazilia chionogaster, Haplophaedia aureliae, Lesbia victoriae, Oreotrochilus estella 
# After adjusting, 0

# [28] = elev_min
# [29] = elev_max
# [30] = elev_midpt

# Patagona gigas 
# Cite Williamson and Witt 2020, ENSM paper 
full1[full1$species=="Patagona gigas",][28] <- 0    # Min elev, determined from ENSM paper 
full1[full1$species=="Patagona gigas",][29] <- 4830 # Max elev, determined from ENSM paper

# Rhodopis vesper 
full1[full1$species=="Rhodopis vesper",][29] <- 3800 # Max elev, determined Schulenberg et al. 2007

# Haplophaedia aureliae
# Schulenberg et al. 2007 doesn't list an elev range (just found in under- and midstory of humid montane forest)
# Stotz min 1,400, MSB collection min is 1051; rounding down to 1050
full1[full1$species=="Haplophaedia aureliae",][28] <- 1050    # Min elev

# Polyonymus caroli
# Schulenberg et al. 2007 lists 2100-3400 m; we know they're common at 3,700-3,800 m in Lima 
# We've collected up to 4,062 m in Potaga, so I'll bump the max elev to 4,100
full1[full1$species=="Polyonymus caroli",][29] <- 4100 # Max elev
# Should lower limit be adjusted from 1,500 m (Stotz) to 2,100 m (Schulenberg)?

# Florisuga mellivora
# Schulenberg et al. 2007 says up to 1,200 m
full1[full1$species=="Florisuga mellivora",][29] <- 1200 # Max elev

# Threnetes leucurus
# Schulenberg et al. 2007 says up to 1200 m but locally to 1800 m; we've collected to ~1,700 m, so I'll go with 1,800 m
full1[full1$species=="Threnetes leucurus",][29] <- 1800 # Max elev

# Phaethornis malaris
# Schulenberg et al. 2007 says max 1,300 m
# MSB voucher shows collected elev of 1673 at Plataforma; adjusting accordingly
full1[full1$species=="Phaethornis malaris",][29] <- 1700 # Max elev

# Phaethornis hispidus
# Schulenberg et al. 2007 says max 850 m, locally to 1400 m
full1[full1$species=="Phaethornis hispidus",][29] <- 850 # Max elev

# Phaethornis philippii
# Schulenberg et al. 2007 says below 500 m
full1[full1$species=="Phaethornis philippii",][29] <- 500 # Max elev

# Doryfera ludovicae
# Schulenberg et al. 2007 1,000-2,500; down to 800 m and up to 2,850 in south 
full1[full1$species=="Doryfera ludovicae",][28] <- 1000 # Min elev
full1[full1$species=="Doryfera ludovicae",][29] <- 2900 # Max elev # bumping up to match Schulenberg & specimen records

# Colibri coruscans
# Stotz seems way off for this well-known generalist; Schulenberg says 400-4,500 m
full1[full1$species=="Colibri coruscans",][28] <- 400 # Min elev
full1[full1$species=="Colibri coruscans",][29] <- 4500 # Max elev 

# Colibri thalassinus (gets changed to Colibri cyanotus below)
# Stotz: Mostly 1300-2800 m, locally 1000-3300 m; also on west slop in NW at 1500-2400 m
# MSB record up to 3,153 m, rounding to 3,200
full1[full1$species=="Colibri thalassinus",][29] <- 3200 # Max elev 

# Heliangelus micraster
# Schulenberg lists range as 2200-3000 
# MSB specimen records reach nearly 3,372
full1[full1$species=="Heliangelus micraster",][29] <- 3400 # Max elev 

# Heliangelus viola
# Schulenberg et al. 2007: 1800-3200 m; note that our specimen records reach 3,279 m
full1[full1$species=="Heliangelus viola",][28] <- 1800 # Min elev
full1[full1$species=="Heliangelus viola",][29] <- 3300 # Max elev

# Adelomyia melanogenys
# Schulenberg et al. 2007: 1000-2900 m; note that our specimen records reach ~2,600 m
full1[full1$species=="Adelomyia melanogenys",][28] <- 1000 # Min elev
full1[full1$species=="Adelomyia melanogenys",][29] <- 2900 # Max elev

# Aglaiocercus kingii
# Schulenberg et al. 2007: 1200-2500 m, locally to 2800 m; our specimen records reach 2,858 m.
full1[full1$species=="Aglaiocercus kingii",][28] <- 1200 # Min elev
full1[full1$species=="Aglaiocercus kingii",][29] <- 2900 # Max elev # Bumping to 2,900 to take into account MSB records

# Lesbia victoriae
# Schulenberg et al. 2007: 2700-4100
# MSB specimen records reach low of 2,338m; rounding down to 2,300 m
full1[full1$species=="Lesbia victoriae",][28] <- 2300 # Min elev
full1[full1$species=="Lesbia victoriae",][29] <- 4100 # Max elev 

# Lesbia nuna
# Schulenberg et al. 2007: 1700-3800 m
# MSB specimen records reach ~4,036 m from Cordillera Vilcanota
full1[full1$species=="Lesbia nuna",][28] <- 1700 # Min elev
full1[full1$species=="Lesbia nuna",][29] <- 4050 # Max elev

# Oreonympha nobilis
# Schulenberg et al. 2007: 2700-3900 m
full1[full1$species=="Oreonympha nobilis",][28] <- 2700 # Min elev
full1[full1$species=="Oreonympha nobilis",][29] <- 3900 # Max elev

# Metallura phoebe
# Schulenberg et al. 2007: 2700-4300
full1[full1$species=="Metallura phoebe",][28] <- 2700 # Min elev
full1[full1$species=="Metallura phoebe",][29] <- 4300 # Max elev

# Metallura tyrianthina
# Schulenberg et al. 2007: 2400-4200 m; ocasionally down to 1900 m
full1[full1$species=="Metallura tyrianthina",][29] <- 4200 # Max elev

# Metallura eupogon
# Schulenberg et al. 2007: 3100-3800
full1[full1$species=="Metallura eupogon",][28] <- 3100 # Min elev
full1[full1$species=="Metallura eupogon",][29] <- 3800 # Max elev

# Eriocnemis vestita
# Schulenberg et al. 2007: 2450-3200 m
# MSB records to 3,293 m, so rounding upper limit to 3,300
full1[full1$species=="Eriocnemis vestita",][28] <- 2450 # Min elev
full1[full1$species=="Eriocnemis vestita",][29] <- 3300 # Max elev # Stotz was set to 3,800; seems high
# Check upper range limit w/ Chris 

# Eriocnemis luciani
# Schulenberg et al. 2007: 2450-3200 m
# MSB records to 3,779 m, so rounding upper limit to 3,800
full1[full1$species=="Eriocnemis luciani",][28] <- 2450 # Min elev
full1[full1$species=="Eriocnemis luciani",][29] <- 3800 # Max elev 

# Aglaeactis cupripennis
# Schulenberg et al. 2007: 2500-4600 m
full1[full1$species=="Aglaeactis cupripennis",][28] <- 2500 # Min elev
full1[full1$species=="Aglaeactis cupripennis",][29] <- 4600 # Max elev 

# Aglaeactis castelnaudii
# Schulenberg et al. 2007: 2500-4100 m
full1[full1$species=="Aglaeactis castelnaudii",][28] <- 2600 # Min elev
full1[full1$species=="Aglaeactis castelnaudii",][29] <- 4600 # Max elev # Bumping up based on specimen records
# Talk w/ Chris about this one - seems to be found a lot higher than recorded 

# Coeligena coeligena
# Schulenberg et al. 2007: 1000-2200; locally to 2600 m
full1[full1$species=="Coeligena coeligena",][28] <- 1000 # Min elev
full1[full1$species=="Coeligena coeligena",][29] <- 2600 # Max elev 

# Coeligena torquata
# Schulenberg et al. 2007: 1800-3000 m
full1[full1$species=="Coeligena torquata",][28] <- 1800 # Min elev
full1[full1$species=="Coeligena torquata",][29] <- 3000 # Max elev 
 
# Coeligena lutetiae
# Schulenberg et al. 2007: 2600-2950; Stotz says 3000-3750; these are two very different ranges
full1[full1$species=="Coeligena lutetiae",][28] <- 2600 # Min elev
full1[full1$species=="Coeligena lutetiae",][29] <- 3700 # Max elev 

# Coeligena violifer
# Schulenberg et al. 2007: 2500-3900 m; occasionally down to 1900 m
full1[full1$species=="Coeligena violifer",][28] <- 2500 # Min elev
full1[full1$species=="Coeligena violifer",][29] <- 3900 # Max elev 
 
# Ensifera ensifera
# Schulenberg et al. 2007: 2400-3600
# MSB records up to 3,835 m; bumping to 3,850
full1[full1$species=="Ensifera ensifera",][28] <- 2400 # Min elev
full1[full1$species=="Ensifera ensifera",][29] <- 3850 # Max elev
 
# Pterophanes cyanopterus
# Stotz and Schulenberg match exactly (2600-3700 m)
# MSB specimen records reach 4200m
full1[full1$species=="Pterophanes cyanopterus",][29] <- 4200 # Max elev 

# Boissonneaua matthewsii
# Schulenberg et al. 2007: 1500-3300 m
full1[full1$species=="Boissonneaua matthewsii",][28] <- 1500 # Min elev
full1[full1$species=="Boissonneaua matthewsii",][29] <- 3300 # Max elev

# Ocreatus underwoodii
# Schulenberg et al. 2007: 1000-2400 m
full1[full1$species=="Ocreatus underwoodii",][28] <- 1000 # Min elev
full1[full1$species=="Ocreatus underwoodii",][29] <- 2400 # Max elev

# Heliodoxa leadbeateri
# Schulenberg et al. 2007: 900-2300 m
full1[full1$species=="Heliodoxa leadbeateri",][28] <- 900 # Min elev

# Heliodoxa aurescens
# Schulenberg et al. 2007: Below 1400 m
full1[full1$species=="Heliodoxa aurescens",][29] <- 1400 # Max elev

# Campylopterus largipennis
# Schulenberg et al. 2007: Up to 1300 m
full1[full1$species=="Campylopterus largipennis",][29] <- 1300 # Max elev

# Campylopterus villaviscensio is interesting
# Schulenberg et al. 2007: from 1050-1400 m
# All MSB vouchers exceed upper range, with highest at 1776 m
full1[full1$species=="Campylopterus villaviscensio",][29] <- 1800 # Max elev

# Leucippus taczanowskii
# Schulenberg et al. 2007: 1000-2400 m
full1[full1$species=="Leucippus taczanowskii",][28] <- 350 # Min elev
full1[full1$species=="Leucippus taczanowskii",][29] <- 2800 # Max elev
# Ask Chris about this one too 

# Amazilia viridicauda
# Schulenberg et al. 2007: 1000-2750 m
# MSB specimen records collected to 3,005 m
full1[full1$species=="Amazilia viridicauda",][28] <- 1000 # Min elev
full1[full1$species=="Amazilia viridicauda",][29] <- 3050 # Max elev # Bumping up based on specimen records

# Amazilia chionogaster
# Schulenberg et al. 2007: primarily 1200-3500, locally as low as 350
# MSB low record is 294 m, so rounding down to 250 m
full1[full1$species=="Amazilia chionogaster",][28] <- 250 # Min elev
full1[full1$species=="Amazilia chionogaster",][29] <- 3500 # Max elev 
# Ask Chris about this one too 

# Taphrospilus hypostictus
# Schulenberg et al. 2007: 750-1500 m, occasionally up to 2800 m in upper Apurimac Valley 
full1[full1$species=="Taphrospilus hypostictus",][29] <- 1500 # Max elev 

# Amazilia amazilia
# Schulenberg et al. 2007: mostly below 1000, locally up to 2400 m
full1[full1$species=="Amazilia amazilia",][29] <- 2400 # Max elev 
# Ask Chris about this one too 

# Chrysuronia oenone
# Schulenberg et al. 2007: Lowlands to 1700 m
full1[full1$species=="Chrysuronia oenone",][29] <- 1700 # Max elev 

# Myrmia micrura
# Stotz says range is just sea level; Schulenberg et al. 2007: sea level to 1200 m locally
full1[full1$species=="Myrmia micrura",][29] <- 1200 # Max elev

# Klais guimeti
# Schulenberg et al. 2007: 500-1550
# MSB collection records go up to 1673 m, so adjusting upper limit to 1700 accordingly
full1[full1$species=="Klais guimeti",][29] <- 1700 # Max elev 

# Phaethornis atrimentalis
# Schulenberg et al. 2007: up to 1,100 m
# MSB collection records go up to 1209 m, so adjusting upper limit to 1250 accordingly
full1[full1$species=="Phaethornis atrimentalis",][29] <- 1250 # Max elev

# Lafresnaya lafresnayi
# Schulenberg et al. 2007: 1800-3200 m; Stotz to 3,350 m
# MSB collection records go up to 3376 m, so adjusting upper limit to 3400 accordingly
full1[full1$species=="Lafresnaya lafresnayi",][29] <- 3400 # Max elev

# Chalcostigma herrani
# Schulenberg et al. 2007: 2450-3100
# MSB collection records go up to 3421 m, so adjusting upper limit to 3450 accordingly
full1[full1$species=="Chalcostigma herrani",][29] <- 3450 # Max elev

# Oreotrochilus estella
# Schulenberg et al. 2007: 3400-4600, locally down to 3,000m
# MSB collection records go down to 3,400. But because Schulenberg uses 3,000, I'll bump down to that as low point.
full1[full1$species=="Oreotrochilus estella",][28] <- 3000 # Min elev


# Eyeball full1 species count to verify you inspected all 
sp.count.full1 <- full1 %>% group_by(species) %>% summarise(count = length(nk))
#write.csv(sp.count.full1, "AllHummingbird_SpeciesList.csv")
# Should be 77 total species 
```


# Assign hummingbirds to approprite clades 
Projecto-Garcia et al. 2013 lists many species and their associated clades. Best phylogeny McGuire et al. 2014. 
For helpful future reference, a list of hummingbirds and their respective clades: https://www.wikiwand.com/en/List_of_hummingbird_species
```{R}
# Note: Colibri thalassinus is now Mexican Violetear and we want Lesser Violetear
# change name to Colibri cyanotus by renaming the level of this factor variable 
levels(full1$species)[levels(full1$species)=="Colibri thalassinus"] <- "Colibri cyanotus"

full1$Clade <- NA # Instantiate new col; capitalize this for easy and nice plot formatting

# add clade labels for 9 hummingbird clades

# EMERALDS
full1$Clade[full1$species=="Amazilia amazilia" | 
           full1$species=="Amazilia chionogaster" | 
           full1$species=="Amazilia franciae" | 
           full1$species=="Amazilia lactea" | 
           full1$species=="Amazilia viridicauda" | 
           full1$species=="Chrysuronia oenone" | 
           full1$species=="Taphrospilus hypostictus" | 
           full1$species=="Thalurania furcata" | 
           full1$species=="Campylopterus largipennis" | 
           full1$species=="Campylopterus villaviscensio" | 
           full1$species=="Leucippus taczanowskii" | 
           full1$species=="Klais guimeti"] <- paste("Emerald")

# BEES 
full1$Clade[full1$species=="Calliphlox amethystina" | 
           full1$species=="Thaumastura cora" | 
           full1$species=="Chaetocercus mulsant" |
           full1$species=="Myrmia micrura" |
           full1$species=="Myrtis fanny" |
           full1$species=="Rhodopis vesper"] <- paste("Bee") 

# MOUNTAIN GEMS 
full1$Clade[full1$species=="Heliomaster longirostris"] <- paste("Mountain-gem")

# PATAGONA 
full1$Clade[full1$species=="Patagona gigas"] <- paste("Patagona") 

# COQUETTES (Andean clade)
full1$Clade[full1$species=="Phlogophilus harterti" | 
           full1$species=="Adelomyia melanogenys" | 
           full1$species=="Aglaiocercus kingii" | 
           full1$species=="Chalcostigma herrani" | 
           full1$species=="Chalcostigma ruficeps" | 
           full1$species=="Chalcostigma stanleyi" | 
           full1$species=="Heliangelus amethysticollis" | 
           full1$species=="Heliangelus micraster" | 
           full1$species=="Heliangelus viola" | 
           full1$species=="Metallura eupogon" | 
           full1$species=="Metallura phoebe" | 
           full1$species=="Metallura tyrianthina" | 
           full1$species=="Oreotrochilus estella" | 
           full1$species=="Oreotrochilus leucopleurus" | 
           full1$species=="Oreotrochilus melanogaster" | 
           full1$species=="Polyonymus caroli" | 
           full1$species=="Sephanoides sephaniodes" | 
           full1$species=="Oreonympha nobilis" |
           full1$species=="Lesbia nuna" |
           full1$species=="Lesbia victoriae"] <- paste("Coquette")

# BRILLIANTS (Andean clade)
full1$Clade[full1$species=="Aglaeactis castelnaudii" | 
           full1$species=="Aglaeactis cupripennis" | 
           full1$species=="Boissonneaua matthewsii" | 
           full1$species=="Coeligena coeligena" | 
           full1$species=="Coeligena iris" | 
           full1$species=="Coeligena lutetiae" | 
           full1$species=="Coeligena torquata" | 
           full1$species=="Coeligena violifer" | 
           full1$species=="Ensifera ensifera" | 
           full1$species=="Eriocnemis aline" | 
           full1$species=="Eriocnemis luciani" | 
           full1$species=="Eriocnemis vestita" | 
           full1$species=="Haplophaedia aureliae" | 
           full1$species=="Heliodoxa aurescens" | 
           full1$species=="Heliodoxa leadbeateri" | 
           full1$species=="Heliodoxa rubinoides" | 
           full1$species=="Lafresnaya lafresnayi" | 
           full1$species=="Ocreatus underwoodii" | 
           full1$species=="Pterophanes cyanopterus"] <- paste("Brilliant")

# MANGOS
full1$Clade[full1$species=="Anthracothorax nigricollis" | 
           full1$species=="Colibri coruscans" | 
           full1$species=="Colibri cyanotus" |  # Remember you changed Colibri thalassinus to cyanotus above 
           full1$species=="Doryfera ludovicae" |
           full1$species=="Schistes geoffroyi"] <- paste("Mango")

# HERMITS 
full1$Clade[full1$species=="Eutoxeres aquila" | 
           full1$species=="Eutoxeres condamini" | 
           full1$species=="Glaucis hirsutus" | 
           full1$species=="Phaethornis atrimentalis" | 
           full1$species=="Phaethornis guy" | 
           full1$species=="Phaethornis hispidus" | 
           full1$species=="Phaethornis malaris" | 
           full1$species=="Phaethornis philippii" | 
           full1$species=="Phaethornis ruber" | 
           full1$species=="Phaethornis stuarti" | 
           full1$species=="Phaethornis syrmatophorus" | 
           full1$species=="Threnetes leucurus"] <- paste("Hermit")

# TOPAZES AND JACOBINS 
full1$Clade[full1$species=="Florisuga mellivora"] <- paste("Topaz") 

# Summarize by clade 
clade.summary <- full1 %>% group_by(Clade) %>% summarise(count = length(species)); clade.summary
```


# Classify elev range breadth, elevational range class, and *P*O2 class
```{r}
# Create elevational range breadth variable: elev_max-elev_min
full1$elev_range <- (full1$elev_max-full1$elev_min)

# Recalculate elev_midpoint
full1$elev_midpt <- ((full1$elev_min + full1$elev_max)/2)

# Make a small summary table with elevational ranges of each species 
# (note: in this case, 'mean' is just a means to get actual range; since only one value, mean will just be elev_range)
elev.range.summary <- full1 %>% group_by(species) %>% summarise(range = mean(elev_range))
# Species w/ ranges <1000 m vertical distance: 15
# species w/ ranges < 2,000 m vertical distance: 63
# species w/ elev ranges > 2,500 m vertical distance: 9

# -------

# Assign Elevational range class ('range_class'): 
# Specialist =  elev_range <1500m
# generalist = elev_range >2,500 m 
# Neither = elev_range 1500-2499 m 
full1$range_class <- NA
full1$range_class[full1$elev_range < 1500] <- "specialist"
full1$range_class[full1$elev_range > 2500] <- "generalist"
full1$range_class[is.na(full1$range_class)] <- "neither"

# PO2_class = degree to which species may specialize on PO2
# Specialists occur only > 2,500 m
# Generalists occur < 1000 m and > 3,000 m 
full1$po2_class <- NA
full1$po2_class[full1$elev_min > 2499] <- "specialist"
full1$po2_class[full1$elev_min < 1000 & full1$elev_max > 2999] <- "generalist"
full1$po2_class[is.na(full1$po2_class)] <- "neither"

# Summarize PO2 classes
# This is a little janky but allows me to sort by po2_class
po2.class.summary <- full1 %>% group_by(species) %>% count(po2_class)
# 6 generalists, 14 specialists, the rest are neither 

# Separate out PO2 specialists and generalists:
# Patagona-free subset:
po2.specialist <- full1[which(full1$po2_class == "specialist"),] 
po2.generalist <- full1[which(full1$po2_class == "generalist"),] 
```


# Calculate elevational range position 
```{r}
# Elev_position = scale of 0-1 for individual positions along a species elev range (e.g. 0=bottom, 0.5=middle, 1=top)
# This is more meaningful than looking at elev_range width comparisons since my analysis is individual-level
# We expect that birds at the top of their range have higher blood values (generally) 
full1$elev_position <- 1-((full1$elev_max-full1$elev)/full1$elev_range)
 
# Any value >1 indicates that birds were collected higher than recorded max_elev; values <1 record the opposite
# Since I checked and vetted these above, all values should fall within 0-1
# Also remember: if I do want to 'automate' more, Ethan has nice function for this  
```


# Save final dataset and adjust data structure for modeling 
```{r}
# Drop lingering rows from Stotz merge (some of which still have NAs due to taxonomic changes)
full1 <- full1[ , -which(names(full1) %in% c("genus","species.y"))] 

# save final dataset 
final <- full1

# All characters need to be made into factors 
str(final)
final$sex <- as.factor(final$sex)
final$age <- as.factor(final$age)
final$month <- as.factor(final$month)
final$locality <- as.factor(final$locality)
final$department <- as.factor(final$department)
final$range_class <- as.factor(final$range_class)
final$po2_class <- as.factor(final$po2_class)
final$Clade <- as.factor(final$Clade)

# Looks good! Should be ready for modeling.
# Best approach for balancing NA omission w/ preserving data is probably to create a subset for each blood parameter
# e.g. subset to include JUST data you need for models, call datasets something like 'final.hb', 'final.hct', etc.
```





#### SPATIAL AND ENVIRONMENTAL DATA ####

# Country polygons, elevation rasters, BioClim download
```{r}
library(raster)
library(sp)

# DOWNLOAD COUNTRY POLYGONS 
# Search country 3-letter ISO code: https://www.iso.org/obp/ui/#search
# Note: Chile's southern coastline makes plotting the outline near impossible; R  crashes every time 
# If you do want to plot outline, write straight to .pdf instead of plotting w/in RStudio
# Best approach seems to be to crop the extent of hte outline to remove Juan Fernandez Islands
# Then to crop and mask climate rasters to Peru and Chile
# THEN to merge these two rasters. 
peru <- raster::getData('GADM', country = "PER", level = 0) # download country shape  
chile <- raster::getData('GADM', country = "CHL", level = 0) 
extent(chile) # look at the spatial extent (lat, lon) of Chile; compare to map 
# Chile's extent goes way out into the Pacific because of Juan Fernandez Islands; get rid of these
extent_mainland <- extent(c(-76, -66.41472, -55.98403,-17.50755)) # Use map to pick a reasonable x min
chile <- crop(chile, extent_mainland) # this takes ~3-5 mins

# Download surrounding countries for visualization/map-making
countries = c("BRA", "ARG", "ECU", "BOL", "VEN", "COL", "GUY", "SUR", "PRY", "URY") 
manycountries = do.call("bind", lapply(countries, function(x) raster::getData('GADM', country=x, level=0)))

# Merge country polygons 
southamerica_countries <- merge(peru, chile, manycountries)

# ELEVATION DATA
# # Note similar weirdness with Chile elevation layer as w/ Chile polygons; namely, Juan Fernandez Islands make extent wonky
# peru_alt <- raster::getData('alt', country = "PER", mask=TRUE) # Peru elev data
# chile_alt <- raster::getData('alt', country = "CHL", mask=TRUE) # Chile elev data; a list of 2
# # In Chile alt list: [[1]] = mainland, [[2]] = Juan Fernandez Islands
# chile_alt <- chile_alt[[1]] # Assign mainland Chile to to Chile_alt because that's all I want 
# proj4string(peru_alt) <- CRS("+init=epsg:4326")
# proj4string(chile_alt) <- CRS("+init=epsg:4326")
# #writeRaster(peru_alt, "peru_alt.grd", overwrite=TRUE) # Save raster files 
# #writeRaster(chile_alt, "chile_alt.grd", overwrite=TRUE) 
# peru_chile_elev <- merge(peru_alt, chile_alt) # Merge Peru and Chile elev layers for map-making
# #writeRaster(peru_chile_elev, "peru_chile_alt.grd", overwrite=TRUE) 
# 
# # Get elev data for surrounding countries to make your map prettier
# bol_alt <- raster::getData('alt', country = "BOL", mask=TRUE) # Bolivia elev data
# arg_alt <- raster::getData('alt', country = "ARG", mask=TRUE) # Argentina elev data
# bra_alt <- raster::getData('alt', country = "BRA", mask=TRUE) # Brazil elev data
# ecu_alt <- raster::getData('alt', country = "ECU", mask=TRUE) # Ecuador elev data
# col_alt <- raster::getData('alt', country = "COL", mask=TRUE) # Colombia elev data
# vez_alt <- raster::getData('alt', country = "VEN", mask=TRUE) # Venezuela elev data
# guy_alt <- raster::getData('alt', country = "GUY", mask=TRUE) # Guyana elev data
# sur_alt <- raster::getData('alt', country = "SUR", mask=TRUE) # Suriname elev data
# pry_alt <- raster::getData('alt', country = "PRY", mask=TRUE) # Paraguay elev data
# uru_alt <- raster::getData('alt', country = "URY", mask=TRUE) # Uruguay elev data
# surrounding_elevs <- merge(bol_alt, arg_alt, bra_alt, ecu_alt, col_alt, vez_alt, guy_alt, sur_alt, pry_alt, uru_alt) 
# # Merge elev layers of surrounding countries 
# south_america_elevs <- merge(peru_chile_elev, surrounding_elevs) # Merge surrounding countries w/ Chile and Peru
# #writeRaster(surrounding_elevs, "surrounding_elevs.grd", overwrite=TRUE) # Save raster files 
#writeRaster(south_america_elevs, "south_america_elevs.grd", overwrite=TRUE) 

# GET CLIMATE DATA, CROP TO COUNTRY SHAPES
climate <- raster::getData("worldclim", var="bio", res = 2.5) # download bioclim data
projection(climate) <- CRS("+init=epsg:4326")
chile_clim <- climate %>% crop(., chile) %>% raster::mask(., chile) # Crop and mask climate raster to Chile
peru_clim <- climate %>% crop(., peru) %>% raster::mask(., peru) # Crop and mask climate raster to Peru
proj4string(chile_clim) == proj4string(peru_clim) #check projections between Peru and Chile files 
peru_chile <- merge(peru_clim, chile_clim) # Merge Peru and Chile rasters; this forms raster brick 
plot(peru_clim$bio1) # Plot this as a test that the above cropping and masking worked
# # NOTE: must call raster::mask()! Otherwise mask() is actually masked by another function and I get an error message. 
# # Masking literally masks the shapes of adjoining countries (i.e. "messy stuff") for a clean look; not masking...doesn't.
plot(peru_chile$layer.12) # plot to verify this worked; layer is calling a bioclim variable
proj4string(climate) == proj4string(peru_chile) #check projections
#writeRaster(climate, "world_climate.grd", overwrite=TRUE) # the world?; write out bioclim brick; takes ~10-15 mins
#writeRaster(peru_chile, "peru_chile_bioclim.grd", overwrite=TRUE) # Just Peru and Chile

# Note: I tried to switch Sabrina's order to merge polygons and THEN clip + mask but this never worked
# For some reason it kept only grabbing Peru, which is weird. Here's the code anyways:
# peru_chile <- merge(peru, chile) # Merge country polygons 
#peru_chile_clim <- climate %>% crop(., peru_chile) %>% raster::mask(., peru_chile) # Crop and mask climate to countries 


# READ IN RASTERS TO AVOID RUNNING ALL CODE ABOVE 
# Need to run BioClim data above to ensure that peru_chile is a raster brick otherwise you can't extract BioClim
#peru_chile_bioclim <- raster("./peru_chile_bioclim.grd") # read in peru_chile BioClim data
peru_alt <- raster("./peru_alt.grd") #read elevation raster
chile_alt <- raster("./chile_alt.grd") #read elevation raster
surrounding_elevs <- raster("./surrounding_elevs.grd") # surrounding South American country elevs
south_america_elevs <- raster("./south_america_elevs.grd") # All South American country elevs combined


###### PROCESS SPATIAL DATA 

# Create a spatial data frame that has our community localities
# NAs in the data will create problems; e.g. you won't be able to make spatial data frame 
locality.coords <- SpatialPointsDataFrame(
                   matrix(c(final$lon, final$lat), ncol = 2),
                   data.frame(ID=1:nrow(final),
                   sampling.locality = final$locality,
                   sampling.elev = final$elev),
                   proj4string=CRS("+init=epsg:4326")
                   )

# Extract mean temp and mean precip for our all localities 
# Use clipped shape of peru_chile to do this
# If you use peru_clim it won't grab Chile points; if you use chile_clim it won't grab Peru points
# With raster brick, need to specify $layer.1 for $bio1; note this name change or you'll get an error
final$bio1 <- raster::extract(peru_chile$layer.1, locality.coords)
final$bio2 <- raster::extract(peru_chile$layer.2, locality.coords)
final$bio3 <- raster::extract(peru_chile$layer.3, locality.coords)
final$bio4 <- raster::extract(peru_chile$layer.4, locality.coords)
final$bio5 <- raster::extract(peru_chile$layer.5, locality.coords)
final$bio6 <- raster::extract(peru_chile$layer.6, locality.coords)
final$bio7 <- raster::extract(peru_chile$layer.7, locality.coords)
final$bio8 <- raster::extract(peru_chile$layer.8, locality.coords)
final$bio9 <- raster::extract(peru_chile$layer.9, locality.coords)
final$bio10 <- raster::extract(peru_chile$layer.10, locality.coords)
final$bio11 <- raster::extract(peru_chile$layer.11, locality.coords)
final$bio12 <- raster::extract(peru_chile$layer.12, locality.coords)
final$bio13 <- raster::extract(peru_chile$layer.13, locality.coords)
final$bio14 <- raster::extract(peru_chile$layer.14, locality.coords)
final$bio15 <- raster::extract(peru_chile$layer.15, locality.coords)
final$bio16 <- raster::extract(peru_chile$layer.16, locality.coords)
final$bio17 <- raster::extract(peru_chile$layer.17, locality.coords)
final$bio18 <- raster::extract(peru_chile$layer.18, locality.coords)
final$bio19 <- raster::extract(peru_chile$layer.19, locality.coords)

# Now write out this dataset, which represents the near-final version of what you'll want for analysis
#write.csv(final, "AllHummingbirdBlood_Wrangled_withBioClim_10-27-20.csv")
```



# Spatial coords PCA 
This is for variable reduction of BioClim variables; look at loadings to understand which directions things work in. 
Collapse temp and aridity variables and use first PCA axis in models. 
```{r}
# Code from Sabrina 
# PCA and scale raster data for temp variables and precip variables

# Temperature PCA using BioClim vars 1-11
tempPCA <- rasterPCA(subset(peru_chile, 1:11), spca = TRUE) #creates a raster of PCAd temp vals
summary(tempPCA$model) #PC1 has 73.4% of variation and 89.5% w/ PC1+PC2
tempPCA$model$loadings #

#Pull PC1 and PC2 values for each community and add to dataset
final$tempPC1 <- raster::extract(tempPCA$map$PC1, locality.coords)
final$tempPC2 <- raster::extract(tempPCA$map$PC2, locality.coords)

plot(bio11 ~ tempPC1, data=final) #PC1 is highly correlated with mean temp
plot(bio11 ~ tempPC1, data=final) #PC1 also highly correlated w/ mean temp of coldest quarter 
# Fairly correlated w/ bio5, bio6
 
## -------

# Precipitation PCA using BioClim vars 12-19
precipPCA <- rasterPCA(subset(peru_chile, 12:19), spca = TRUE)
summary(precipPCA$model) #PC1 has 79.8% of variation and 89.1% of variation w/ PC1+PC2
precipPCA$model$loadings #

# Add PC1 and PC2 values to dataset 
final$precipPC1 <- raster::extract(precipPCA$map$PC1, locality.coords)
final$precipPC2 <- raster::extract(precipPCA$map$PC2, locality.coords)

plot(bio12 ~ precipPC1, data=final) # Moderate correlation between PC1 and mean annual precip
# Overal not many correlations here 
plot(bio19 ~ tempPC2, data=final) 

# Write out PCA rasters for later map generation:
# writeRaster(tempPCA$map$PC1, "tempPC1.grd", overwrite=TRUE)
# writeRaster(tempPCA$map$PC2, "tempPC2.grd", overwrite=TRUE)
# writeRaster(precipPCA$map$PC1, "precipPC1.grd", overwrite=TRUE)
# writeRaster(precipPCA$map$PC2, "precipPC2.grd", overwrite=TRUE)


# Look at distributions of BioClim variables and transform if necessary 
p <- ggpairs(subset(final, select = c(elev, elev_position, mass, hb, hct, trbc, mcv, mch, mchc, tempPC1, tempPC2, precipPC1, precipPC2))) 
print(p)
# PC variables look a little wonky 

hist(final$tempPC1)
qqPlot(final$tempPC1) # off at tails 

hist(final$tempPC2)
qqPlot(final$tempPC2) # off at left tail  

hist(final$precipPC1)
qqPlot(final$precipPC2) # weird skew 

hist(final$precipPC2)
qqPlot(final$precipPC2) # off at left tail  
```


WORLDCLIM CODEBOOK 
```

They are coded as follows:

BIO1 = Annual Mean Temperature
BIO2 = Mean Diurnal Range (Mean of monthly (max temp - min temp))
BIO3 = Isothermality (BIO2/BIO7) (* 100)
BIO4 = Temperature Seasonality (standard deviation *100)
BIO5 = Max Temperature of Warmest Month
BIO6 = Min Temperature of Coldest Month
BIO7 = Temperature Annual Range (BIO5-BIO6)
BIO8 = Mean Temperature of Wettest Quarter
BIO9 = Mean Temperature of Driest Quarter (*transformed)
BIO10 = Mean Temperature of Warmest Quarter
BIO11 = Mean Temperature of Coldest Quarter
BIO12 = Annual Precipitation
BIO13 = Precipitation of Wettest Month
BIO14 = Precipitation of Driest Month (*transformed)
BIO15 = Precipitation Seasonality (Coefficient of Variation)
BIO16 = Precipitation of Wettest Quarter
BIO17 = Precipitation of Driest Quarter (*transformed)
BIO18 = Precipitation of Warmest Quarter
BIO19 = Precipitation of Coldest Quarter

```


# SAMPLING LOCALITY MAP 
Mapping tips here: https://www.r-spatial.org/r/2018/10/25/ggplot2-sf.html
```{r}
library(viridis)
library(RColorBrewer)

# CROP MAP EXTENT FOR BETTER VISUALIZATION 
# Wide version: 
# xmin = -90
# xmax = 41.7
# y min = -45.3
# y max = 0.8

# Zoomed version: 
# xmin = -86
# xmax = 47.4
# y min = -37
# y max = 0.8

# crop raster and country lines to a smaller box (hard to get right, requires a lot of tinkering)
# south_america_elevs_cropped = all elevation rasters combined 
# southamerica_cropped = all country polygons combined 
# These are perfect but fairly zoomed out to capture more of southern Chile 
#ex <- extent(c(-90, -41.7, -45.3, 0.8)) # These go: xmin (left), xmax (right), ymin (bottom), ymax (top)
# Zoomed in version: 
ex <- extent(c(-86, -47.4, -37, 0.8)) # These go: xmin (left), xmax (right), ymin (bottom), ymax (top)
south_america_elevs_cropped <- crop(south_america_elevs, ex)
extent(south_america_elevs_cropped)
southamerica_cropped <- crop(southamerica_countries, ex)

# FINAL MAP
# Peru and Chile sampling localities w/ rest of South America 
pdf("./Peru_and_Chile_SamplingMap_WithSouthAmerica_Cropped_Outlines.pdf", useDingbats = F)
#plot(south_america_elevs_cropped, col = colorRampPalette(c("grey20","grey80"))(100), cex.axis=1.5, ext=ex)
plot(south_america_elevs_cropped, col=magma(200), cex.axis=1.5, ext=ex)
plot(manycountries, add=TRUE, border = "gray92", lwd = 0.05) # outlines surrounding countries; must be first to "layer"
#plot(manycountries, add=TRUE, border = "gray28", lwd = 0.5) # outlines surrounding countries; must be first to "layer"
plot(peru, add=TRUE, border = "white", lwd = 1.5) # This outlines Peru 
plot(chile, add=TRUE, border = "white", lwd = 1.5) # This outlines Peru 
#plot(locality.coords, add=TRUE, pch=19, cex=0.8, col="olivedrab1") # olivedrab1 points w/ no outline # doesn't look great
plot(locality.coords, add=TRUE, pch=21, cex=1.1, bg="olivedrab1") # pch=21 is circle w/ black outline
# olivedrab1 is electric and amazing; olivedrab2 *slightly* subtler and nice, but doesn't 'pop' quite as much
# olivedrab3 looks fantastic; yellow looks nice
addscalebar(plotepsg = 4326, widthhint=0.20, labelpadin = 0.08, label.cex = 0.8, label.col = "black", pos = "bottomleft")
#with(CommunitySpatial, text(CommunitySpatial,
#  labels=CommunitySpatial$ID, pos=4, cex=1.3))
dev.off()

# Played around a lot with aesthetics, namely: w/out country outlines, it's a little hard to orient in South America. 
# EBL thinks 2 colors to denote country lines is a little distracting
# I sorta like thin gray28 spiderweb lines for surrounding countries and thick white lines for Peru and Chile
# I also played around with ALL white outlines but thick for Peru and Chile and super thin for surrounding countries
# I think the best approach might be THICK white Peru + Chile outlines w/ slightly dulled bright gray spiderwebs
# that are much thinner; this makes it look like all the same color, but spiderwebs won't 'pop' as much to distract

# Notes on elevation color palettes, ranked: 
# magma is stunning, sophisticated
# cividis is stunning and lovely; equivalent topography resolution to inferno but subtler palette and nice 
# inferno looks great as well but doesn't pick up as much subtlety in topography as magma 
# Viridis is stunning but is what Sabrina used for Peru malaria paper; magma also picks up more detail in topography
# plasma is ok but looks a little 'cheaper' than magma, which is very sophisticated
# terrain.colors: base R option, looks cheap
# topo.colors: actually painful for the eyes 

# See allhum_blood_comp_CUT.Rmd for other map versions that I cut/didn't include here 
```


# IMPUTE MISSING VALUES? = NO! 
If we don't eliminate NAs or impute missing values, brms won't give proper estimates
```{r}
# Notes on imputation: 
# Don't impute categorical variables
# Don’t transform skewed variables: when you transform a variable to meet normality assumptions before imputing, you
    # change the distribution of that variable AND the relationship between that variable and the others you use to impute.
    # Doing so can lead to imputing outliers, creating more bias than just imputing the skewed variable
# The more imputations, the more robust the estimates. Bodner (2008) recommends as many imputations as % missing data.
    # Since our data is 57% complete, I'll run 50 imputations (43% incomplete + buffer)

#library(mice)
#imp <- mice(final, m=5, print=TRUE, maxit=5, method="cart", seed=500) # m= # imputations; maxit= # iterations/datasets
# In order to get this to work, specify method='cart' for 'classification and regression trees'
# High number of unbalanced factors in the data mean default linear regression methods won't work 
# method=cart means 'mice' won't do X matrix inversion and imputation will work 
# mice() returns a list that you can pass directly into brms with:  
# fit_imp1 <- brm_multiple(bmi ~ age*chl, data = imp, chains = 2)

# BUT, AS A GENERAL RULE, DON'T IMPUTE DATA IF YOU HAVE >5% MISSING VALUES!
percent.missing <- function(x){sum(is.na(x))/length(x)*100} # Function to calculate % missing values 
apply(final, 2, percent.missing)
# apply(final.sub, 1, percent.missing)

# We have a lot of missing blood data! % missing data: 
# mass = 1.52%(this would be ok for imputation)
# hb: 13.3% 
# hct 8.88%
# mcv: 38%
# mchc: 21.8%
# mch: 40.4%

# This is really *too* much missing data to be able to impute well. 
# It's clunker, but probably more reliable to create subsets for each blood parameter model set, then remove NAs from those

# SECOND BUT: What we can do is as Chris suggested and impute mass values by taking means across species and sex
# Start with a summary of species means for each sex 
library(dplyr)
species.mass.summary <- final %>% group_by(species, sex) %>%
                                  dplyr::summarize(mean_mass = mean(mass, na.rm = TRUE), 
                                  count =n())
species.mass.summary
# Call summarize from dplyr with dplyr::summarize or R pulls from something else 

# Create a teeny subset with all mass=NA values 
nomasses <- final[-which(!is.na(final$mass)),] # 19 observations missing masses that you'll "impute" w/ species mean values
final.nomasses <- final[-which(is.na(final$mass)),] # Drop missing mass records from final dataset 

# Now, impute missing values manually based on species mean values for each sex 
# There has *got* to be a less manual and clunky as hell way to do this...
colnames(nomasses) # col 16 is mass 
nomasses[nomasses$species=="Aglaiocercus kingii" & nomasses$sex=="female",][16] <- 4.465556 # Mean female mass
nomasses[nomasses$species=="Amazilia chionogaster" & nomasses$sex=="female",][16] <- 5.513333
nomasses[nomasses$species=="Chrysuronia oenone" & nomasses$sex=="male",][16] <- 4.599000
nomasses[nomasses$species=="Coeligena violifer" & nomasses$sex=="male",][16] <- 7.894615
nomasses[nomasses$species=="Coeligena violifer" & nomasses$sex=="female",][16] <- 7.693182
nomasses[nomasses$species=="Colibri coruscans" & nomasses$sex=="male",][16] <- 8.466667
nomasses[nomasses$species=="Eriocnemis luciani" & nomasses$sex=="female",][16] <- 6.491429 
nomasses[nomasses$species=="Eriocnemis luciani" & nomasses$sex=="male",][16] <- 7.133333
nomasses[nomasses$species=="Florisuga mellivora" & nomasses$sex=="male",][16] <- 6.882500
nomasses[nomasses$species=="Metallura tyrianthina" & nomasses$sex=="male",][16] <- 3.784615
nomasses[nomasses$species=="Myrtis fanny" & nomasses$sex=="female",][16] <- 2.648000
nomasses[nomasses$species=="Oreotrochilus melanogaster" & nomasses$sex=="female",][16] <- 7.078448
nomasses[nomasses$species=="Patagona gigas" & nomasses$sex=="female",][16] <- 19.972381
nomasses[nomasses$species=="Pterophanes cyanopterus" & nomasses$sex=="female",][16] <- 10.465714
nomasses[nomasses$species=="Pterophanes cyanopterus" & nomasses$sex=="male",][16] <- 11.550000
nomasses[nomasses$species=="Taphrospilus hypostictus" & nomasses$sex=="female",][16] <- 7.150000

# Recombine datasets 
final <- rbind(final.nomasses, nomasses)

# Check percent missing mass with this, should now be zero 
percent.missing <- function(x){sum(is.na(x))/length(x)*100} # Function to calculate % missing values 
apply(final, 2, percent.missing)

# yay! No missing mass data. 

# Final dataset without Hb genotypes:
write.csv(final, "AllHummingbirdBlood_Wrangled_withBioClim_andPCA_NoHbGenotypes_10-27-20.csv")

```

Vignette on handling missing values w/ brms: 
https://cran.r-project.org/web/packages/brms/vignettes/brms_missings.html

Another imputation tutorial: 
https://stefvanbuuren.name/fimd/sec-toomany.html

More important imputation stuff: 
https://www.theanalysisfactor.com/multiple-imputation-5-recent-findings-that-change-how-to-use-it/

Imputation help: https://www.kaggle.com/c/house-prices-advanced-regression-techniques/discussion/24586



# ADD IN HEMOGLOBIN GENOTYPES
```{r}
# Instantiate new columns 
final$b13 <- NA
final$b83 <- NA 

# Add in beta13 position amino acids 

# SER at beta 13
final$b13[final$species=="Patagona gigas" | 
          final$species=="Chalcostigma stanleyi" | 
          final$species=="Oreotrochilus melanogaster" | 
          final$species=="Oreotrochilus estella" | 
          final$species=="Oreotrochilus leucopleurus"] <- paste("Ser")

# SER at beta 83
final$b83[final$species=="Patagona gigas" | 
          final$species=="Chalcostigma stanleyi" | 
          final$species=="Oreotrochilus melanogaster" | 
          final$species=="Oreotrochilus estella" | 
          final$species=="Oreotrochilus leucopleurus"] <- paste("Ser")

# GLY at beta 13
final$b13[final$species=="Amazilia viridicauda" | 
          final$species=="Amazilia chionogaster" | 
          final$species=="Amazilia lactea" | 
          final$species=="Amazilia amazilia" | 
          final$species=="Chrysuronia oenone" | 
          final$species=="Taphrospilus hypostictus" | 
          final$species=="Thalurania furcata" | 
          final$species=="Thaumastura cora" | 
          final$species=="Calliphlox amethystina" |
          final$species=="Metallura tyrianthina" | 
          final$species=="Metallura phoebe" |
          final$species=="Chalcostigma ruficeps" | 
          final$species=="Chalcostigma herrani" | 
          final$species=="Polyonymus caroli" | 
          final$species=="Aglaiocercus kingii" | 
          final$species=="Adelomyia melanogenys" |
          final$species=="Heliangelus micraster" |
          final$species=="Heliangelus viola" |
          final$species=="Heliangelus amethysticollis" |
          final$species=="Phlogophilus harterti" |
          final$species=="Coeligena lutetiae" |
          final$species=="Coeligena iris" |
          final$species=="Coeligena violifer" |
          final$species=="Coeligena torquata" |
          final$species=="Coeligena coeligena" |
          final$species=="Heliodoxa leadbeateri" |
          final$species=="Heliodoxa rubinoides" |
          final$species=="Ocreatus underwoodii" |
          final$species=="Boissonneaua matthewsii" |
          final$species=="Pterophanes cyanopterus" |
          final$species=="Ensifera ensifera" |
          final$species=="Aglaeactis castelnaudii" |
          final$species=="Lafresnaya lafresnayi" |
          final$species=="Eriocnemis luciani" |
          final$species=="Eriocnemis aline" |
          final$species=="Haplophaedia aureliae" |
          final$species=="Anthracothorax nigricollis" |
          final$species=="Colibri coruscans" |
          final$species=="Schistes geoffroyi" |
          final$species=="Doryfera ludovicae" |
          final$species=="Phaethornis ruber" |
          final$species=="Phaethornis hispidus" |
          final$species=="Phaethornis malaris" |
          final$species=="Phaethornis guy" |
          final$species=="Phaethornis syrmatophorus" |
          final$species=="Glaucis hirsutus" |
          final$species=="Threnetes leucurus" |
          final$species=="Eutoxeres condamini" | 
          final$species=="Florisuga mellivora" |  # Everything below F. mellivora was added w/ Chris data or inference 
          final$species=="Amazilia franciae" | 
          final$species=="Campylopterus villaviscensio" | 
          final$species=="Eutoxeres aquila" | 
          final$species=="Klais guimeti" | 
          final$species=="Leucippus taczanowskii" | 
          final$species=="Myrmia micrura" | 
          final$species=="Phaethornis atrimentalis" | 
          final$species=="Sephanoides sephaniodes" | 
          final$species=="Campylopterus largipennis" | 
          final$species=="Heliodoxa aurescens" | 
          final$species=="Heliomaster longirostris" | 
          final$species=="Phaethornis philippii" | 
          final$species=="Phaethornis stuarti" | 
          final$species=="Chaetocercus mulsant" | 
          final$species=="Eriocnemis vestita" | 
          final$species=="Myrtis fanny" |
          final$species=="Aglaeactis cupripennis" |
          final$species=="Colibri cyanotus" |
          final$species=="Lesbia nuna" |
          final$species=="Lesbia victoriae" |
          final$species=="Metallura eupogon" |
          final$species=="Oreonympha nobilis" |
          final$species=="Rhodopis vesper" ] <- paste("Gly")

# GLY at beta 83 (black tips in Projecto-Garcia tree)
final$b83[final$species=="Amazilia lactea" | 
          final$species=="Amazilia amazilia" | 
          final$species=="Chrysuronia oenone" | 
          final$species=="Taphrospilus hypostictus" | 
          final$species=="Calliphlox amethystina" |
          final$species=="Aglaiocercus kingii" | 
          final$species=="Adelomyia melanogenys" |
          final$species=="Heliangelus micraster" |
          final$species=="Heliangelus viola" |
          final$species=="Heliangelus amethysticollis" |
          final$species=="Phlogophilus harterti" |
          final$species=="Coeligena torquata" |
          final$species=="Coeligena coeligena" |
          final$species=="Heliodoxa leadbeateri" |
          final$species=="Heliodoxa rubinoides" |
          final$species=="Haplophaedia aureliae" |
          final$species=="Anthracothorax nigricollis" |
          final$species=="Schistes geoffroyi" |
          final$species=="Phaethornis ruber" |
          final$species=="Phaethornis hispidus" |
          final$species=="Phaethornis malaris" |
          final$species=="Phaethornis guy" |
          final$species=="Phaethornis syrmatophorus" |
          final$species=="Glaucis hirsutus" |
          final$species=="Threnetes leucurus" |
          final$species=="Eutoxeres condamini" | 
          final$species=="Florisuga mellivora" |  # Everything below F. mellivora added w/ Chris data or inference 
          final$species=="Amazilia franciae" |
          final$species=="Campylopterus villaviscensio" |
          final$species=="Eutoxeres aquila" |
          final$species=="Klais guimeti" |
          final$species=="Leucippus taczanowskii" |
          final$species=="Myrmia micrura" |
          final$species=="Phaethornis atrimentalis" |
          final$species=="Sephanoides sephaniodes" |
          final$species=="Campylopterus largipennis" |
          final$species=="Heliodoxa aurescens" |
          final$species=="Heliomaster longirostris" |
          final$species=="Phaethornis philippii" |
          final$species=="Phaethornis stuarti" ] <- paste("Gly")

# SER at beta 83 (gray tips in Projecto-Garcia tree)
final$b83[final$species=="Amazilia viridicauda" | 
          final$species=="Amazilia chionogaster" | 
          final$species=="Thalurania furcata" |
          final$species=="Thaumastura cora" |
          final$species=="Metallura tyrianthina" |
          final$species=="Metallura phoebe" |
          final$species=="Chalcostigma ruficeps" |
          final$species=="Chalcostigma herrani" |
          final$species=="Polyonymus caroli" |
          final$species=="Coeligena lutetiae" |
          final$species=="Coeligena iris" |
          final$species=="Coeligena violifer" |
          final$species=="Ocreatus underwoodii" |
          final$species=="Boissonneaua matthewsii" |
          final$species=="Pterophanes cyanopterus" |
          final$species=="Ensifera ensifera" |
          final$species=="Aglaeactis castelnaudii" |
          final$species=="Lafresnaya lafresnayi" |
          final$species=="Eriocnemis luciani" |
          final$species=="Eriocnemis aline" |
          final$species=="Colibri coruscans" |
          final$species=="Doryfera ludovicae" |  # Everything below D. ludoviciae added w/ Chris data or inference 
          final$species=="Chaetocercus mulsant" | 
          final$species=="Eriocnemis vestita" | 
          final$species=="Myrtis fanny" | 
          final$species=="Aglaeactis cupripennis" | 
          final$species=="Colibri cyanotus" | 
          final$species=="Lesbia nuna" | 
          final$species=="Lesbia victoriae" | 
          final$species=="Metallura eupogon" | 
          final$species=="Oreonympha nobilis" | 
          final$species=="Rhodopis vesper" ] <- paste("Ser")

# Combine b13 and b83 to make a b13-b83 genotype: 
final$b13b83.genotype <- paste(final$b13, final$b83,sep="-") # specify dash separator; can change this 

# Hb genotype summary 
# Summarize by clade 
genotype.summary <- final %>% group_by(b13b83.genotype) %>% summarise(count = length(species)); genotype.summary
```

How we did this: We first added Hb genotypes from Projecto-Garcia et al. 2013. If any were missing, Jessie looked through old Hbba sequence files from Chris, assembled alignments with 5 GenBank Reference sequences from Projecto-Garcia, and determined Hb genotypes that way (see description of this workflow in Hbba-genotyping-workflow-hummingbird-comparative-data_08-27-20.doc). Then, if species were still missing, Jessie took Hb genotypes from the McGuire et al. .nex file that Chris sent over that has a nice little alignment of b13-b83 genotypes for many species. 

After doing this, we were still left with 15 species missing Hb genotypes. We made educated best guesses based on genotypes of sister taxa, and elevational ranges. We think most of these are very solidly-inferred, though we know this is not a foolproof method. 
Rationale for best guesses is provided in: ComparativeHummingbirdData_SpeciesMissingHbGenotypes_Data&BestGuesses_08-28-20.xlsx

Here are the list of 15 species for which we used "best guess" Hb genotype info: 
Amazilia franciae
Campylopterus villaviscensio
Eutoxeres aquila
Klais guimeti
Leucippus taczanowskii
Myrmia micrura
Phaethornis atrimentalis
Sephanoides sephaniodes
Chaetocercus mulsant
Eriocnemis vestita
Myrtis fanny
Oreotrochilus leucopleurus


# In case you end up modeling species-level data, create a species mean dataset
```{r}
# Figure out why there are NaNs! In at least a few cases it comes from only 1 individual per species
# drop species with fewer than 2? 
## You'll need to figure this out if you do continuous trait mapping

# species mean data
species.means <- final %>% group_by(species) %>%
                                  dplyr::summarize(
                                    mean_mass = mean(mass, na.rm = TRUE),
                                    mean_hb = mean(hb, na.rm=TRUE),
                                    mean_hct = mean(hct, na.rm=TRUE),
                                    mean_trbc = mean(trbc, na.rm=TRUE),
                                    mean_mcv = mean(mcv, na.rm=TRUE),
                                    mean_mch = mean(mch, na.rm=TRUE),
                                    mean_mchc = mean(mchc, na.rm=TRUE),
                                    mean_elev = mean(elev, na.rm=TRUE),
                                    elev_min = mean(elev_min, na.rm=TRUE), # Mean will just be elev_min cause all same 
                                    elev_max = mean(elev_max, na.rm=TRUE), # Mean will just be elev_max cause all same 
                                    range = mean(elev_range, na.rm=TRUE),  # Mean will just be elev_range cause all same 
                                    count=n())                              #Gives a count of total samples/species

# Assign species to clades 
species.means$Clade <- NA # Instantiate new col; capitalize this for easy and nice plot formatting

# add clade labels for 9 hummingbird clades

# EMERALDS
species.means$Clade[species.means$species=="Amazilia amazilia" | 
           species.means$species=="Amazilia chionogaster" | 
           species.means$species=="Amazilia franciae" | 
           species.means$species=="Amazilia lactea" | 
           species.means$species=="Amazilia viridicauda" | 
           species.means$species=="Chrysuronia oenone" | 
           species.means$species=="Taphrospilus hypostictus" | 
           species.means$species=="Thalurania furcata" | 
           species.means$species=="Campylopterus largipennis" | 
           species.means$species=="Campylopterus villaviscensio" | 
           species.means$species=="Leucippus taczanowskii" | 
           species.means$species=="Klais guimeti"] <- paste("Emerald")

# BEES 
species.means$Clade[species.means$species=="Calliphlox amethystina" | 
           species.means$species=="Thaumastura cora" | 
           species.means$species=="Chaetocercus mulsant" |
           species.means$species=="Myrmia micrura" |
           species.means$species=="Myrtis fanny" |
           species.means$species=="Rhodopis vesper"] <- paste("Bee") 

# MOUNTAIN GEMS 
species.means$Clade[species.means$species=="Heliomaster longirostris"] <- paste("Mountain-gem")

# PATAGONA 
species.means$Clade[species.means$species=="Patagona gigas"] <- paste("Patagona") 

# COQUETTES (Andean clade)
species.means$Clade[species.means$species=="Phlogophilus harterti" | 
           species.means$species=="Adelomyia melanogenys" | 
           species.means$species=="Aglaiocercus kingii" | 
           species.means$species=="Chalcostigma herrani" | 
           species.means$species=="Chalcostigma ruficeps" | 
           species.means$species=="Chalcostigma stanleyi" | 
           species.means$species=="Heliangelus amethysticollis" | 
           species.means$species=="Heliangelus micraster" | 
           species.means$species=="Heliangelus viola" | 
           species.means$species=="Metallura eupogon" | 
           species.means$species=="Metallura phoebe" | 
           species.means$species=="Metallura tyrianthina" | 
           species.means$species=="Oreotrochilus estella" | 
           species.means$species=="Oreotrochilus leucopleurus" | 
           species.means$species=="Oreotrochilus melanogaster" | 
           species.means$species=="Polyonymus caroli" | 
           species.means$species=="Sephanoides sephaniodes" | 
           species.means$species=="Oreonympha nobilis" |
           species.means$species=="Lesbia nuna" |
           species.means$species=="Lesbia victoriae"] <- paste("Coquette")

# BRILLIANTS (Andean clade)
species.means$Clade[species.means$species=="Aglaeactis castelnaudii" | 
           species.means$species=="Aglaeactis cupripennis" | 
           species.means$species=="Boissonneaua matthewsii" | 
           species.means$species=="Coeligena coeligena" | 
           species.means$species=="Coeligena iris" | 
           species.means$species=="Coeligena lutetiae" | 
           species.means$species=="Coeligena torquata" | 
           species.means$species=="Coeligena violifer" | 
           species.means$species=="Ensifera ensifera" | 
           species.means$species=="Eriocnemis aline" | 
           species.means$species=="Eriocnemis luciani" | 
           species.means$species=="Eriocnemis vestita" | 
           species.means$species=="Haplophaedia aureliae" | 
           species.means$species=="Heliodoxa aurescens" | 
           species.means$species=="Heliodoxa leadbeateri" | 
           species.means$species=="Heliodoxa rubinoides" | 
           species.means$species=="Lafresnaya lafresnayi" | 
           species.means$species=="Ocreatus underwoodii" | 
           species.means$species=="Pterophanes cyanopterus"] <- paste("Brilliant")

# MANGOS
species.means$Clade[species.means$species=="Anthracothorax nigricollis" | 
           species.means$species=="Colibri coruscans" | 
           species.means$species=="Colibri cyanotus" |  # Remember you changed Colibri thalassinus to cyanotus above 
           species.means$species=="Doryfera ludovicae" |
           species.means$species=="Schistes geoffroyi"] <- paste("Mango")

# HERMITS 
species.means$Clade[species.means$species=="Eutoxeres aquila" | 
           species.means$species=="Eutoxeres condamini" | 
           species.means$species=="Glaucis hirsutus" | 
           species.means$species=="Phaethornis atrimentalis" | 
           species.means$species=="Phaethornis guy" | 
           species.means$species=="Phaethornis hispidus" | 
           species.means$species=="Phaethornis malaris" | 
           species.means$species=="Phaethornis philippii" | 
           species.means$species=="Phaethornis ruber" | 
           species.means$species=="Phaethornis stuarti" | 
           species.means$species=="Phaethornis syrmatophorus" | 
           species.means$species=="Threnetes leucurus"] <- paste("Hermit")

# TOPAZES AND JACOBINS 
species.means$Clade[species.means$species=="Florisuga mellivora"] <- paste("Topaz") 


# Create order based on phylogeny branch order (for elev_range plotting)
species.means$phylo_order <- NA # Instantiate new col; capitalize this for easy and nice plot formatting

species.means$phylo_order[species.means$species=="Phlogophilus harterti"] <- paste(.1) # Weird decimals only way to keep order
species.means$phylo_order[species.means$species=="Oreotrochilus leucopleurus"] <- paste(.2)
species.means$phylo_order[species.means$species=="Oreotrochilus estella"] <- paste(.3)
species.means$phylo_order[species.means$species=="Oreotrochilus melanogaster"] <- paste(.4)
species.means$phylo_order[species.means$species=="Polyonymus caroli"] <- paste(.5)
species.means$phylo_order[species.means$species=="Chalcostigma herrani"] <- paste(.6)
species.means$phylo_order[species.means$species=="Chalcostigma ruficeps"] <- paste(.7)
species.means$phylo_order[species.means$species=="Oreonympha nobilis"] <- paste(.8)
species.means$phylo_order[species.means$species=="Chalcostigma stanleyi"] <- paste(.9)
species.means$phylo_order[species.means$species=="Metallura eupogon"] <- paste(10)
species.means$phylo_order[species.means$species=="Metallura tyrianthina"] <- paste(11)
species.means$phylo_order[species.means$species=="Metallura phoebe"] <- paste(12)
species.means$phylo_order[species.means$species=="Lesbia nuna"] <- paste(13)
species.means$phylo_order[species.means$species=="Lesbia victoriae"] <- paste(14)
species.means$phylo_order[species.means$species=="Adelomyia melanogenys"] <- paste(15)
species.means$phylo_order[species.means$species=="Aglaiocercus kingii"] <- paste(16)
species.means$phylo_order[species.means$species=="Heliangelus amethysticollis"] <- paste(17)
species.means$phylo_order[species.means$species=="Heliangelus micraster"] <- paste(18)
species.means$phylo_order[species.means$species=="Heliangelus viola"] <- paste(19)
species.means$phylo_order[species.means$species=="Sephanoides sephaniodes"] <- paste(20) # Coquettes  

species.means$phylo_order[species.means$species=="Lafresnaya lafresnayi"] <- paste(21)
species.means$phylo_order[species.means$species=="Aglaeactis castelnaudii"] <- paste(22)
species.means$phylo_order[species.means$species=="Aglaeactis cupripennis"] <- paste(23)
species.means$phylo_order[species.means$species=="Heliodoxa leadbeateri"] <- paste(24)
species.means$phylo_order[species.means$species=="Heliodoxa rubinoides"] <- paste(25)
species.means$phylo_order[species.means$species=="Heliodoxa aurescens"] <- paste(26)
species.means$phylo_order[species.means$species=="Boissonneaua matthewsii"] <- paste(27)
species.means$phylo_order[species.means$species=="Ocreatus underwoodii"] <- paste(28)
species.means$phylo_order[species.means$species=="Pterophanes cyanopterus"] <- paste(29)
species.means$phylo_order[species.means$species=="Ensifera ensifera"] <- paste(30)
species.means$phylo_order[species.means$species=="Coeligena coeligena"] <- paste(31)
species.means$phylo_order[species.means$species=="Coeligena violifer"] <- paste(32)
species.means$phylo_order[species.means$species=="Coeligena lutetiae"] <- paste(33)
species.means$phylo_order[species.means$species=="Coeligena iris"] <- paste(34)
species.means$phylo_order[species.means$species=="Coeligena torquata"] <- paste(35)
species.means$phylo_order[species.means$species=="Eriocnemis vestita"] <- paste(36)
species.means$phylo_order[species.means$species=="Eriocnemis luciani"] <- paste(37)
species.means$phylo_order[species.means$species=="Eriocnemis aline"] <- paste(38)
species.means$phylo_order[species.means$species=="Haplophaedia aureliae"] <- paste(39) # Brilliants

species.means$phylo_order[species.means$species=="Patagona gigas"] <- paste(40) # Patagona

species.means$phylo_order[species.means$species=="Calliphlox amethystina"] <- paste(41)
species.means$phylo_order[species.means$species=="Rhodopis vesper"] <- paste(42) 
species.means$phylo_order[species.means$species=="Myrtis fanny"] <- paste(43) 
species.means$phylo_order[species.means$species=="Thaumastura cora"] <- paste(44) 
species.means$phylo_order[species.means$species=="Myrmia micrura"] <- paste(45) 
species.means$phylo_order[species.means$species=="Chaetocercus mulsant"] <- paste(46) # Bees 

species.means$phylo_order[species.means$species=="Heliomaster longirostris"] <- paste(47) # Mountain-gems 

species.means$phylo_order[species.means$species=="Campylopterus villaviscensio"] <- paste(48) 
species.means$phylo_order[species.means$species=="Campylopterus largipennis"] <- paste(49) 
species.means$phylo_order[species.means$species=="Klais guimeti"] <- paste(50) 
species.means$phylo_order[species.means$species=="Amazilia lactea"] <- paste(51) 
species.means$phylo_order[species.means$species=="Amazilia viridicauda"] <- paste(52) 
species.means$phylo_order[species.means$species=="Amazilia chionogaster"] <- paste(53) 
species.means$phylo_order[species.means$species=="Chrysuronia oenone"] <- paste(54) 
species.means$phylo_order[species.means$species=="Amazilia franciae"] <- paste(55) 
species.means$phylo_order[species.means$species=="Amazilia amazilia"] <- paste(56) 
species.means$phylo_order[species.means$species=="Leucippus taczanowskii"] <- paste(57) 
species.means$phylo_order[species.means$species=="Taphrospilus hypostictus"] <- paste(58) 
species.means$phylo_order[species.means$species=="Thalurania furcata"] <- paste(59) # Emeralds

species.means$phylo_order[species.means$species=="Anthracothorax nigricollis"] <- paste(60) 
species.means$phylo_order[species.means$species=="Doryfera ludovicae"] <- paste(61) 
species.means$phylo_order[species.means$species=="Colibri cyanotus"] <- paste(62) 
species.means$phylo_order[species.means$species=="Colibri coruscans"] <- paste(63) 
species.means$phylo_order[species.means$species=="Schistes geoffroyi"] <- paste(64) # Mangos

species.means$phylo_order[species.means$species=="Florisuga mellivora"] <- paste(65) # Topazes

species.means$phylo_order[species.means$species=="Eutoxeres condamini"] <- paste(66) 
species.means$phylo_order[species.means$species=="Eutoxeres aquila"] <- paste(67) 
species.means$phylo_order[species.means$species=="Glaucis hirsutus"] <- paste(68) 
species.means$phylo_order[species.means$species=="Threnetes leucurus"] <- paste(69) 
species.means$phylo_order[species.means$species=="Phaethornis guy"] <- paste(70) 
species.means$phylo_order[species.means$species=="Phaethornis malaris"] <- paste(71) 
species.means$phylo_order[species.means$species=="Phaethornis syrmatophorus"] <- paste(72) 
species.means$phylo_order[species.means$species=="Phaethornis hispidus"] <- paste(73)
species.means$phylo_order[species.means$species=="Phaethornis philippii"] <- paste(74) 
species.means$phylo_order[species.means$species=="Phaethornis stuarti"] <- paste(75) 
species.means$phylo_order[species.means$species=="Phaethornis atrimentalis"] <- paste(76) 
species.means$phylo_order[species.means$species=="Phaethornis ruber"] <- paste(77) # Hermits

# Consider adding elev_position, BioClim variables, temp and precip PCAs, other stuff
# If you do want to use this and color by clades, need to manually add in clades 

# Calculate elevation range midpoints and midpoint offsets: 
species.means$elevmidpt <- species.means$range/2 # Make variable for elevational midpoint for each species 
species.means$elevmidpt_offset <- species.means$elevmidpt + species.means$elev_min # IMPORTANT!!!!
# Because midpoints will be plotted with correct numbers but on full elev x-axis scale, midpoints don't align properly w/in 
# their associated elev ranges (i.e. if range = 1000-2000, midpoint=500, but point will be plotted AT 500 instead of 1,500)
# To correct for this, I'll add the elev_min "offset" to each midpoint, which makes its starting point the elev_min #
# After testing this out, thins look correct: All elev midpoints plot accurately at their midpoints within their elev range (!)

write.csv(species.means, "AllHummingbird_SpeciesMeanBloodData_09-17-20.csv")


# Now make copy of 'final' dataset that you can finagle to get offset collection elevations to represent sampling: 
sampling <- final 

# Add phylo order to this big samping dataset
sampling$phylo_order <- NA # Instantiate new col; capitalize this for easy and nice plot formatting

sampling$phylo_order[sampling$species=="Phlogophilus harterti"] <- paste(.1) # Weird decimals only way to keep order
sampling$phylo_order[sampling$species=="Oreotrochilus leucopleurus"] <- paste(.2)
sampling$phylo_order[sampling$species=="Oreotrochilus estella"] <- paste(.3)
sampling$phylo_order[sampling$species=="Oreotrochilus melanogaster"] <- paste(.4)
sampling$phylo_order[sampling$species=="Polyonymus caroli"] <- paste(.5)
sampling$phylo_order[sampling$species=="Chalcostigma herrani"] <- paste(.6)
sampling$phylo_order[sampling$species=="Chalcostigma ruficeps"] <- paste(.7)
sampling$phylo_order[sampling$species=="Oreonympha nobilis"] <- paste(.8)
sampling$phylo_order[sampling$species=="Chalcostigma stanleyi"] <- paste(.9)
sampling$phylo_order[sampling$species=="Metallura eupogon"] <- paste(10)
sampling$phylo_order[sampling$species=="Metallura tyrianthina"] <- paste(11)
sampling$phylo_order[sampling$species=="Metallura phoebe"] <- paste(12)
sampling$phylo_order[sampling$species=="Lesbia nuna"] <- paste(13)
sampling$phylo_order[sampling$species=="Lesbia victoriae"] <- paste(14)
sampling$phylo_order[sampling$species=="Adelomyia melanogenys"] <- paste(15)
sampling$phylo_order[sampling$species=="Aglaiocercus kingii"] <- paste(16)
sampling$phylo_order[sampling$species=="Heliangelus amethysticollis"] <- paste(17)
sampling$phylo_order[sampling$species=="Heliangelus micraster"] <- paste(18)
sampling$phylo_order[sampling$species=="Heliangelus viola"] <- paste(19)
sampling$phylo_order[sampling$species=="Sephanoides sephaniodes"] <- paste(20) # Coquettes  

sampling$phylo_order[sampling$species=="Lafresnaya lafresnayi"] <- paste(21)
sampling$phylo_order[sampling$species=="Aglaeactis castelnaudii"] <- paste(22)
sampling$phylo_order[sampling$species=="Aglaeactis cupripennis"] <- paste(23)
sampling$phylo_order[sampling$species=="Heliodoxa leadbeateri"] <- paste(24)
sampling$phylo_order[sampling$species=="Heliodoxa rubinoides"] <- paste(25)
sampling$phylo_order[sampling$species=="Heliodoxa aurescens"] <- paste(26)
sampling$phylo_order[sampling$species=="Boissonneaua matthewsii"] <- paste(27)
sampling$phylo_order[sampling$species=="Ocreatus underwoodii"] <- paste(28)
sampling$phylo_order[sampling$species=="Pterophanes cyanopterus"] <- paste(29)
sampling$phylo_order[sampling$species=="Ensifera ensifera"] <- paste(30)
sampling$phylo_order[sampling$species=="Coeligena coeligena"] <- paste(31)
sampling$phylo_order[sampling$species=="Coeligena violifer"] <- paste(32)
sampling$phylo_order[sampling$species=="Coeligena lutetiae"] <- paste(33)
sampling$phylo_order[sampling$species=="Coeligena iris"] <- paste(34)
sampling$phylo_order[sampling$species=="Coeligena torquata"] <- paste(35)
sampling$phylo_order[sampling$species=="Eriocnemis vestita"] <- paste(36)
sampling$phylo_order[sampling$species=="Eriocnemis luciani"] <- paste(37)
sampling$phylo_order[sampling$species=="Eriocnemis aline"] <- paste(38)
sampling$phylo_order[sampling$species=="Haplophaedia aureliae"] <- paste(39) # Brilliants

sampling$phylo_order[sampling$species=="Patagona gigas"] <- paste(40) # Patagona

sampling$phylo_order[sampling$species=="Calliphlox amethystina"] <- paste(41)
sampling$phylo_order[sampling$species=="Rhodopis vesper"] <- paste(42) 
sampling$phylo_order[sampling$species=="Myrtis fanny"] <- paste(43) 
sampling$phylo_order[sampling$species=="Thaumastura cora"] <- paste(44) 
sampling$phylo_order[sampling$species=="Myrmia micrura"] <- paste(45) 
sampling$phylo_order[sampling$species=="Chaetocercus mulsant"] <- paste(46) # Bees 

sampling$phylo_order[sampling$species=="Heliomaster longirostris"] <- paste(47) # Mountain-gems 

sampling$phylo_order[sampling$species=="Campylopterus villaviscensio"] <- paste(48) 
sampling$phylo_order[sampling$species=="Campylopterus largipennis"] <- paste(49) 
sampling$phylo_order[sampling$species=="Klais guimeti"] <- paste(50) 
sampling$phylo_order[sampling$species=="Amazilia lactea"] <- paste(51) 
sampling$phylo_order[sampling$species=="Amazilia viridicauda"] <- paste(52) 
sampling$phylo_order[sampling$species=="Amazilia chionogaster"] <- paste(53) 
sampling$phylo_order[sampling$species=="Chrysuronia oenone"] <- paste(54) 
sampling$phylo_order[sampling$species=="Amazilia franciae"] <- paste(55) 
sampling$phylo_order[sampling$species=="Amazilia amazilia"] <- paste(56) 
sampling$phylo_order[sampling$species=="Leucippus taczanowskii"] <- paste(57) 
sampling$phylo_order[sampling$species=="Taphrospilus hypostictus"] <- paste(58) 
sampling$phylo_order[sampling$species=="Thalurania furcata"] <- paste(59) # Emeralds

sampling$phylo_order[sampling$species=="Anthracothorax nigricollis"] <- paste(60) 
sampling$phylo_order[sampling$species=="Doryfera ludovicae"] <- paste(61) 
sampling$phylo_order[sampling$species=="Colibri cyanotus"] <- paste(62) 
sampling$phylo_order[sampling$species=="Colibri coruscans"] <- paste(63) 
sampling$phylo_order[sampling$species=="Schistes geoffroyi"] <- paste(64) # Mangos

sampling$phylo_order[sampling$species=="Florisuga mellivora"] <- paste(65) # Topazes

sampling$phylo_order[sampling$species=="Eutoxeres condamini"] <- paste(66) 
sampling$phylo_order[sampling$species=="Eutoxeres aquila"] <- paste(67) 
sampling$phylo_order[sampling$species=="Glaucis hirsutus"] <- paste(68) 
sampling$phylo_order[sampling$species=="Threnetes leucurus"] <- paste(69) 
sampling$phylo_order[sampling$species=="Phaethornis guy"] <- paste(70) 
sampling$phylo_order[sampling$species=="Phaethornis malaris"] <- paste(71) 
sampling$phylo_order[sampling$species=="Phaethornis syrmatophorus"] <- paste(72) 
sampling$phylo_order[sampling$species=="Phaethornis hispidus"] <- paste(73)
sampling$phylo_order[sampling$species=="Phaethornis philippii"] <- paste(74) 
sampling$phylo_order[sampling$species=="Phaethornis stuarti"] <- paste(75) 
sampling$phylo_order[sampling$species=="Phaethornis atrimentalis"] <- paste(76) 
sampling$phylo_order[sampling$species=="Phaethornis ruber"] <- paste(77) # Hermits

#### playing around, this needs to go in plots and figs script

#sort by complete ENSM magnitude (descending)
species.means$phylo_order <- as.numeric(species.means$phylo_order) # Need to make this numeric
sampling$phylo_order <- as.numeric(sampling$phylo_order) # Need to make this numeric
# species.means$range <- as.numeric(species.means$range)

# Species elevational ranges in order of phylogeny 
p1 <- ggplot(data=species.means, aes(y=range, x=reorder(phylo_order, -phylo_order))) + # Add fill=Clade to color by clade 
            # x=reorder is saying "reorder common_name on x-axis by magnitude of elev shift in DESCENDING"
      # geom_bar(stat="identity", position=position_dodge(), width=0.2) + # trick: don't plot these! 
     # geom_line(aes(ymin=elev_min, ymax=elev_max)) +
      geom_errorbar(aes(ymin=elev_min, ymax=elev_max), width=0.5) + # Specifies exact elev range; width controls error bar caps
      geom_point(data=sampling, position=position_jitter(width=0.08, height=0.5), # SAMPLING POINTS, jittered 
                 # WIDTH=height and HEIGHT=width w/ coordflip
                 aes(y=elev, x=reorder(phylo_order, -phylo_order)), 
                 shape=16, color="gray28", alpha=0.5, size=1.0) + # or gray28 for a bit lighter? 
      geom_point(aes(y=elevmidpt_offset, x=reorder(phylo_order, -phylo_order)), # ELEVATION MIDPOINT OF RANGE 
                 shape=18, color="firebrick1", size=1.4) +
      coord_flip() + # This flips x and y axes
      scale_y_continuous(position="right") + # Because of coord flip, "right" moves x-axis to top of plot 
      theme(axis.text.x=element_text(angle=0, hjust=0.5),
           panel.grid.major=element_blank(), 
           panel.grid.minor=element_blank(),
           panel.background = element_blank()) +
      labs(title="",x="", y = "Elevational Range (m)") + # y = species in phylo order 
      theme(plot.title = element_text(face="bold")) + # This makes panel header bold 
      # This is good for labeling figure panels! Avoids having to manually toy w/ hjust and vjust
      theme(axis.text.y=element_text(size=8), axis.text.x=element_text(size=10), axis.title=element_text(size=8))
print(p1)
ggsave(p1, filename="AllSpecies_ElevRange_PhylogenyOrder_ForDoctoringInIllustrator_WholeRange_09-17-20.pdf", height=6.35, width=2.8, units="in")
# Same units as phylogeny for easy wrangling? 

#####
# TRIAL AND ERROR NOTES ON PLOTTING ELEV RANGES: 
# geom_pointrante() plots the midpoint of your line segment IF the ymin and ymax are -/+ some number from the y variable
# ymin and ymax are meant to represent standard deviation, so must be manipulating y variable somehow
# If you do like I need to and plot ymin and ymax and entirely separate variables, the "midpoint" of geom_pointrange is off 
# Example of geom_pointrange working correctly: 
# ggplot(data=species.means) +
#   aes(
#    x = phylo_order, # reorder(phylo_order, -phylo_order),
#    y = range,
#    ymin = range-500, # BUT, if I specify ymin=elev_min and ymax=elev_max, points are all wonky 
#    ymax = range+500
#   ) +
#   geom_pointrange()

# Help on coord_flip and moving axis positions after coord_flip:
# https://stackoverflow.com/questions/63863398/in-ggplot2-how-can-i-move-the-x-axis-on-top-in-this-case-then-rotate-the-axes-w
# change geom_bar style; make thin w/ caps at ends

```

Helpful data summarizing tips: http://kbroman.org/datacarpentry_R_2016-06-01/03-dplyr.html



# READ OUT FINAL DATASETS FOR DOWNSTREAM ANALYSES: 
Since Patagona is a huge body size outlier, make a non-Patagona subset to be able to do some body size comparisons below
```{r}
# FINAL FINAL: This is the 'final' dataset w/ PCA reduced-BioClimVars
write.csv(final, "AllHummingbirdBlood_Wrangled_withBioClim_andPCA_10-27-20.csv")

# Patagona-free subset (for possible modeling of mass, where Patagona is a huge body size outlier):
final.nopgig <- final[-which(final$species == "Patagona gigas"),] # 161 Patagona observations 
write.csv(final.nopgig, "HumBlood.final.nopgig.csv")
```



###########



## END 